// @totvs/thf-ui/esm2015/lib/components/thf-field/thf-rich-text/thf-rich-text-body/thf-rich-text-body.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ElementRef, EventEmitter, Input, Output, ViewChild } from '@angular/core';
/** @type {?} */
const thfRichTextBodyCommands = [
    'bold', 'italic', 'underline', 'justifyleft', 'justifycenter', 'justifyright', 'justifyfull', 'insertUnorderedList'
];
export class ThfRichTextBodyComponent {
    constructor() {
        this.commands = new EventEmitter();
        this.value = new EventEmitter();
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this.bodyElement.nativeElement.designMode = 'on';
        // timeout necessário para setar o valor vindo do writeValue do componente principal.
        setTimeout((/**
         * @return {?}
         */
        () => { this.updateValueWithModelValue(); }));
    }
    /**
     * @param {?} command
     * @return {?}
     */
    executeCommand(command) {
        this.bodyElement.nativeElement.focus();
        document.execCommand(command, false, null);
        this.updateModel();
        this.value.emit(this.modelValue);
    }
    /**
     * @return {?}
     */
    onClick() {
        this.emitSelectionCommands();
    }
    /**
     * @return {?}
     */
    onKeyUp() {
        // Tratamento necessário para eliminar a tag <br> criada no firefox quando o body for limpo.
        if (!this.bodyElement.nativeElement.innerText.trim() && this.bodyElement.nativeElement.firstChild) {
            this.bodyElement.nativeElement.firstChild.remove();
        }
        this.updateModel();
        this.emitSelectionCommands();
    }
    /**
     * @return {?}
     */
    update() {
        setTimeout((/**
         * @return {?}
         */
        () => { this.updateModel(); }));
        setTimeout((/**
         * @return {?}
         */
        () => { this.onKeyUp(); }));
    }
    /**
     * @private
     * @return {?}
     */
    emitSelectionCommands() {
        /** @type {?} */
        const commands = thfRichTextBodyCommands.filter((/**
         * @param {?} command
         * @return {?}
         */
        command => {
            return document.queryCommandState(command);
        }));
        this.commands.emit(commands);
    }
    /**
     * @private
     * @return {?}
     */
    updateModel() {
        this.modelValue = this.bodyElement.nativeElement.innerHTML;
        this.value.emit(this.modelValue);
    }
    /**
     * @private
     * @return {?}
     */
    updateValueWithModelValue() {
        if (this.modelValue) {
            this.bodyElement.nativeElement.insertAdjacentHTML('afterbegin', this.modelValue);
        }
    }
}
ThfRichTextBodyComponent.decorators = [
    { type: Component, args: [{
                selector: 'thf-rich-text-body',
                template: "<div #bodyElement\n  class=\"thf-rich-text-body\"\n  [attr.contenteditable]=\"!readonly\"\n  [attr.data-placeholder]=\"placeholder\"\n  [style.height.px]=\"height\"\n  (click)=\"onClick()\"\n  (cut)=\"update()\"\n  (keyup)=\"onKeyUp()\"\n  (paste)=\"update()\">\n</div>\n"
            }] }
];
ThfRichTextBodyComponent.propDecorators = {
    bodyElement: [{ type: ViewChild, args: ['bodyElement', { static: true },] }],
    height: [{ type: Input, args: ['t-height',] }],
    modelValue: [{ type: Input, args: ['t-model-value',] }],
    placeholder: [{ type: Input, args: ['t-placeholder',] }],
    readonly: [{ type: Input, args: ['t-readonly',] }],
    commands: [{ type: Output, args: ['t-commands',] }],
    value: [{ type: Output, args: ['t-value',] }]
};
if (false) {
    /** @type {?} */
    ThfRichTextBodyComponent.prototype.bodyElement;
    /** @type {?} */
    ThfRichTextBodyComponent.prototype.height;
    /** @type {?} */
    ThfRichTextBodyComponent.prototype.modelValue;
    /** @type {?} */
    ThfRichTextBodyComponent.prototype.placeholder;
    /** @type {?} */
    ThfRichTextBodyComponent.prototype.readonly;
    /** @type {?} */
    ThfRichTextBodyComponent.prototype.commands;
    /** @type {?} */
    ThfRichTextBodyComponent.prototype.value;
}
