// @totvs/thf-ui/esm2015/lib/components/thf-modal/thf-modal.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ChangeDetectorRef, Component, ElementRef, Renderer2, ViewChild } from '@angular/core';
import { v4 as uuid } from 'uuid';
import { ThfModalBaseComponent } from './thf-modal-base.component';
import { ThfModalService } from './thf-modal-service';
/**
 * \@docsExtends ThfModalBaseComponent
 *
 * \@example
 *
 * <example name="thf-modal-basic" title="Totvs Modal Basic">
 *  <file name="sample-thf-modal-basic/sample-thf-modal-basic.component.html"> </file>
 *  <file name="sample-thf-modal-basic/sample-thf-modal-basic.component.ts"> </file>
 * </example>
 *
 * <example name="thf-modal-labs" title="Totvs Modal Labs">
 *  <file name="sample-thf-modal-labs/sample-thf-modal-labs.component.html"> </file>
 *  <file name="sample-thf-modal-labs/sample-thf-modal-labs.component.ts"> </file>
 * </example>
 *
 * <example name="thf-modal-fruits-salad" title="Totvs Modal - Fruits Salad">
 *  <file name="sample-thf-modal-fruits-salad/sample-thf-modal-fruits-salad.component.html"> </file>
 *  <file name="sample-thf-modal-fruits-salad/sample-thf-modal-fruits-salad.component.ts"> </file>
 * </example>
 */
export class ThfModalComponent extends ThfModalBaseComponent {
    /**
     * @param {?} thfModalService
     * @param {?} renderer
     * @param {?} changeDetector
     */
    constructor(thfModalService, renderer, changeDetector) {
        super();
        this.thfModalService = thfModalService;
        this.renderer = renderer;
        this.changeDetector = changeDetector;
        this.focusableElements = 'input, select, textarea, button:not([disabled]), a';
        this.id = uuid();
    }
    /**
     * @param {?=} xClosed
     * @return {?}
     */
    close(xClosed = false) {
        this.thfModalService.modalActive = undefined;
        super.close(xClosed);
        this.removeEventListeners();
        if (this.sourceElement) {
            this.sourceElement.focus();
        }
    }
    /**
     * @param {?} event
     * @return {?}
     */
    closeModalOnEscapeKey(event) {
        if (!this.hideClose) {
            event.preventDefault();
            event.stopPropagation();
            this.close();
        }
    }
    /**
     * @return {?}
     */
    getPrimaryActionButtonType() {
        return this.primaryAction.danger ? 'danger' : 'primary';
    }
    /**
     * @return {?}
     */
    getSecondaryActionButtonType() {
        return this.secondaryAction && this.secondaryAction.danger && !this.primaryAction.danger ? 'danger' : 'default';
    }
    /**
     * @param {?} event
     * @return {?}
     */
    onClickOut(event) {
        if (this.clickOut && !this.modalContent.nativeElement.contains(event.target)) {
            this.close();
        }
    }
    /**
     * @return {?}
     */
    open() {
        this.sourceElement = document.activeElement;
        super.open();
        this.handleFocus();
    }
    /**
     * @private
     * @return {?}
     */
    handleFocus() {
        this.thfModalService.modalActive = this.id;
        setTimeout((/**
         * @return {?}
         */
        () => {
            if (this.modalContent) {
                this.initFocus();
                document.addEventListener('focus', this.focusFunction, true);
            }
        }));
    }
    /**
     * @private
     * @return {?}
     */
    initFocus() {
        this.focusFunction = (/**
         * @param {?} event
         * @return {?}
         */
        (event) => {
            this.thfModalService.modalActive = this.thfModalService.modalActive || this.id;
            /** @type {?} */
            const modalElement = this.modalContent.nativeElement;
            if (!modalElement.contains(event.target) && this.thfModalService.modalActive === this.id) {
                event.stopPropagation();
                this.firstElement.focus();
            }
        });
        this.setFirstElement();
        if (this.hideClose) {
            this.firstElement.focus();
        }
        else {
            /** @type {?} */
            const firstFieldElement = this.modalContent.nativeElement.querySelectorAll(this.focusableElements)[1] ||
                this.modalContent.nativeElement;
            firstFieldElement.focus();
        }
    }
    /**
     * @private
     * @return {?}
     */
    removeEventListeners() {
        document.removeEventListener('focus', this.focusFunction, true);
    }
    /**
     * @private
     * @return {?}
     */
    setFirstElement() {
        this.firstElement = this.modalContent.nativeElement.querySelector(this.focusableElements) || this.modalContent.nativeElement;
    }
}
ThfModalComponent.decorators = [
    { type: Component, args: [{
                selector: 'thf-modal',
                template: "<div *ngIf=\"!isHidden\"\n  class=\"thf-modal\"\n  tabindex=\"0\"\n  (keydown.esc)=\"closeModalOnEscapeKey($event)\">\n\n  <div class=\"thf-modal-overlay\">\n    <div class=\"thf-modal-container thf-pb-2 thf-pt-2\" (click)=\"onClickOut($event)\">\n\n      <div class=\"thf-modal-vertical-align\">\n        <div #modalContent\n          class=\"thf-modal-content thf-modal-{{ size }}\"\n          tabindex=\"-1\">\n\n          <div class=\"thf-modal-header\">\n            <div class=\"thf-modal-title thf-text-ellipsis\">\n              {{ title }}\n            </div>\n\n            <a *ngIf=\"!hideClose\"\n              class=\"thf-modal-header-close-button\"\n              tabindex=\"0\"\n              (click)=\"close(true)\">\n              <span class=\"thf-icon thf-icon-close\"></span>\n            </a>\n          </div>\n\n          <div class=\"thf-modal-body\">\n            <ng-content></ng-content>\n          </div>\n\n          <div class=\"thf-modal-footer\">\n            <thf-button *ngIf=\"secondaryAction\"\n              [t-disabled]=\"secondaryAction.disabled\"\n              [t-label]=\"secondaryAction.label\"\n              [t-loading]=\"secondaryAction.loading\"\n              [t-type]=\"getSecondaryActionButtonType()\"\n              (t-click)=\"secondaryAction.action()\">\n            </thf-button>\n\n            <thf-button\n              class=\"thf-button-modal-first-action\"\n              [t-disabled]=\"primaryAction.disabled\"\n              [t-label]=\"primaryAction.label\"\n              [t-loading]=\"primaryAction.loading\"\n              [t-type]=\"getPrimaryActionButtonType()\"\n              (t-click)=\"primaryAction.action()\">\n            </thf-button>\n          </div>\n\n        </div>\n      </div>\n\n    </div>\n  </div>\n</div>\n\n"
            }] }
];
/** @nocollapse */
ThfModalComponent.ctorParameters = () => [
    { type: ThfModalService },
    { type: Renderer2 },
    { type: ChangeDetectorRef }
];
ThfModalComponent.propDecorators = {
    modalContent: [{ type: ViewChild, args: ['modalContent', { read: ElementRef, static: false },] }]
};
if (false) {
    /** @type {?} */
    ThfModalComponent.prototype.modalContent;
    /**
     * @type {?}
     * @private
     */
    ThfModalComponent.prototype.firstElement;
    /**
     * @type {?}
     * @private
     */
    ThfModalComponent.prototype.focusFunction;
    /**
     * @type {?}
     * @private
     */
    ThfModalComponent.prototype.focusableElements;
    /**
     * @type {?}
     * @private
     */
    ThfModalComponent.prototype.id;
    /**
     * @type {?}
     * @private
     */
    ThfModalComponent.prototype.sourceElement;
    /** @type {?} */
    ThfModalComponent.prototype.onResizeListener;
    /**
     * @type {?}
     * @private
     */
    ThfModalComponent.prototype.thfModalService;
    /**
     * @type {?}
     * @private
     */
    ThfModalComponent.prototype.renderer;
    /**
     * @type {?}
     * @private
     */
    ThfModalComponent.prototype.changeDetector;
}
