// @totvs/thf-ui/esm5/lib/components/thf-toolbar/thf-toolbar.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { ThfAvatarModule } from './../thf-avatar/thf-avatar.module';
import { ThfPopupModule } from '../thf-popup/thf-popup.module';
import { ThfToolbarActionsComponent } from './thf-toolbar-actions/thf-toolbar-actions.component';
import { ThfToolbarComponent } from './thf-toolbar.component';
import { ThfToolbarNotificationComponent } from './thf-toolbar-notification/thf-toolbar-notification.component';
import { ThfToolbarProfileComponent } from './thf-toolbar-profile/thf-toolbar-profile.component';
/**
 * \@description
 *
 * Módulo do componente thf-toolbar
 *
 */
var ThfToolbarModule = /** @class */ (function () {
    function ThfToolbarModule() {
    }
    ThfToolbarModule.decorators = [
        { type: NgModule, args: [{
                    imports: [
                        CommonModule,
                        ThfAvatarModule,
                        ThfPopupModule
                    ],
                    declarations: [
                        ThfToolbarActionsComponent,
                        ThfToolbarComponent,
                        ThfToolbarNotificationComponent,
                        ThfToolbarProfileComponent
                    ],
                    exports: [
                        ThfToolbarComponent
                    ],
                    providers: [],
                    schemas: []
                },] }
    ];
    return ThfToolbarModule;
}());
export { ThfToolbarModule };
