// @totvs/thf-templates/esm2015/lib/components/thf-page-job-scheduler/thf-page-job-scheduler-parameters/thf-page-job-scheduler-parameters.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, EventEmitter, Input, Output, ViewChild } from '@angular/core';
import { NgForm } from '@angular/forms';
export class ThfPageJobSchedulerParametersComponent {
    constructor() {
        this.literals = (/** @type {?} */ ({}));
        this.parameters = [];
        this.valueChange = new EventEmitter();
    }
    /**
     * @return {?}
     */
    ngAfterViewInit() {
        if (this.form) {
            setTimeout((/**
             * @return {?}
             */
            () => {
                this.form.valueChanges.subscribe((/**
                 * @param {?} value
                 * @return {?}
                 */
                value => {
                    this.valueChange.emit(value);
                }));
            }));
        }
    }
}
ThfPageJobSchedulerParametersComponent.decorators = [
    { type: Component, args: [{
                selector: 'thf-page-job-scheduler-parameters',
                template: "<ng-container *ngIf=\"parameters && parameters.length; then formFieldsTemplate; else parametersNotFoundTemplate\">\n</ng-container>\n\n<ng-template #parametersNotFoundTemplate>\n\n  <div class=\"thf-text-center\">\n    <span class=\"thf-icon thf-icon-info\"></span>\n    <span class=\"thf-font-text-large\">\n      {{ literals.parametersNotFound }}\n    </span>\n  </div>\n\n</ng-template>\n\n<ng-template #formFieldsTemplate>\n  <form #parametersForm=\"ngForm\">\n    <thf-dynamic-form\n      t-group-form\n      [t-fields]=\"parameters\"\n      [t-value]=\"value\">\n    </thf-dynamic-form>\n  </form>\n</ng-template>\n"
            }] }
];
ThfPageJobSchedulerParametersComponent.propDecorators = {
    form: [{ type: ViewChild, args: ['parametersForm', { static: true },] }],
    literals: [{ type: Input, args: ['t-literals',] }],
    parameters: [{ type: Input, args: ['t-parameters',] }],
    value: [{ type: Input, args: ['t-value',] }],
    valueChange: [{ type: Output, args: ['t-valueChange',] }]
};
if (false) {
    /** @type {?} */
    ThfPageJobSchedulerParametersComponent.prototype.form;
    /** @type {?} */
    ThfPageJobSchedulerParametersComponent.prototype.literals;
    /** @type {?} */
    ThfPageJobSchedulerParametersComponent.prototype.parameters;
    /** @type {?} */
    ThfPageJobSchedulerParametersComponent.prototype.value;
    /** @type {?} */
    ThfPageJobSchedulerParametersComponent.prototype.valueChange;
}
