// @totvs/thf-templates/esm2015/lib/components/thf-page-blocked-user/thf-page-blocked-user.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ActivatedRoute, Router } from '@angular/router';
import { Component } from '@angular/core';
import { browserLanguage, isExternalLink, thfLocaleDefault } from '../../utils/util';
import { ThfPageBlockedUserBaseComponent } from './thf-page-blocked-user-base.component';
/** @type {?} */
export const thfPageBlockedUserButtonLiterals = {
    en: (/** @type {?} */ ({
        primaryButton: 'Back to home screen',
    })),
    es: (/** @type {?} */ ({
        primaryButton: 'Volver al inicio',
    })),
    pt: (/** @type {?} */ ({
        primaryButton: 'Voltar para o início',
    }))
};
/**
 * \@docsExtends ThfPageBlockedUserBaseComponent
 *
 * \@example
 *
 * <example name="thf-page-blocked-user-basic" title="Totvs Page Blocked User Basic">
 *  <file name="sample-thf-page-blocked-user-basic/sample-thf-page-blocked-user-basic.component.html"> </file>
 *  <file name="sample-thf-page-blocked-user-basic/sample-thf-page-blocked-user-basic.component.ts"> </file>
 * </example>
 *
 * <example name="thf-page-blocked-user-labs" title="Totvs Page Blocked User Labs">
 *  <file name="sample-thf-page-blocked-user-labs/sample-thf-page-blocked-user-labs.component.html"> </file>
 *  <file name="sample-thf-page-blocked-user-labs/sample-thf-page-blocked-user-labs.component.ts"> </file>
 * </example>
 *
 * <example name="thf-page-blocked-user-exceeded-attempts" title="Totvs Page Blocked User Exceeded Attempts">
 *  <file name="sample-thf-page-blocked-user-exceeded-attempts/sample-thf-page-blocked-user-exceeded-attempts.component.html"> </file>
 *  <file name="sample-thf-page-blocked-user-exceeded-attempts/sample-thf-page-blocked-user-exceeded-attempts.component.ts"> </file>
 * </example>
 *
 * <example name="thf-page-blocked-user-expired-password" title="Totvs Page Blocked User Expired Password">
 *  <file name="sample-thf-page-blocked-user-expired-password/sample-thf-page-blocked-user-expired-password.component.html"> </file>
 *  <file name="sample-thf-page-blocked-user-expired-password/sample-thf-page-blocked-user-expired-password.component.ts"> </file>
 * </example>
 */
export class ThfPageBlockedUserComponent extends ThfPageBlockedUserBaseComponent {
    /**
     * @param {?} activatedRoute
     * @param {?} router
     */
    constructor(activatedRoute, router) {
        super();
        this.activatedRoute = activatedRoute;
        this.router = router;
        this.literals = Object.assign({}, thfPageBlockedUserButtonLiterals[thfLocaleDefault], thfPageBlockedUserButtonLiterals[browserLanguage()]);
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this.checkingForRouteMetadata(this.activatedRoute.snapshot.data);
    }
    /**
     * @param {?} url
     * @return {?}
     */
    navigateTo(url) {
        isExternalLink(url) ? window.open(url) : this.router.navigate([url || '/']);
    }
    /**
     * @private
     * @param {?} object
     * @param {?} property
     * @return {?}
     */
    checkingForMetadataProperty(object, property) {
        if (Object.prototype.hasOwnProperty.call(object, property)) {
            return object[property];
        }
    }
    /**
     * @private
     * @param {?} data
     * @return {?}
     */
    checkingForRouteMetadata(data) {
        if (Object.keys(data).length !== 0) {
            this.contactEmail = this.checkingForMetadataProperty(data, 'contactEmail') || this.contactEmail;
            this.contactPhone = this.checkingForMetadataProperty(data, 'contactPhone') || this.contactPhone;
            this.reason = this.checkingForMetadataProperty(data, 'reason') || this.reason;
            this.urlBack = this.checkingForMetadataProperty(data, 'urlBack') || this.urlBack;
        }
    }
}
ThfPageBlockedUserComponent.decorators = [
    { type: Component, args: [{
                selector: 'thf-page-blocked-user',
                template: "<div class=\"thf-page-blocked-user-wrapper\">\n  <div class=\"thf-page-background-wrap\">\n    <thf-page-background\n      [t-hide-logo]=\"true\"\n      [t-logo]=\"logo\"\n      [t-secondary-logo]=\"secondaryLogo\">\n\n      <thf-page-blocked-user-reason\n        [t-params]=\"params\"\n        [t-reason]=\"reason\">\n      </thf-page-blocked-user-reason>\n\n      <div class=\"thf-page-blocked-user-mobile-bottom-alignment\">\n        <thf-page-blocked-user-contacts\n          [t-email]=\"contactEmail\"\n          [t-phone]=\"contactPhone\">\n        </thf-page-blocked-user-contacts>\n\n        <div class=\"thf-row\">\n          <thf-button\n            class=\"thf-lg-12 thf-sm-12 thf-page-blocked-user-button\"\n            t-type=\"primary\"\n            [t-label]=\"literals.primaryButton\"\n            (t-click)=\"navigateTo(urlBack)\">\n          </thf-button>\n        </div>\n      </div>\n    </thf-page-background>\n  </div>\n</div>\n"
            }] }
];
/** @nocollapse */
ThfPageBlockedUserComponent.ctorParameters = () => [
    { type: ActivatedRoute },
    { type: Router }
];
if (false) {
    /** @type {?} */
    ThfPageBlockedUserComponent.prototype.literals;
    /**
     * @type {?}
     * @private
     */
    ThfPageBlockedUserComponent.prototype.activatedRoute;
    /**
     * @type {?}
     * @private
     */
    ThfPageBlockedUserComponent.prototype.router;
}
