// @totvs/thf-ui/esm2015/lib/components/thf-dropdown/thf-dropdown-base.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Input } from '@angular/core';
import { convertToBoolean } from './../../utils/util';
/**
 * \@description
 *
 * O componente `thf-dropdown` pode ser utilizado como um agrupador de ações e / ou opções.
 *
 * > Caso não haja configuração de rotas em sua aplicação, se faz necessário importar o `RouterModule`
 * no módulo principal para o correto funcionamento deste componente:
 *
 * ```
 * import { RouterModule } from '\@angular/router';
 *
 * \@NgModule({
 *   imports: [
 *     ...
 *     RouterModule.forRoot([]),
 *     ThfModule
 *   ],
 *   declarations: [
 *     AppComponent
 *   ],
 *   exports: [],
 *   providers: [],
 *   bootstrap: [
 *     AppComponent
 *   ]
 * })
 * export class AppModule { }
 * ```
 * > Para maiores dúvidas referente à configuração de rotas, acesse em nosso portal /Guias /Começando
 * [/Configurando as rotas do thf-menu](/guides/getting-started).
 */
export class ThfDropdownBaseComponent {
    constructor() {
        this._disabled = false;
        this.icon = 'thf-icon-arrow-down';
        this.open = false;
    }
    /**
     * Lista de ações que serão exibidas no componente.
     * @param {?} value
     * @return {?}
     */
    set actions(value) {
        this._actions = Array.isArray(value) ? value : [];
    }
    /**
     * @return {?}
     */
    get actions() {
        return this._actions;
    }
    /**
     * \@optional
     *
     * \@description
     *
     * Desabilita o campo.
     *
     * \@default `false`
     * @param {?} value
     * @return {?}
     */
    set disabled(value) {
        this._disabled = convertToBoolean(value);
    }
    /**
     * @return {?}
     */
    get disabled() {
        return this._disabled;
    }
}
ThfDropdownBaseComponent.propDecorators = {
    actions: [{ type: Input, args: ['t-actions',] }],
    disabled: [{ type: Input, args: ['t-disabled',] }],
    label: [{ type: Input, args: ['t-label',] }]
};
if (false) {
    /**
     * @type {?}
     * @private
     */
    ThfDropdownBaseComponent.prototype._actions;
    /**
     * @type {?}
     * @private
     */
    ThfDropdownBaseComponent.prototype._disabled;
    /** @type {?} */
    ThfDropdownBaseComponent.prototype.icon;
    /** @type {?} */
    ThfDropdownBaseComponent.prototype.open;
    /**
     * Adiciona um rótulo ao `dropdown`.
     * @type {?}
     */
    ThfDropdownBaseComponent.prototype.label;
}
