// @totvs/thf-ui/esm2015/lib/components/thf-dropdown/thf-dropdown.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ElementRef, Renderer2, ViewChild } from '@angular/core';
import { isKeyCodeEnter } from './../../utils/util';
import { ThfDropdownBaseComponent } from './thf-dropdown-base.component';
/**
 * \@docsExtends ThfDropdownBaseComponent
 *
 * \@example
 *
 * <example name="thf-dropdown-basic" title="Totvs Dropdown Basic" >
 *  <file name="sample-thf-dropdown-basic/sample-thf-dropdown-basic.component.html"> </file>
 *  <file name="sample-thf-dropdown-basic/sample-thf-dropdown-basic.component.ts"> </file>
 *  <file name="sample-thf-dropdown-basic/sample-thf-dropdown-basic.component.e2e-spec.ts"> </file>
 *  <file name="sample-thf-dropdown-basic/sample-thf-dropdown-basic.component.po.ts"> </file>
 * </example>
 *
 * <example name="thf-dropdown-labs" title="Totvs Dropdown Labs" >
 *  <file name="sample-thf-dropdown-labs/sample-thf-dropdown-labs.component.html"> </file>
 *  <file name="sample-thf-dropdown-labs/sample-thf-dropdown-labs.component.ts"> </file>
 * </example>
 *
 * <example name="thf-dropdown-social-network" title="Totvs Dropdown - Social Network" >
 *  <file name="sample-thf-dropdown-social-network/sample-thf-dropdown-social-network.component.html"> </file>
 *  <file name="sample-thf-dropdown-social-network/sample-thf-dropdown-social-network.component.ts"> </file>
 * </example>
 */
export class ThfDropdownComponent extends ThfDropdownBaseComponent {
    /**
     * @param {?} renderer
     */
    constructor(renderer) {
        super();
        this.renderer = renderer;
        this.onScroll = (/**
         * @return {?}
         */
        () => {
            if (this.open) {
                this.hideDropdown();
            }
        });
    }
    /**
     * @param {?} event
     * @return {?}
     */
    onKeyDown(event) {
        if (isKeyCodeEnter(event)) {
            this.toggleDropdown();
        }
    }
    /**
     * @return {?}
     */
    toggleDropdown() {
        this.dropdownRef && !this.open && !this.disabled ? this.showDropdown() : this.hideDropdown();
    }
    /**
     * @private
     * @param {?} event
     * @return {?}
     */
    checkClickArea(event) {
        return this.dropdownRef && this.dropdownRef.nativeElement.contains(event.target);
    }
    /**
     * @private
     * @return {?}
     */
    hideDropdown() {
        this.icon = 'thf-icon-arrow-down';
        this.removeListeners();
        this.popupRef.close();
        this.open = false;
    }
    /**
     * @private
     * @return {?}
     */
    initializeListeners() {
        this.clickoutListener = this.renderer.listen('document', 'click', (/**
         * @param {?} event
         * @return {?}
         */
        (event) => {
            this.wasClickedOnDropdown(event);
        }));
        this.resizeListener = this.renderer.listen('window', 'resize', (/**
         * @return {?}
         */
        () => {
            this.hideDropdown();
        }));
        window.addEventListener('scroll', this.onScroll, true);
    }
    /**
     * @private
     * @return {?}
     */
    removeListeners() {
        if (this.clickoutListener) {
            this.clickoutListener();
        }
        if (this.resizeListener) {
            this.resizeListener();
        }
        window.removeEventListener('scroll', this.onScroll, true);
    }
    /**
     * @private
     * @return {?}
     */
    showDropdown() {
        this.icon = 'thf-icon-arrow-up';
        this.initializeListeners();
        this.popupRef.open();
        this.open = true;
    }
    /**
     * @private
     * @param {?} event
     * @return {?}
     */
    wasClickedOnDropdown(event) {
        /** @type {?} */
        const clickedOnDropdown = this.checkClickArea(event);
        if (!clickedOnDropdown) {
            this.hideDropdown();
        }
    }
}
ThfDropdownComponent.decorators = [
    { type: Component, args: [{
                selector: 'thf-dropdown',
                template: "<div #dropdownRef\n  class=\"thf-dropdown\"\n  tabindex=\"{{ disabled ? -1 : 0 }}\"\n  (click)=\"toggleDropdown()\"\n  (keydown)=\"onKeyDown($event)\">\n\n  <div class=\"thf-dropdown-button\" [ngClass]=\"{'thf-dropdown-button-disabled' : disabled}\">\n    {{ label }}\n    <div class=\"thf-dropdown-icon thf-icon {{icon}}\"></div>\n  </div>\n\n  <thf-popup #popupRef\n    t-hide-arrow\n    t-is-corner-align\n    t-position=\"bottom-left\"\n    [t-actions]=\"actions\"\n    [t-custom-positions]=\"['bottom-left', 'top-left']\"\n    [t-target]=\"dropdownRef\">\n  </thf-popup>\n\n</div>\n"
            }] }
];
/** @nocollapse */
ThfDropdownComponent.ctorParameters = () => [
    { type: Renderer2 }
];
ThfDropdownComponent.propDecorators = {
    dropdownRef: [{ type: ViewChild, args: ['dropdownRef', { read: ElementRef, static: true },] }],
    popupRef: [{ type: ViewChild, args: ['popupRef', { static: false },] }]
};
if (false) {
    /**
     * @type {?}
     * @private
     */
    ThfDropdownComponent.prototype.clickoutListener;
    /**
     * @type {?}
     * @private
     */
    ThfDropdownComponent.prototype.resizeListener;
    /** @type {?} */
    ThfDropdownComponent.prototype.dropdownRef;
    /** @type {?} */
    ThfDropdownComponent.prototype.popupRef;
    /**
     * @type {?}
     * @private
     */
    ThfDropdownComponent.prototype.onScroll;
    /**
     * @type {?}
     * @private
     */
    ThfDropdownComponent.prototype.renderer;
}
