// @totvs/thf-ui/esm5/lib/components/thf-list-view/thf-list-view.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { RouterModule } from '@angular/router';
import { ThfButtonModule } from '../thf-button/thf-button.module';
import { ThfPopupModule } from '../thf-popup/thf-popup.module';
import { ThfListViewComponent } from './thf-list-view.component';
import { ThfListViewContentTemplateDirective } from './thf-list-view-content-template/thf-list-view-content-template.directive';
import { ThfListViewDetailTemplateDirective } from './thf-list-view-detail-template/thf-list-view-detail-template.directive';
/**
 * \@description
 *
 * Módulo do componente `thf-list-view`.
 *
 * > Para o correto funcionamento do componente `thf-list-view`, deve ser importado o módulo `BrowserAnimationsModule` no
 * > módulo principal da sua aplicação.
 *
 * Módulo da aplicação:
 * ```
 * import { BrowserAnimationsModule } from '\@angular/platform-browser/animations';
 * import { ThfModule } from '\@totvs/thf-ui';
 * ...
 *
 * \@NgModule({
 *   imports: [
 *     BrowserModule,
 *     BrowserAnimationsModule,
 *     ...
 *     ThfModule
 *   ],
 *   declarations: [
 *     AppComponent,
 *     ...
 *   ],
 *   providers: [],
 *   bootstrap: [AppComponent]
 * })
 * export class AppModule { }
 * ```
 */
var ThfListViewModule = /** @class */ (function () {
    function ThfListViewModule() {
    }
    ThfListViewModule.decorators = [
        { type: NgModule, args: [{
                    imports: [
                        CommonModule,
                        RouterModule,
                        ThfButtonModule,
                        ThfPopupModule
                    ],
                    declarations: [
                        ThfListViewComponent,
                        ThfListViewContentTemplateDirective,
                        ThfListViewDetailTemplateDirective
                    ],
                    exports: [
                        ThfListViewComponent,
                        ThfListViewContentTemplateDirective,
                        ThfListViewDetailTemplateDirective
                    ],
                    providers: [],
                    schemas: []
                },] }
    ];
    return ThfListViewModule;
}());
export { ThfListViewModule };
