// @totvs/thf-templates/esm5/lib/components/thf-page-dynamic-search/thf-advanced-filter/thf-advanced-filter.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, ViewChild } from '@angular/core';
import { ThfDynamicFormComponent, ThfModalComponent } from '@totvs/thf-ui';
import { ThfAdvancedFilterBaseComponent } from './thf-advanced-filter-base.component';
/**
 * \@docsPrivate
 *
 * \@docsExtends ThfAdvancedFilterBaseComponent
 *
 * \@examplePrivate
 *
 * <example-private name="thf-advanced-filter" title="Totvs Busca Avançada">
 *   <file name="sample-thf-advanced-filter.component.html"> </file>
 *   <file name="sample-thf-advanced-filter.component.ts"> </file>
 * </example-private>
 */
var ThfAdvancedFilterComponent = /** @class */ (function (_super) {
    tslib_1.__extends(ThfAdvancedFilterComponent, _super);
    function ThfAdvancedFilterComponent() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.filter = {};
        _this.primaryAction = {
            action: (/**
             * @return {?}
             */
            function () {
                /** @type {?} */
                var models = _this.getValuesFromForm();
                _this.searchEvent.emit(models);
                _this.thfModal.close();
            }),
            label: _this.literals.primaryActionLabel
        };
        _this.secondaryAction = {
            action: (/**
             * @return {?}
             */
            function () {
                _this.thfModal.close();
            }),
            label: _this.literals.secondaryActionLabel
        };
        return _this;
    }
    // Retorna os models dos campos preenchidos
    // Retorna os models dos campos preenchidos
    /**
     * @return {?}
     */
    ThfAdvancedFilterComponent.prototype.getValuesFromForm = 
    // Retorna os models dos campos preenchidos
    /**
     * @return {?}
     */
    function () {
        var _this = this;
        Object.keys(this.filter).forEach((/**
         * @param {?} property
         * @return {?}
         */
        function (property) {
            if (_this.filter[property] === undefined || _this.filter[property] === '') {
                delete _this.filter[property];
            }
        }));
        return this.filter;
    };
    /**
     * @return {?}
     */
    ThfAdvancedFilterComponent.prototype.open = /**
     * @return {?}
     */
    function () {
        this.filter = {};
        this.thfModal.open();
    };
    ThfAdvancedFilterComponent.decorators = [
        { type: Component, args: [{
                    selector: 'thf-advanced-filter',
                    template: "<thf-modal\n  t-hide-close\n  [t-primary-action]=\"primaryAction\"\n  [t-secondary-action]=\"secondaryAction\"\n  [t-title]=\"literals.title\">\n\n  <thf-dynamic-form\n    [t-fields]=\"filters\"\n    [t-value]=\"filter\">\n  </thf-dynamic-form>\n\n</thf-modal>\n"
                }] }
    ];
    ThfAdvancedFilterComponent.propDecorators = {
        thfModal: [{ type: ViewChild, args: [ThfModalComponent, { static: true },] }],
        thfDynamicForm: [{ type: ViewChild, args: [ThfDynamicFormComponent, { static: true },] }]
    };
    return ThfAdvancedFilterComponent;
}(ThfAdvancedFilterBaseComponent));
export { ThfAdvancedFilterComponent };
if (false) {
    /** @type {?} */
    ThfAdvancedFilterComponent.prototype.filter;
    /** @type {?} */
    ThfAdvancedFilterComponent.prototype.thfModal;
    /** @type {?} */
    ThfAdvancedFilterComponent.prototype.thfDynamicForm;
    /** @type {?} */
    ThfAdvancedFilterComponent.prototype.primaryAction;
    /** @type {?} */
    ThfAdvancedFilterComponent.prototype.secondaryAction;
}
