// @totvs/thf-ui/esm2015/lib/components/thf-stepper/thf-stepper-step/thf-stepper-step.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, EventEmitter, Input, Output } from '@angular/core';
import { browserLanguage, convertToBoolean, isTypeof, thfLocaleDefault } from './../../../utils/util';
import { ThfStepperOrientation } from '../enums/thf-stepper-orientation.enum';
import { ThfStepperStatus } from '../enums/thf-stepper-status.enum';
/** @type {?} */
const thfStepperStepSizeDefault = 24;
/** @type {?} */
const thfStepperStepSizeMax = 64;
/** @type {?} */
const thfStepLiteralsDefault = {
    en: { label: 'Step' },
    es: { label: 'Paso' },
    pt: { label: 'Passo' }
};
/**
 * \@docsPrivate
 *
 * \@description
 *
 * Componente que representa um *step* contendo as ligações das laterais (`thf-stepper-step-bar`) e a label.
 */
export class ThfStepperStepComponent {
    constructor() {
        this._stepIcons = false;
        this._stepSize = thfStepperStepSizeDefault;
        this.literals = Object.assign({}, thfStepLiteralsDefault[thfLocaleDefault], thfStepLiteralsDefault[browserLanguage()]);
        // Evento que será emitido quando o status do *step* estiver ativo (`ThfStepperStatus.Active`).
        this.activated = new EventEmitter();
        // Evento que será emitido ao clicar no *step*.
        this.click = new EventEmitter();
        // Evento que será emitido ao focar no *step* e pressionar a tecla *enter*.
        this.enter = new EventEmitter();
    }
    // Label do *step*.
    /**
     * @param {?} value
     * @return {?}
     */
    set label(value) {
        this._label = isTypeof(value, 'string') ? value : `${this.literals.label} ${this.circleContent}`;
    }
    /**
     * @return {?}
     */
    get label() {
        return this._label;
    }
    // Define o estado de exibição.
    /**
     * @param {?} value
     * @return {?}
     */
    set status(value) {
        this._status = ((/** @type {?} */ (value))) ? value : undefined;
        if (this.status === ThfStepperStatus.Active) {
            this.activated.emit();
        }
    }
    /**
     * @return {?}
     */
    get status() {
        return this._status;
    }
    /**
     * @param {?} stepIcons
     * @return {?}
     */
    set stepIcons(stepIcons) {
        this._stepIcons = convertToBoolean(stepIcons);
    }
    /**
     * @return {?}
     */
    get stepIcons() {
        return this._stepIcons;
    }
    // Tamanho do `step` em *pixels*, possibilitando um maior destaque.
    // O valor informado deve ser entre `24` e `64`.
    /**
     * @param {?} value
     * @return {?}
     */
    set stepSize(value) {
        this._stepSize = (value >= thfStepperStepSizeDefault && value <= thfStepperStepSizeMax) ? value : thfStepperStepSizeDefault;
    }
    /**
     * @return {?}
     */
    get stepSize() {
        return this._stepSize;
    }
    /**
     * @return {?}
     */
    get halfStepSize() {
        return this.stepSize / 2;
    }
    /**
     * @return {?}
     */
    get isVerticalOrientation() {
        return this.orientation === ThfStepperOrientation.Vertical;
    }
    /**
     * @return {?}
     */
    get marginHorizontalBar() {
        return this.isVerticalOrientation ? undefined : this.halfStepSize;
    }
    /**
     * @param {?} status
     * @return {?}
     */
    getStatusClass(status) {
        switch (status) {
            case ThfStepperStatus.Active:
                return 'thf-stepper-step-active';
            case ThfStepperStatus.Disabled:
                return 'thf-stepper-step-disabled';
            case ThfStepperStatus.Done:
                return 'thf-stepper-step-done';
            case ThfStepperStatus.Error:
                return 'thf-stepper-step-error';
            default:
                return 'thf-stepper-step-default';
        }
    }
    /**
     * @return {?}
     */
    onClick() {
        if (this.status !== ThfStepperStatus.Disabled) {
            this.click.emit();
        }
    }
    /**
     * @return {?}
     */
    onEnter() {
        if (this.status !== ThfStepperStatus.Disabled) {
            this.enter.emit();
        }
    }
}
ThfStepperStepComponent.decorators = [
    { type: Component, args: [{
                selector: 'thf-stepper-step',
                template: "<div class=\"thf-stepper-step\"\n  [ngClass]=\"getStatusClass(status)\"\n  (click)=\"onClick()\"\n  (keydown.enter)=\"onEnter()\">\n\n  <div class=\"thf-stepper-step-container\" [style.width.px]=\"isVerticalOrientation ? stepSize : undefined\">\n\n    <div\n      [class.thf-stepper-step-bar-top]=\"isVerticalOrientation\"\n      [class.thf-stepper-step-bar-left]=\"!isVerticalOrientation\"\n      [style.margin-right.px]=\"marginHorizontalBar\">\n    </div>\n\n    <thf-stepper-circle\n      [t-content]=\"circleContent\"\n      [t-icons]=\"stepIcons\"\n      [t-size]=\"stepSize\"\n      [t-status]=\"status\">\n    </thf-stepper-circle>\n\n    <div\n      [class.thf-stepper-step-bar-bottom]=\"isVerticalOrientation\"\n      [class.thf-stepper-step-bar-right]=\"!isVerticalOrientation\"\n      [style.margin-left.px]=\"marginHorizontalBar\">\n    </div>\n\n  </div>\n\n  <thf-stepper-label\n    class=\"thf-stepper-step-label-position\"\n    [t-content]=\"label\">\n  </thf-stepper-label>\n</div>\n"
            }] }
];
ThfStepperStepComponent.propDecorators = {
    circleContent: [{ type: Input, args: ['t-circle-content',] }],
    label: [{ type: Input, args: ['t-label',] }],
    orientation: [{ type: Input, args: ['t-orientation',] }],
    status: [{ type: Input, args: ['t-status',] }],
    stepIcons: [{ type: Input, args: ['t-step-icons',] }],
    stepSize: [{ type: Input, args: ['t-step-size',] }],
    activated: [{ type: Output, args: ['t-activated',] }],
    click: [{ type: Output, args: ['t-click',] }],
    enter: [{ type: Output, args: ['t-enter',] }]
};
if (false) {
    /**
     * @type {?}
     * @private
     */
    ThfStepperStepComponent.prototype._label;
    /**
     * @type {?}
     * @private
     */
    ThfStepperStepComponent.prototype._status;
    /**
     * @type {?}
     * @private
     */
    ThfStepperStepComponent.prototype._stepIcons;
    /**
     * @type {?}
     * @private
     */
    ThfStepperStepComponent.prototype._stepSize;
    /** @type {?} */
    ThfStepperStepComponent.prototype.literals;
    /** @type {?} */
    ThfStepperStepComponent.prototype.circleContent;
    /** @type {?} */
    ThfStepperStepComponent.prototype.orientation;
    /** @type {?} */
    ThfStepperStepComponent.prototype.activated;
    /** @type {?} */
    ThfStepperStepComponent.prototype.click;
    /** @type {?} */
    ThfStepperStepComponent.prototype.enter;
}
