// @totvs/thf-ui/esm2015/lib/components/thf-page/thf-page-detail/thf-page-detail.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ViewContainerRef } from '@angular/core';
import { ThfPageDetailBaseComponent } from './thf-page-detail-base.component';
import { callAction, hasAction } from '../thf-page-util/thf-page-util';
/**
 * \@docsExtends ThfPageDetailBaseComponent
 *
 * \@example
 *
 * <example name="thf-page-detail-basic" title="Totvs Page Detail Basic">
 *  <file name="sample-thf-page-detail-basic/sample-thf-page-detail-basic.component.html"> </file>
 *  <file name="sample-thf-page-detail-basic/sample-thf-page-detail-basic.component.ts"> </file>
 * </example>
 *
 * <example name="thf-page-detail-labs" title="Totvs Page Detail Labs">
 *  <file name="sample-thf-page-detail-labs/sample-thf-page-detail-labs.component.html"> </file>
 *  <file name="sample-thf-page-detail-labs/sample-thf-page-detail-labs.component.ts"> </file>
 * </example>
 *
 * <example name="thf-page-detail-user" title="Totvs Page Detail - User">
 *  <file name="sample-thf-page-detail-user/sample-thf-page-detail-user.component.html"> </file>
 *  <file name="sample-thf-page-detail-user/sample-thf-page-detail-user.component.ts"> </file>
 * </example>
 */
export class ThfPageDetailComponent extends ThfPageDetailBaseComponent {
    /**
     * @param {?} viewRef
     */
    constructor(viewRef) {
        super();
        this.callActionFn = callAction;
        this.hasActionFn = hasAction;
        this.parentContext = viewRef['_view']['component'];
    }
    /**
     * @return {?}
     */
    hasAnyAction() {
        return this.hasActionFn('back', this.parentContext) ||
            this.hasActionFn('edit', this.parentContext) ||
            this.hasActionFn('remove', this.parentContext);
    }
    /**
     * @param {?} property
     * @return {?}
     */
    hasEditFn(property) {
        if (property === 'icon') {
            return this.hasActionFn('edit', this.parentContext) ? '' : 'thf-icon-delete';
        }
        else if (property === 'type') {
            return this.hasActionFn('edit', this.parentContext) ? 'default' : 'primary';
        }
        else {
            return '';
        }
    }
    /**
     * @param {?} property
     * @return {?}
     */
    hasEditOrRemoveFn(property) {
        if (property === 'icon') {
            return this.hasActionFn('edit', this.parentContext) || this.hasActionFn('remove', this.parentContext) ? '' : 'thf-icon-arrow-left';
        }
        else if (property === 'type') {
            return this.hasActionFn('edit', this.parentContext) || this.hasActionFn('remove', this.parentContext) ? 'default' : 'primary';
        }
        else {
            return '';
        }
    }
    /**
     * @return {?}
     */
    hasPageHeader() {
        return !!(this.title || this.hasAnyAction() || (this.breadcrumb && this.breadcrumb.items.length));
    }
}
ThfPageDetailComponent.decorators = [
    { type: Component, args: [{
                selector: 'thf-page-detail',
                template: "<thf-page>\n  <!-- HEADER -->\n  <thf-page-header\n    *ngIf=\"hasPageHeader()\"\n    [t-breadcrumb]=\"breadcrumb\"\n    [t-title]=\"title\">\n\n    <!-- OPERATIONS -->\n    <div *ngIf=\"hasAnyAction()\" class=\"thf-page-header-actions\">\n      <thf-button *ngIf=\"hasActionFn('back', parentContext)\"\n        [t-icon]=\"hasEditOrRemoveFn('icon')\"\n        [t-label]=\"literals.back\"\n        [t-type]=\"hasEditOrRemoveFn('type')\"\n        (t-click)=\"callActionFn('back', parentContext)\">\n      </thf-button>\n\n      <thf-button *ngIf=\"hasActionFn('remove', parentContext)\"\n        [t-icon]=\"hasEditFn('icon')\"\n        [t-label]=\"literals.remove\"\n        [t-type]=\"hasEditFn('type')\"\n        (t-click)=\"callActionFn('remove', parentContext)\">\n      </thf-button>\n\n      <thf-button *ngIf=\"hasActionFn('edit', parentContext)\"\n        t-icon=\"thf-icon-edit\"\n        t-type=\"primary\"\n        [t-label]=\"literals.edit\"\n        (t-click)=\"callActionFn('edit', parentContext)\">\n      </thf-button>\n    </div>\n\n  </thf-page-header>\n\n  <!-- CONTENT -->\n  <thf-page-content>\n    <ng-content>\n    </ng-content>\n  </thf-page-content>\n\n</thf-page>\n"
            }] }
];
/** @nocollapse */
ThfPageDetailComponent.ctorParameters = () => [
    { type: ViewContainerRef }
];
if (false) {
    /** @type {?} */
    ThfPageDetailComponent.prototype.callActionFn;
    /** @type {?} */
    ThfPageDetailComponent.prototype.hasActionFn;
    /** @type {?} */
    ThfPageDetailComponent.prototype.parentContext;
}
