// @totvs/thf-ui/esm2015/lib/components/thf-toolbar/thf-toolbar-notification/thf-toolbar-notification.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input } from '@angular/core';
import { ThfControlPositionService } from '../../../services/thf-control-position/thf-control-position.service';
/**
 * \@docsPrivate
 *
 * \@usedBy ThfToolbarComponent
 *
 * \@description
 *
 * O componente `thf-toolbar-notification` tem como objetivo notificar o usuário de novas ações da aplicação que necessitam de atenção,
 * como por exemplo um alerta de nova mensagem.
 *
 * O mesmo também permite que a cada nova notificação seja incrementado e exibido, ou não, este número em uma *tag* ao lado do
 * ícone de notificações.
 */
export class ThfToolbarNotificationComponent {
    constructor() {
        this._notificationNumber = 0;
    }
    /**
     * Define o número de notificações.
     * @param {?} value
     * @return {?}
     */
    set notificationNumber(value) {
        this._notificationNumber = Number.isInteger(value) ? value : 0;
    }
    /**
     * @return {?}
     */
    get notificationNumber() {
        return this._notificationNumber;
    }
}
ThfToolbarNotificationComponent.decorators = [
    { type: Component, args: [{
                selector: 'thf-toolbar-notification',
                template: "<div class=\"thf-toolbar-notification thf-clickable\" (click)=\"popup.toggle()\">\n  <span #notification class=\"thf-icon thf-icon-notification thf-toolbar-icon\"></span>\n  <div *ngIf=\"notificationNumber\" class=\"thf-toolbar-notification-badge\">\n    {{ notificationNumber }}\n  </div>\n</div>\n\n<thf-popup #popup\n  [t-actions]=\"notificationActions\"\n  [t-target]=\"notification\">\n</thf-popup>\n",
                providers: [ThfControlPositionService]
            }] }
];
ThfToolbarNotificationComponent.propDecorators = {
    notificationActions: [{ type: Input, args: ['t-notification-actions',] }],
    notificationNumber: [{ type: Input, args: ['t-notification-number',] }]
};
if (false) {
    /**
     * @type {?}
     * @private
     */
    ThfToolbarNotificationComponent.prototype._notificationNumber;
    /**
     * Define uma lista de ações.
     * @type {?}
     */
    ThfToolbarNotificationComponent.prototype.notificationActions;
}
