// @totvs/thf-ui/esm5/lib/components/thf-toolbar/thf-toolbar-profile/thf-toolbar-profile.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input } from '@angular/core';
import { ThfControlPositionService } from '../../../services/thf-control-position/thf-control-position.service';
/** @type {?} */
var thfToolbarProfileDefaultAvatar = './assets/images/totvs-logo-user.svg';
/**
 * \@docsPrivate
 *
 * \@docsExtends ThfToolbarProfileComponent
 *
 * \@description
 *
 * Componente `thf-toolbar-profile`.
 */
var ThfToolbarProfileComponent = /** @class */ (function () {
    function ThfToolbarProfileComponent() {
        this.defaultAvatar = thfToolbarProfileDefaultAvatar;
    }
    ThfToolbarProfileComponent.decorators = [
        { type: Component, args: [{
                    selector: 'thf-toolbar-profile',
                    template: "<div #profileElement class=\"thf-toolbar-profile thf-clickable\" (click)=\"popup.toggle()\">\n  <ng-container *ngIf=\"profile.avatar; then avatar; else userIcon\"></ng-container>\n</div>\n\n<thf-popup #popup\n  [t-actions]=\"profileActions\"\n  [t-target]=\"profileElement\">\n\n  <div t-popup-header-template\n    *ngIf=\"profile?.title\"\n    class=\"thf-toolbar-profile-item-header\">\n\n    <thf-avatar\n      class=\"thf-toolbar-profile-item-avatar\"\n      t-size=\"sm\"\n      [t-src]=\"profile.avatar || defaultAvatar\">\n    </thf-avatar>\n\n    <div>\n      <div class=\"thf-toolbar-profile-item-header-title\">\n        {{ profile.title }}\n      </div>\n\n      <div *ngIf=\"profile.subtitle\" class=\"thf-toolbar-profile-item-header-subtitle\">\n        {{ profile.subtitle }}\n      </div>\n    </div>\n  </div>\n\n</thf-popup>\n\n<ng-template #avatar>\n  <thf-avatar t-size=\"xs\" [t-src]=\"profile.avatar\"></thf-avatar>\n</ng-template>\n\n<ng-template #userIcon>\n  <span class=\"thf-icon thf-icon-user thf-toolbar-icon\"></span>\n</ng-template>\n",
                    providers: [ThfControlPositionService]
                }] }
    ];
    ThfToolbarProfileComponent.propDecorators = {
        profile: [{ type: Input, args: ['t-profile',] }],
        profileActions: [{ type: Input, args: ['t-profile-actions',] }]
    };
    return ThfToolbarProfileComponent;
}());
export { ThfToolbarProfileComponent };
if (false) {
    /**
     * Objeto que implementa a interface `ThfToolbarProfile`.
     * @type {?}
     */
    ThfToolbarProfileComponent.prototype.profile;
    /**
     * Define uma lista de ações.
     * @type {?}
     */
    ThfToolbarProfileComponent.prototype.profileActions;
    /** @type {?} */
    ThfToolbarProfileComponent.prototype.defaultAvatar;
}
