// @totvs/thf-ui/esm2015/lib/services/thf-notification/thf-notification.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { ThfComponentInjectorService } from './../thf-component-injector/thf-component-injector.service';
import { ThfNotificationBaseService } from './thf-notification-base.service';
import { ThfToasterOrientation } from './thf-toaster/thf-toaster-orientation.enum';
import { ThfToasterComponent } from './thf-toaster/thf-toaster.component';
import * as i0 from "@angular/core";
import * as i1 from "../thf-component-injector/thf-component-injector.service";
/**
 * \@docsExtends ThfNotificationBaseService
 *
 * \@example
 *
 * <example name="thf-notification-basic" title="Totvs Notification Basic">
 *  <file name="sample-thf-notification-basic/sample-thf-notification-basic.component.html"> </file>
 *  <file name="sample-thf-notification-basic/sample-thf-notification-basic.component.ts"> </file>
 * </example>
 *
 * <example name="thf-notification-labs" title="Totvs Notification Labs">
 *  <file name="sample-thf-notification-labs/sample-thf-notification-labs.component.html"> </file>
 *  <file name="sample-thf-notification-labs/sample-thf-notification-labs.component.ts"> </file>
 * </example>
 *
 * <example name="thf-notification-sales" title="Totvs Notification - Sales">
 *  <file name="sample-thf-notification-sales/sample-thf-notification-sales.component.html"> </file>
 *  <file name="sample-thf-notification-sales/sample-thf-notification-sales.component.ts"> </file>
 * </example>
 */
export class ThfNotificationService extends ThfNotificationBaseService {
    /**
     * @param {?} thfComponentInjector
     */
    constructor(thfComponentInjector) {
        super();
        this.thfComponentInjector = thfComponentInjector;
    }
    /**
     * @param {?} toaster
     * @return {?}
     */
    createToaster(toaster) {
        /** @type {?} */
        const componentRef = this.thfComponentInjector.createComponentInApplication(ThfToasterComponent);
        toaster.componentRef = componentRef;
        componentRef.changeDetectorRef.detectChanges();
        componentRef.instance.configToaster(toaster);
        if (toaster.orientation === ThfToasterOrientation.Top) {
            this.stackTop.push(componentRef);
        }
        else {
            this.stackBottom.push(componentRef);
        }
        this.observableOnClose(componentRef);
        if (toaster.action === undefined) {
            setTimeout((/**
             * @return {?}
             */
            () => {
                this.destroyToaster(componentRef);
            }), this.durationToaster);
        }
    }
    /**
     * @param {?} toaster
     * @return {?}
     */
    destroyToaster(toaster) {
        /** @type {?} */
        let stack;
        if (toaster.instance.orientation === ThfToasterOrientation.Top) {
            stack = this.stackTop;
        }
        else {
            stack = this.stackBottom;
        }
        /** @type {?} */
        const index = stack.indexOf(toaster);
        stack.splice(index, 1);
        this.thfComponentInjector.destroyComponentInApplication(toaster);
        for (let count = 0; count < stack.length; count++) {
            stack[count].instance.changePosition(count);
        }
    }
    /**
     * @private
     * @param {?} componentRef
     * @return {?}
     */
    observableOnClose(componentRef) {
        componentRef.instance.observableOnClose.subscribe((/**
         * @return {?}
         */
        () => {
            this.destroyToaster(componentRef);
        }));
    }
}
ThfNotificationService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] }
];
/** @nocollapse */
ThfNotificationService.ctorParameters = () => [
    { type: ThfComponentInjectorService }
];
/** @nocollapse */ ThfNotificationService.ngInjectableDef = i0.ɵɵdefineInjectable({ factory: function ThfNotificationService_Factory() { return new ThfNotificationService(i0.ɵɵinject(i1.ThfComponentInjectorService)); }, token: ThfNotificationService, providedIn: "root" });
if (false) {
    /**
     * @type {?}
     * @private
     */
    ThfNotificationService.prototype.thfComponentInjector;
}
