// @totvs/thf-ui/esm5/lib/components/components.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { ThfAvatarModule } from './thf-avatar/thf-avatar.module';
import { ThfBreadcrumbModule } from './thf-breadcrumb/thf-breadcrumb.module';
import { ThfButtonModule } from './thf-button/thf-button.module';
import { ThfButtonGroupModule } from './thf-button-group/thf-button-group.module';
import { ThfCalendarModule } from './thf-calendar/thf-calendar.module';
import { ThfChartModule } from './thf-chart/thf-chart.module';
import { ThfContainerModule } from './thf-container/thf-container.module';
import { ThfDisclaimerGroupModule } from './thf-disclaimer-group/thf-disclaimer-group.module';
import { ThfDisclaimerModule } from './thf-disclaimer/thf-disclaimer.module';
import { ThfDividerModule } from './thf-divider/thf-divider.module';
import { ThfDropdownModule } from './thf-dropdown/thf-dropdown.module';
import { ThfDynamicModule } from './thf-dynamic/thf-dynamic.module';
import { ThfFieldModule } from './thf-field/thf-field.module';
import { ThfGridModule } from './thf-grid/thf-grid.module';
import { ThfInfoModule } from './thf-info/thf-info.module';
import { ThfListViewModule } from './thf-list-view/thf-list-view.module';
import { ThfLoadingModule } from './thf-loading/thf-loading.module';
import { ThfMenuModule } from './thf-menu/thf-menu.module';
import { ThfMenuPanelModule } from './thf-menu-panel/thf-menu-panel.module';
import { ThfModalModule } from './thf-modal/thf-modal.module';
import { ThfNavbarModule } from './thf-navbar/thf-navbar.module';
import { ThfPageModule } from './thf-page/thf-page.module';
import { ThfPopoverModule } from './thf-popover/thf-popover.module';
import { ThfPopupModule } from './thf-popup/thf-popup.module';
import { ThfSlideModule } from './thf-slide/thf-slide.module';
import { ThfStepperModule } from './thf-stepper/thf-stepper.module';
import { ThfTableModule } from './thf-table/thf-table.module';
import { ThfTabsModule } from './thf-tabs/thf-tabs.module';
import { ThfTagModule } from './thf-tag/thf-tag.module';
import { ThfToolbarModule } from './thf-toolbar/thf-toolbar.module';
import { ThfWidgetModule } from './thf-widget/thf-widget.module';
var ThfComponentsModule = /** @class */ (function () {
    function ThfComponentsModule() {
    }
    ThfComponentsModule.decorators = [
        { type: NgModule, args: [{
                    declarations: [],
                    imports: [
                        ThfAvatarModule,
                        ThfBreadcrumbModule,
                        ThfButtonModule,
                        ThfButtonGroupModule,
                        ThfCalendarModule,
                        ThfChartModule,
                        ThfContainerModule,
                        ThfDisclaimerGroupModule,
                        ThfDisclaimerModule,
                        ThfDividerModule,
                        ThfDropdownModule,
                        ThfDynamicModule,
                        ThfFieldModule,
                        ThfGridModule,
                        ThfInfoModule,
                        ThfListViewModule,
                        ThfLoadingModule,
                        ThfMenuModule,
                        ThfMenuPanelModule,
                        ThfModalModule,
                        ThfNavbarModule,
                        ThfPageModule,
                        ThfPopoverModule,
                        ThfPopupModule,
                        ThfSlideModule,
                        ThfStepperModule,
                        ThfTableModule,
                        ThfTabsModule,
                        ThfTagModule,
                        ThfToolbarModule,
                        ThfWidgetModule
                    ],
                    exports: [
                        ThfAvatarModule,
                        ThfBreadcrumbModule,
                        ThfButtonModule,
                        ThfButtonGroupModule,
                        ThfCalendarModule,
                        ThfChartModule,
                        ThfContainerModule,
                        ThfDisclaimerGroupModule,
                        ThfDisclaimerModule,
                        ThfDividerModule,
                        ThfDropdownModule,
                        ThfDynamicModule,
                        ThfFieldModule,
                        ThfGridModule,
                        ThfInfoModule,
                        ThfListViewModule,
                        ThfLoadingModule,
                        ThfMenuModule,
                        ThfMenuPanelModule,
                        ThfModalModule,
                        ThfNavbarModule,
                        ThfPageModule,
                        ThfPopoverModule,
                        ThfPopupModule,
                        ThfSlideModule,
                        ThfStepperModule,
                        ThfTableModule,
                        ThfTabsModule,
                        ThfTagModule,
                        ThfToolbarModule,
                        ThfWidgetModule
                    ],
                    providers: [],
                    bootstrap: []
                },] }
    ];
    return ThfComponentsModule;
}());
export { ThfComponentsModule };
