// @totvs/thf-ui/esm5/lib/components/thf-page/thf-page-default/thf-page-default.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, Renderer2, ViewContainerRef } from '@angular/core';
import { Router } from '@angular/router';
import { callFunction, isTypeof } from '../../../utils/util';
import { ThfPageDefaultBaseComponent } from './thf-page-default-base.component';
/**
 * \@docsExtends ThfPageDefaultBaseComponent
 *
 * \@example
 *
 * <example name="thf-page-default-basic" title="Totvs Page Default Basic">
 *  <file name="sample-thf-page-default-basic/sample-thf-page-default-basic.component.html"> </file>
 *  <file name="sample-thf-page-default-basic/sample-thf-page-default-basic.component.ts"> </file>
 * </example>
 *
 * <example name="thf-page-default-labs" title="Totvs Page Default Labs">
 *  <file name="sample-thf-page-default-labs/sample-thf-page-default-labs.component.html"> </file>
 *  <file name="sample-thf-page-default-labs/sample-thf-page-default-labs.component.ts"> </file>
 * </example>
 *
 * <example name="thf-page-default-dashboard" title="Totvs Page Default - Dashboard">
 *  <file name="sample-thf-page-default-dashboard/sample-thf-page-default-dashboard.component.html"> </file>
 *  <file name="sample-thf-page-default-dashboard/sample-thf-page-default-dashboard.component.ts"> </file>
 *  <file name="sample-thf-page-default-dashboard/sample-thf-page-default-dashboard.service.ts"> </file>
 * </example>
 */
var ThfPageDefaultComponent = /** @class */ (function (_super) {
    tslib_1.__extends(ThfPageDefaultComponent, _super);
    function ThfPageDefaultComponent(viewRef, renderer, router) {
        var _this = _super.call(this) || this;
        _this.renderer = renderer;
        _this.router = router;
        _this.limitPrimaryActions = 3;
        _this.maxWidthMobile = 480;
        _this.parentRef = viewRef['_view']['component'];
        return _this;
    }
    /**
     * @return {?}
     */
    ThfPageDefaultComponent.prototype.ngAfterContentInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this.setIsMobile();
        this.setDropdownActions();
        this.renderer.listen('window', 'resize', (/**
         * @param {?} event
         * @return {?}
         */
        function (event) {
            _this.onResize(event);
        }));
    };
    /**
     * @param {?} changes
     * @return {?}
     */
    ThfPageDefaultComponent.prototype.ngOnChanges = /**
     * @param {?} changes
     * @return {?}
     */
    function (changes) {
        this.setDropdownActions();
    };
    /**
     * @param {?} action
     * @return {?}
     */
    ThfPageDefaultComponent.prototype.actionIsDisabled = /**
     * @param {?} action
     * @return {?}
     */
    function (action) {
        return isTypeof(action.disabled, 'function') ? action.disabled(action) : action.disabled;
    };
    /**
     * @param {?} item
     * @return {?}
     */
    ThfPageDefaultComponent.prototype.callAction = /**
     * @param {?} item
     * @return {?}
     */
    function (item) {
        if (item.url) {
            this.router.navigate([item.url]);
        }
        else if (item.action) {
            callFunction(item.action, this.parentRef);
        }
    };
    /**
     * @return {?}
     */
    ThfPageDefaultComponent.prototype.hasPageHeader = /**
     * @return {?}
     */
    function () {
        return !!(this.title || (this.actions && this.actions.length) || (this.breadcrumb && this.breadcrumb.items.length));
    };
    /**
     * @private
     * @param {?} event
     * @return {?}
     */
    ThfPageDefaultComponent.prototype.onResize = /**
     * @private
     * @param {?} event
     * @return {?}
     */
    function (event) {
        /** @type {?} */
        var width = ((/** @type {?} */ (event.target))).innerWidth;
        if (width < this.maxWidthMobile) {
            this.isMobile = true;
            this.limitPrimaryActions = 2;
            this.setDropdownActions();
        }
        else {
            this.isMobile = false;
            this.limitPrimaryActions = 3;
            this.setDropdownActions();
        }
    };
    /**
     * @private
     * @return {?}
     */
    ThfPageDefaultComponent.prototype.setIsMobile = /**
     * @private
     * @return {?}
     */
    function () {
        if (window.innerWidth < this.maxWidthMobile) {
            this.isMobile = true;
            this.limitPrimaryActions = 2;
        }
    };
    /**
     * @return {?}
     */
    ThfPageDefaultComponent.prototype.setDropdownActions = /**
     * @return {?}
     */
    function () {
        if (this.actions.length > this.limitPrimaryActions) {
            this.dropdownActions = this.actions.slice(this.limitPrimaryActions - 1);
        }
    };
    ThfPageDefaultComponent.decorators = [
        { type: Component, args: [{
                    selector: 'thf-page-default',
                    template: "<thf-page>\n\n  <!-- HEADER -->\n  <thf-page-header\n    *ngIf=\"hasPageHeader()\"\n    [t-breadcrumb]=\"breadcrumb\"\n    [t-title]=\"title\">\n\n    <!-- OPERATIONS -->\n    <div class=\"thf-page-header-actions\">\n      <thf-dropdown *ngIf=\"actions.length > limitPrimaryActions\"\n        t-label=\"Outras a\u00E7\u00F5es\"\n        [t-actions]=\"dropdownActions\">\n      </thf-dropdown>\n\n      <thf-button *ngIf=\"actions.length === 3 && actions[2] && !isMobile\"\n        [t-disabled]=\"actionIsDisabled(actions[2])\"\n        [t-label]=\"actions[2].label\"\n        (t-click)=\"callAction(actions[2])\">\n      </thf-button>\n\n      <thf-button *ngIf=\"actions[1] && (actions.length === 2 || !isMobile)\"\n        [t-disabled]=\"actionIsDisabled(actions[1])\"\n        [t-label]=\"actions[1].label\"\n        (t-click)=\"callAction(actions[1])\">\n      </thf-button>\n\n      <thf-button *ngIf=\"actions[0]\"\n        t-type=\"primary\"\n        [t-disabled]=\"actionIsDisabled(actions[0])\"\n        [t-icon]=\"actions[0].icon\"\n        [t-label]=\"actions[0].label\"\n        (t-click)=\"callAction(actions[0])\">\n      </thf-button>\n    </div>\n\n  </thf-page-header>\n\n  <!-- CONTENT -->\n  <thf-page-content>\n    <ng-content></ng-content>\n  </thf-page-content>\n\n</thf-page>\n"
                }] }
    ];
    /** @nocollapse */
    ThfPageDefaultComponent.ctorParameters = function () { return [
        { type: ViewContainerRef },
        { type: Renderer2 },
        { type: Router }
    ]; };
    return ThfPageDefaultComponent;
}(ThfPageDefaultBaseComponent));
export { ThfPageDefaultComponent };
if (false) {
    /** @type {?} */
    ThfPageDefaultComponent.prototype.limitPrimaryActions;
    /** @type {?} */
    ThfPageDefaultComponent.prototype.dropdownActions;
    /** @type {?} */
    ThfPageDefaultComponent.prototype.isMobile;
    /**
     * @type {?}
     * @private
     */
    ThfPageDefaultComponent.prototype.maxWidthMobile;
    /**
     * @type {?}
     * @private
     */
    ThfPageDefaultComponent.prototype.parentRef;
    /**
     * @type {?}
     * @private
     */
    ThfPageDefaultComponent.prototype.renderer;
    /**
     * @type {?}
     * @private
     */
    ThfPageDefaultComponent.prototype.router;
}
