// @totvs/thf-ui/esm2015/lib/components/thf-field/thf-clean/thf-clean.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ViewContainerRef } from '@angular/core';
import { ThfCleanBaseComponent } from './thf-clean-base.component';
/**
 * \@docsPrivate
 *
 * \@docsExtends ThfCleanBaseComponent
 *
 * \@examplePrivate
 *
 * <example-private name="thf-clean-labs" title="Totvs Clean Labs">
 *   <file name="sample-thf-clean-labs.component.html"> </file>
 *   <file name="sample-thf-clean-labs.component.ts"> </file>
 * </example-private>
 */
export class ThfCleanComponent extends ThfCleanBaseComponent {
    /**
     * @param {?} viewRef
     */
    constructor(viewRef) {
        super();
        this.viewRef = viewRef;
        this.parentComponent = this.viewRef['_view']['component'];
    }
    /**
     * @param {?=} value
     * @return {?}
     */
    setInputValue(value) {
        if (this.inputRef && this.inputRef.nativeElement) {
            this.inputRef.nativeElement.value = value;
        }
    }
    /**
     * @return {?}
     */
    getInputValue() {
        if (this.inputRef && this.inputRef.nativeElement) {
            return this.inputRef.nativeElement.value;
        }
    }
}
ThfCleanComponent.decorators = [
    { type: Component, args: [{
                selector: 'thf-clean',
                template: "<span *ngIf=\"showIcon()\"\n  class=\"thf-icon thf-field-icon thf-icon-close\"\n  (click)=\"clear()\">\n</span>\n"
            }] }
];
/** @nocollapse */
ThfCleanComponent.ctorParameters = () => [
    { type: ViewContainerRef }
];
if (false) {
    /**
     * @type {?}
     * @private
     */
    ThfCleanComponent.prototype.viewRef;
}
