// @totvs/thf-ui/esm5/lib/components/thf-tag/thf-tag.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component } from '@angular/core';
import { ThfTagBaseComponent } from './thf-tag-base.component';
/**
 * \@docsExtends ThfTagBaseComponent
 *
 * \@example
 *
 * <example name="thf-tag-basic" title="Totvs Tag Basic">
 *  <file name="sample-thf-tag-basic/sample-thf-tag-basic.component.html"> </file>
 *  <file name="sample-thf-tag-basic/sample-thf-tag-basic.component.ts"> </file>
 * </example>
 *
 * <example name="thf-tag-labs" title="Totvs Tag Labs">
 *  <file name="sample-thf-tag-labs/sample-thf-tag-labs.component.html"> </file>
 *  <file name="sample-thf-tag-labs/sample-thf-tag-labs.component.ts"> </file>
 * </example>
 *
 * <example name="thf-tag-bank-account" title="Totvs Tag - Bank Account">
 *  <file name="sample-thf-tag-bank-account/sample-thf-tag-bank-account.component.html"> </file>
 *  <file name="sample-thf-tag-bank-account/sample-thf-tag-bank-account.component.ts"> </file>
 * </example>
 */
var ThfTagComponent = /** @class */ (function (_super) {
    tslib_1.__extends(ThfTagComponent, _super);
    function ThfTagComponent() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /**
     * @return {?}
     */
    ThfTagComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.isClickable = this.click.observers.length > 0;
    };
    /**
     * @return {?}
     */
    ThfTagComponent.prototype.onClick = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var submittedTagItem = { value: this.value, type: this.type };
        this.click.emit(submittedTagItem);
    };
    ThfTagComponent.decorators = [
        { type: Component, args: [{
                    selector: 'thf-tag',
                    template: "<div class=\"thf-tag-container\" [class.thf-tag-container-horizontal]=\"orientation === thfTagOrientation.Horizontal\">\n  <div *ngIf=\"label\" class=\"thf-tag-title thf-text-nowrap\">\n    <span class=\"thf-tag-label\">{{ orientation === thfTagOrientation.Horizontal ? label + ':' : label }}</span>\n  </div>\n  <div \n    class=\"thf-tag thf-tag-{{ type }}\" \n    [ngClass]=\"{'thf-clickable': isClickable}\"\n    (click)=\"onClick()\">\n    <span *ngIf=\"icon\" class=\"thf-icon thf-icon-{{ iconFromType }}\"></span>\n    <span class=\"thf-tag-value\">{{value}}</span>\n  </div>\n</div>\n"
                }] }
    ];
    return ThfTagComponent;
}(ThfTagBaseComponent));
export { ThfTagComponent };
if (false) {
    /** @type {?} */
    ThfTagComponent.prototype.isClickable;
}
