// @totvs/thf-ui/esm2015/lib/components/thf-stepper/thf-stepper-circle/thf-stepper-circle.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input } from '@angular/core';
import { ThfStepperStatus } from '../enums/thf-stepper-status.enum';
/** @type {?} */
const thfMediumStepSize = 32;
/** @type {?} */
const thfLargeStepSize = 48;
/**
 * \@docsPrivate
 *
 * \@description
 *
 * Componente responsável por exibir os círculos dos *steps*.
 */
export class ThfStepperCircleComponent {
    /**
     * @return {?}
     */
    get isActive() {
        return this.status === ThfStepperStatus.Active;
    }
    /**
     * @return {?}
     */
    get isDefault() {
        return this.status === ThfStepperStatus.Default;
    }
    /**
     * @return {?}
     */
    get isDisabled() {
        return this.status === ThfStepperStatus.Disabled;
    }
    /**
     * @return {?}
     */
    get isDone() {
        return this.status === ThfStepperStatus.Done;
    }
    /**
     * @return {?}
     */
    get isError() {
        return this.status === ThfStepperStatus.Error;
    }
    /**
     * @return {?}
     */
    get isLargeStep() {
        return this.size >= thfLargeStepSize;
    }
    /**
     * @return {?}
     */
    get isMediumStep() {
        return this.size >= thfMediumStepSize && !this.isLargeStep;
    }
}
ThfStepperCircleComponent.decorators = [
    { type: Component, args: [{
                selector: 'thf-stepper-circle',
                template: "<div class=\"thf-stepper-circle\"\n  [class.thf-stepper-circle-with-icon]=\"icons\"\n  [style.height.px]=\"size\"\n  [style.width.px]=\"size\"\n  [tabindex]=\"isDisabled ? -1 : 0\">\n\n  <span class=\"thf-stepper-circle-content\"\n    [class.thf-icon]=\"icons\"\n    [class.thf-icon-exclamation]=\"icons && isError\"\n    [class.thf-icon-info]=\"icons && (isActive || isDefault || isDisabled)\"\n    [class.thf-icon-ok]=\"icons && isDone\"\n    [class.thf-stepper-circle-content-lg]=\"isLargeStep\"\n    [class.thf-stepper-circle-content-md]=\"isMediumStep\">\n\n    {{ !icons ? content: '' }}\n  </span>\n</div>\n"
            }] }
];
ThfStepperCircleComponent.propDecorators = {
    content: [{ type: Input, args: ['t-content',] }],
    icons: [{ type: Input, args: ['t-icons',] }],
    size: [{ type: Input, args: ['t-size',] }],
    status: [{ type: Input, args: ['t-status',] }]
};
if (false) {
    /** @type {?} */
    ThfStepperCircleComponent.prototype.content;
    /** @type {?} */
    ThfStepperCircleComponent.prototype.icons;
    /** @type {?} */
    ThfStepperCircleComponent.prototype.size;
    /** @type {?} */
    ThfStepperCircleComponent.prototype.status;
}
