// @totvs/thf-ui/esm2015/lib/components/thf-field/thf-datepicker-range/thf-datepicker-range.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ElementRef, forwardRef, ViewChild } from '@angular/core';
import { NG_VALIDATORS, NG_VALUE_ACCESSOR } from '@angular/forms';
import { ThfDatepickerRangeBaseComponent } from './thf-datepicker-range-base.component';
import { ThfDateService } from './../../../services/thf-date/thf-date.service';
import { ThfMask } from '../thf-input/thf-mask';
/** @type {?} */
const arrowLeftKey = 37;
/** @type {?} */
const arrowRightKey = 39;
/** @type {?} */
const backspaceKey = 8;
/** @type {?} */
const thfDatepickerRangeDateLengthDefault = 10;
/**
 * \@docsExtends ThfDatepickerRangeBaseComponent
 *
 * \@example
 *
 * <example name="thf-datepicker-range-basic" title="Totvs Datepicker Range Basic">
 *  <file name="sample-thf-datepicker-range-basic/sample-thf-datepicker-range-basic.component.html"> </file>
 *  <file name="sample-thf-datepicker-range-basic/sample-thf-datepicker-range-basic.component.ts"> </file>
 * </example>
 *
 * <example name="thf-datepicker-range-labs" title="Totvs Datepicker Range Labs">
 *  <file name="sample-thf-datepicker-range-labs/sample-thf-datepicker-range-labs.component.html"> </file>
 *  <file name="sample-thf-datepicker-range-labs/sample-thf-datepicker-range-labs.component.ts"> </file>
 * </example>
 *
 * <example name="thf-datepicker-range-vacations" title="Totvs Datepicker Range - Vacations Suggestion">
 *  <file name="sample-thf-datepicker-range-vacations/sample-thf-datepicker-range-vacations.component.html"> </file>
 *  <file name="sample-thf-datepicker-range-vacations/sample-thf-datepicker-range-vacations.component.ts"> </file>
 * </example>
 *
 * <example name="thf-datepicker-range-vacations-reactive-form" title="Totvs Datepicker Range - Vacations Reactive Form">
 *  <file name="sample-thf-datepicker-range-vacations-reactive-form/sample-thf-datepicker-range-vacations-reactive-form.component.html">
 *  </file>
 *  <file name="sample-thf-datepicker-range-vacations-reactive-form/sample-thf-datepicker-range-vacations-reactive-form.component.ts">
 *  </file>
 * </example>
 */
export class ThfDatepickerRangeComponent extends ThfDatepickerRangeBaseComponent {
    /**
     * @param {?} thfDateService
     * @param {?} thfDatepickerRangeElement
     */
    constructor(thfDateService, thfDatepickerRangeElement) {
        super(thfDateService);
        this.thfDatepickerRangeElement = thfDatepickerRangeElement;
    }
    /**
     * @return {?}
     */
    get enableCleaner() {
        return (this.startDateInputValue || this.endDateInputValue) && !this.disabled && !this.readonly;
    }
    /**
     * @return {?}
     */
    get endDateInputName() {
        return 'end-date';
    }
    /**
     * @return {?}
     */
    get endDateInputValue() {
        return this.endDateInput.nativeElement.value;
    }
    /**
     * @return {?}
     */
    get getErrorMessage() {
        return (this.errorMessage !== '' && this.hasInvalidClass()) ? this.errorMessage : '';
    }
    /**
     * @return {?}
     */
    get isDateRangeInputUncompleted() {
        return this.endDateInputValue.length < thfDatepickerRangeDateLengthDefault
            && this.startDateInputValue.length < thfDatepickerRangeDateLengthDefault;
    }
    /**
     * @return {?}
     */
    get isDirtyDateRangeInput() {
        return this.endDateInputValue.length > 0 || this.startDateInputValue.length > 0;
    }
    /**
     * @return {?}
     */
    get startDateInputName() {
        return 'start-date';
    }
    /**
     * @return {?}
     */
    get startDateInputValue() {
        return this.startDateInput.nativeElement.value;
    }
    /**
     * @param {?} event
     * @return {?}
     */
    static getKeyCode(event) {
        return event.keyCode || event.which;
    }
    /**
     * @param {?} event
     * @return {?}
     */
    static getTargetElement(event) {
        return event.target || event.srcElement;
    }
    /**
     * @param {?} keyCode
     * @return {?}
     */
    static isValidKey(keyCode) {
        /** @type {?} */
        const isNumericKey = keyCode >= 48 && keyCode <= 57;
        /** @type {?} */
        const isNumericNumpadKey = keyCode >= 96 && keyCode <= 105;
        return isNumericKey || isNumericNumpadKey;
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        // Classe de máscara
        this.thfMaskObject = this.buildMask();
    }
    /**
     * @return {?}
     */
    clear() {
        this.resetDateRangeInputValidation();
        this.dateRange = { start: '', end: '' };
        this.updateScreenByModel(this.dateRange);
        this.updateModel(this.dateRange);
    }
    /**
     * @return {?}
     */
    onBlur() {
        this.removeFocusFromDatePickerRangeField();
    }
    /**
     * @return {?}
     */
    onFocus() {
        this.applyFocusOnDatePickerRangeField();
    }
    /**
     * @param {?=} event
     * @return {?}
     */
    onKeydown(event) {
        if (this.readonly) {
            return;
        }
        if (this.isSetFocusOnBackspace(event)) {
            event.preventDefault();
            this.setFocusOnBackspace();
        }
        else {
            this.thfMaskObject.keydown(event);
        }
    }
    /**
     * @param {?} event
     * @return {?}
     */
    onKeyup(event) {
        if (this.readonly) {
            return;
        }
        /** @type {?} */
        const isStartDateTargetEvent = event.target.name === this.startDateInputName;
        this.setFocus(event);
        this.thfMaskObject.keyup(event);
        this.updateModelByScreen(isStartDateTargetEvent);
    }
    /**
     * @return {?}
     */
    resetDateRangeInputValidation() {
        this.isStartDateRangeInputValid = true;
        this.isDateRangeInputFormatValid = true;
    }
    /**
     * @param {?} model
     * @return {?}
     */
    updateScreenByModel(model) {
        /** @type {?} */
        const dateRange = { start: (/** @type {?} */ (model.start)), end: (/** @type {?} */ (model.end)) };
        /** @type {?} */
        const isStartDateValid = this.thfDateService.isDateRangeValid(dateRange.end, dateRange.start);
        /** @type {?} */
        const isDateValid = (/**
         * @param {?} date
         * @return {?}
         */
        date => !this.dateFormatFailed(date) && isStartDateValid);
        /** @type {?} */
        const endDateFormated = isDateValid(dateRange.end) ? this.formatModelToScreen(dateRange.end) : '';
        /** @type {?} */
        const startDateFormated = isDateValid(dateRange.start) ? this.formatModelToScreen(dateRange.start) : '';
        this.endDateInput.nativeElement.value = endDateFormated;
        this.startDateInput.nativeElement.value = startDateFormated;
    }
    /**
     * @private
     * @return {?}
     */
    applyFocusOnDatePickerRangeField() {
        this.dateRangeField.nativeElement.classList.add('thf-datepicker-range-field-focused');
    }
    // Retorna um objeto do tipo ThfMask com a mascara configurada.
    /**
     * @private
     * @return {?}
     */
    buildMask() {
        /** @type {?} */
        let mask = this.format.toUpperCase();
        mask = mask.replace(/DD/g, '99');
        mask = mask.replace(/MM/g, '99');
        mask = mask.replace(/YYYY/g, '9999');
        return new ThfMask(mask, true);
    }
    /**
     * @private
     * @param {?} format
     * @param {?=} day
     * @param {?=} month
     * @param {?=} year
     * @return {?}
     */
    formatDate(format, day = '', month = '', year = '') {
        /** @type {?} */
        let dateFormatted = format;
        day = day && day.includes('T') ? day.slice(0, 2) : day;
        dateFormatted = dateFormatted.replace('dd', ('0' + day).slice(-2));
        dateFormatted = dateFormatted.replace('mm', ('0' + (month)).slice(-2));
        dateFormatted = dateFormatted.replace('yyyy', String(year));
        return dateFormatted;
    }
    /**
     * @private
     * @param {?=} value
     * @return {?}
     */
    formatScreenToModel(value = '') {
        const [day, month, year] = value.split('/');
        return value ? this.formatDate('yyyy-mm-dd', day, month, year) : '';
    }
    /**
     * @private
     * @param {?=} value
     * @return {?}
     */
    formatModelToScreen(value = '') {
        const [year, month, day] = value.split('-');
        return value ? this.formatDate(this.format, day, month, year) : '';
    }
    /**
     * @private
     * @param {?} startDate
     * @param {?} endDate
     * @param {?} isStartDateTargetEvent
     * @return {?}
     */
    getDateRangeFormatValidation(startDate, endDate, isStartDateTargetEvent) {
        this.setDateRangeInputValidation(startDate, endDate);
        return {
            isValid: this.isDateRangeInputFormatValid && this.isStartDateRangeInputValid,
            dateRangeModel: this.getValidatedModel(startDate, endDate, isStartDateTargetEvent)
        };
    }
    /**
     * @private
     * @param {?} startDate
     * @param {?} endDate
     * @param {?} isStartDateTargetEvent
     * @return {?}
     */
    getValidatedModel(startDate, endDate, isStartDateTargetEvent) {
        /** @type {?} */
        const dateRangeModel = { start: '', end: '' };
        dateRangeModel.end = (isStartDateTargetEvent || this.isStartDateRangeInputValid)
            && !this.dateFormatFailed(endDate) ? endDate : '';
        dateRangeModel.start = (!isStartDateTargetEvent || this.isStartDateRangeInputValid)
            && !this.dateFormatFailed(startDate) ? startDate : '';
        return dateRangeModel;
    }
    /**
     * @private
     * @return {?}
     */
    hasInvalidClass() {
        return (this.thfDatepickerRangeElement.nativeElement.classList.contains('ng-invalid') &&
            this.thfDatepickerRangeElement.nativeElement.classList.contains('ng-dirty'));
    }
    /**
     * @private
     * @param {?} startDate
     * @param {?} endDate
     * @return {?}
     */
    isEqualBeforeValue(startDate, endDate) {
        return this.isDateRangeInputFormatValid && endDate === this.dateRange.end && startDate === this.dateRange.start;
    }
    /**
     * @private
     * @param {?} event
     * @return {?}
     */
    isSetFocusOnBackspace(event) {
        return event.target.name === this.endDateInputName &&
            this.endDateInput.nativeElement.selectionStart === 0 && this.endDateInput.nativeElement.selectionEnd === 0 &&
            event.keyCode === backspaceKey;
    }
    /**
     * @private
     * @return {?}
     */
    removeFocusFromDatePickerRangeField() {
        this.dateRangeField.nativeElement.classList.remove('thf-datepicker-range-field-focused');
    }
    /**
     * @private
     * @param {?} startDate
     * @param {?} endDate
     * @return {?}
     */
    setDateRangeInputValidation(startDate, endDate) {
        this.isStartDateRangeInputValid = this.thfDateService.isDateRangeValid(endDate, startDate);
        this.isDateRangeInputFormatValid = !this.dateFormatFailed(startDate) && !this.dateFormatFailed(endDate);
    }
    /**
     * @private
     * @param {?} event
     * @return {?}
     */
    setFocus(event) {
        /** @type {?} */
        const inputElement = ThfDatepickerRangeComponent.getTargetElement(event);
        /** @type {?} */
        const keyCode = ThfDatepickerRangeComponent.getKeyCode(event);
        /** @type {?} */
        const inputName = inputElement['name'];
        this.setFocusOnArrowLeft(keyCode, inputName);
        this.setFocusOnArrowRight(keyCode, inputName, inputElement);
        this.setFocusOnStartDateCompleted(keyCode, inputName);
    }
    /**
     * @private
     * @param {?} position
     * @param {?} inputElement
     * @param {?} selectionRange
     * @return {?}
     */
    setFocusAndPosition(position, inputElement, selectionRange) {
        this.focusOnElement(inputElement);
        setTimeout((/**
         * @return {?}
         */
        () => {
            inputElement.nativeElement.setSelectionRange(selectionRange, selectionRange);
            this.thfMaskObject.initialPosition = position;
            this.thfMaskObject.finalPosition = position;
        }));
    }
    /**
     * @private
     * @param {?} inputElement
     * @return {?}
     */
    focusOnElement(inputElement) {
        inputElement.nativeElement.focus();
    }
    /**
     * @private
     * @param {?} keyCode
     * @param {?} inputName
     * @return {?}
     */
    setFocusOnArrowLeft(keyCode, inputName) {
        /** @type {?} */
        const isCursorAtStartOfInput = this.endDateInput.nativeElement.selectionStart === 0;
        if (inputName === this.endDateInputName && isCursorAtStartOfInput && keyCode === arrowLeftKey) {
            /** @type {?} */
            const inputLength = this.startDateInput.nativeElement.value.length;
            this.setFocusAndPosition(inputLength, this.startDateInput, inputLength);
        }
    }
    /**
     * @private
     * @param {?} keyCode
     * @param {?} inputName
     * @param {?} inputElement
     * @return {?}
     */
    setFocusOnArrowRight(keyCode, inputName, inputElement) {
        /** @type {?} */
        const isCursorAtEndOfInput = this.startDateInput.nativeElement.selectionStart === inputElement.value.length;
        if (inputName === this.startDateInputName && isCursorAtEndOfInput && keyCode === arrowRightKey) {
            this.setFocusAndPosition(0, this.endDateInput, 0);
        }
    }
    /**
     * @private
     * @return {?}
     */
    setFocusOnBackspace() {
        /** @type {?} */
        const inputLength = this.startDateInput.nativeElement.value.length;
        this.startDateInput.nativeElement.value = this.startDateInputValue.slice(0, -1);
        this.setFocusAndPosition(inputLength, this.startDateInput, inputLength);
    }
    /**
     * @private
     * @param {?} keyCode
     * @param {?} inputName
     * @return {?}
     */
    setFocusOnStartDateCompleted(keyCode, inputName) {
        /** @type {?} */
        const isLastKeyPressed = this.startDateInput.nativeElement.selectionStart === thfDatepickerRangeDateLengthDefault;
        /** @type {?} */
        const isNewDateCompleted = this.startDateInputValue.length === thfDatepickerRangeDateLengthDefault && isLastKeyPressed;
        /** @type {?} */
        const isValidKey = ThfDatepickerRangeComponent.isValidKey(keyCode);
        if (inputName === this.startDateInputName && isNewDateCompleted && isValidKey) {
            this.setFocusAndPosition(0, this.endDateInput, 0);
        }
    }
    /**
     * @private
     * @param {?} isStartDateTargetEvent
     * @return {?}
     */
    updateModelByScreen(isStartDateTargetEvent) {
        /** @type {?} */
        const endDateFormatted = this.formatScreenToModel(this.endDateInputValue);
        /** @type {?} */
        const startDateFormatted = this.formatScreenToModel(this.startDateInputValue);
        if (this.isDateRangeInputUncompleted && this.isDirtyDateRangeInput) {
            this.updateModel(this.dateRange);
            return;
        }
        if (this.isEqualBeforeValue(startDateFormatted, endDateFormatted)) {
            this.resetDateRangeInputValidation();
            this.validateModel(this.dateRange);
            return;
        }
        /** @type {?} */
        const dateFormatValidation = this.getDateRangeFormatValidation(startDateFormatted, endDateFormatted, isStartDateTargetEvent);
        if (dateFormatValidation.isValid) {
            this.dateRange = { start: startDateFormatted, end: endDateFormatted };
            this.updateModel(this.dateRange);
            this.onChange.emit(Object.assign({}, this.dateRange));
        }
        if (!dateFormatValidation.isValid) {
            this.dateRange = Object.assign({}, dateFormatValidation.dateRangeModel);
            this.updateModel(dateFormatValidation.dateRangeModel);
        }
    }
}
ThfDatepickerRangeComponent.decorators = [
    { type: Component, args: [{
                selector: 'thf-datepicker-range',
                template: "<thf-field-container\n  [t-help]=\"help\"\n  [t-label]=\"label\"\n  [t-optional]=\"!required && optional\">\n\n  <div\n    #dateRangeField\n    class=\"thf-datepicker-range-field thf-input\"\n    [class.thf-datepicker-range-field-disabled]=\"disabled\">\n\n    <div class=\"thf-datepicker-range-start-date\">\n      <input\n        #startDateInput\n        class=\"thf-datepicker-range-input\"\n        maxlength=\"10\"\n        type=\"text\"\n        [disabled]=\"disabled\"\n        [name]=\"startDateInputName\"\n        [readonly]=\"readonly\"\n        (blur)=\"onBlur()\"\n        (focus)=\"onFocus()\"\n        (keydown)=\"onKeydown($event)\"\n        (keyup)=\"onKeyup($event)\">\n    </div>\n\n    <div class=\"thf-datepicker-range-separator\">-</div>\n\n    <div class=\"thf-datepicker-range-end-date\">\n      <input\n        #endDateInput\n        class=\"thf-datepicker-range-input\"\n        maxlength=\"10\"\n        type=\"text\"\n        [disabled]=\"disabled\"\n        [name]=\"endDateInputName\"\n        [readonly]=\"readonly\"\n        (blur)=\"onBlur()\"\n        (focus)=\"onFocus()\"\n        (keydown)=\"onKeydown($event)\"\n        (keyup)=\"onKeyup($event)\">\n    </div>\n\n    <div class=\"thf-datepicker-range-icon\">\n      <thf-clean *ngIf=\"enableCleaner\" (t-change-event)=\"clear($event)\"></thf-clean>\n    </div>\n\n    <div class=\"thf-datepicker-range-icon\">\n      <span\n        class=\"thf-icon thf-field-icon thf-icon-calendar\"\n        [class.thf-field-icon-disabled]=\"disabled || readonly\">\n      </span>\n    </div>\n\n  </div>\n\n  <thf-field-container-bottom [t-error-pattern]=\"getErrorMessage\"></thf-field-container-bottom>\n</thf-field-container>\n",
                providers: [
                    {
                        provide: NG_VALUE_ACCESSOR,
                        useExisting: forwardRef((/**
                         * @return {?}
                         */
                        () => ThfDatepickerRangeComponent)),
                        multi: true,
                    },
                    {
                        provide: NG_VALIDATORS,
                        useExisting: forwardRef((/**
                         * @return {?}
                         */
                        () => ThfDatepickerRangeComponent)),
                        multi: true,
                    }
                ]
            }] }
];
/** @nocollapse */
ThfDatepickerRangeComponent.ctorParameters = () => [
    { type: ThfDateService },
    { type: ElementRef }
];
ThfDatepickerRangeComponent.propDecorators = {
    dateRangeField: [{ type: ViewChild, args: ['dateRangeField', { read: ElementRef, static: true },] }],
    endDateInput: [{ type: ViewChild, args: ['endDateInput', { read: ElementRef, static: true },] }],
    startDateInput: [{ type: ViewChild, args: ['startDateInput', { read: ElementRef, static: true },] }]
};
if (false) {
    /**
     * @type {?}
     * @private
     */
    ThfDatepickerRangeComponent.prototype.thfDatepickerRangeElement;
    /**
     * @type {?}
     * @private
     */
    ThfDatepickerRangeComponent.prototype.thfMaskObject;
    /** @type {?} */
    ThfDatepickerRangeComponent.prototype.dateRangeField;
    /** @type {?} */
    ThfDatepickerRangeComponent.prototype.endDateInput;
    /** @type {?} */
    ThfDatepickerRangeComponent.prototype.startDateInput;
}
