// @totvs/thf-templates/esm5/lib/components/thf-page-login/thf-page-login.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { CommonModule } from '@angular/common';
import { FormsModule } from '@angular/forms';
import { NgModule } from '@angular/core';
import { RouterModule } from '@angular/router';
import { ThfModule } from '@totvs/thf-ui';
import { ThfModalPasswordRecoveryComponent } from '../thf-modal-password-recovery/thf-modal-password-recovery.component';
import { ThfPageBackgroundModule } from '../thf-page-background/index';
import { ThfPageLoginComponent } from './thf-page-login.component';
import { ThfPageLoginPopoverComponent } from './thf-page-login-popover/thf-page-login-popover.component';
import { ThfPageLoginService } from './thf-page-login.service';
/**
 * \@description
 *
 * Módulo do template do thf-page-login.
 */
var ThfPageLoginModule = /** @class */ (function () {
    function ThfPageLoginModule() {
    }
    ThfPageLoginModule.decorators = [
        { type: NgModule, args: [{
                    imports: [
                        CommonModule,
                        FormsModule,
                        RouterModule,
                        ThfPageBackgroundModule,
                        ThfModule
                    ],
                    declarations: [
                        ThfPageLoginComponent,
                        ThfPageLoginPopoverComponent
                    ],
                    entryComponents: [
                        ThfModalPasswordRecoveryComponent
                    ],
                    exports: [
                        ThfPageLoginComponent
                    ],
                    providers: [ThfPageLoginService]
                },] }
    ];
    return ThfPageLoginModule;
}());
export { ThfPageLoginModule };
