// @totvs/thf-templates/esm5/lib/components/thf-page-login/thf-page-login.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { NgForm } from '@angular/forms';
import { ActivatedRoute, Router } from '@angular/router';
import { ChangeDetectorRef, Component, IterableDiffers, ViewChild, ViewContainerRef } from '@angular/core';
import { isExternalLink } from '../../utils/util';
import { ThfComponentInjectorService } from '@totvs/thf-ui';
import { ThfModalPasswordRecoveryComponent } from '../thf-modal-password-recovery/thf-modal-password-recovery.component';
import { ThfModalPasswordRecoveryType } from '../thf-modal-password-recovery/enums/thf-modal-password-recovery-type.enum';
import { ThfPageLoginBaseComponent } from './thf-page-login-base.component';
import { ThfPageLoginService } from './thf-page-login.service';
/**
 * \@docsExtends ThfPageLoginBaseComponent
 *
 * \@example
 *
 * <example name="thf-page-login-basic" title="Totvs Page Login Basic">
 *  <file name="sample-thf-page-login-basic/sample-thf-page-login-basic.component.html"> </file>
 *  <file name="sample-thf-page-login-basic/sample-thf-page-login-basic.component.ts"> </file>
 * </example>
 *
 * <example name="thf-page-login-labs" title="Totvs Page Login Labs">
 *  <file name="sample-thf-page-login-labs/sample-thf-page-login-labs.component.html"> </file>
 *  <file name="sample-thf-page-login-labs/sample-thf-page-login-labs.component.ts"> </file>
 * </example>
 *
 * <example name="thf-page-login-human-resources" title="Totvs Page Login - Human Resources">
 *  <file name="sample-thf-page-login-human-resources/sample-thf-page-login-human-resources.component.html"> </file>
 *  <file name="sample-thf-page-login-human-resources/sample-thf-page-login-human-resources.component.ts"> </file>
 *  <file name="sample-thf-page-login-human-resources/sample-thf-page-login-human-resources.module.ts"> </file>
 * </example>
 *
 * <example name="thf-page-login-automatic-service" title="Totvs Page Login - Automatic Service">
 *  <file name="sample-thf-page-login-automatic-service/sample-thf-page-login-automatic-service.component.html"> </file>
 *  <file name="sample-thf-page-login-automatic-service/sample-thf-page-login-automatic-service.component.ts"> </file>
 * </example>
 */
var ThfPageLoginComponent = /** @class */ (function (_super) {
    tslib_1.__extends(ThfPageLoginComponent, _super);
    function ThfPageLoginComponent(changeDetector, activatedRoute, thfComponentInjector, differs, loginService, router) {
        var _this = _super.call(this, loginService, router) || this;
        _this.changeDetector = changeDetector;
        _this.activatedRoute = activatedRoute;
        _this.thfComponentInjector = thfComponentInjector;
        _this.componentRef = null;
        _this.customPasswordError = { custom: false };
        _this.differ = differs.find([]).create(null);
        return _this;
    }
    /**
     * @return {?}
     */
    ThfPageLoginComponent.prototype.ngAfterViewChecked = /**
     * @return {?}
     */
    function () {
        if (this.differ) {
            this.validateArrayChanges(this.differ, [
                { array: this.loginErrors, callback: this.generateLoginError.bind(this) },
                { array: this.passwordErrors, callback: this.generatePasswordError.bind(this) }
            ]);
        }
    };
    /**
     * @return {?}
     */
    ThfPageLoginComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.checkingForRouteMetadata(this.activatedRoute.snapshot.data);
    };
    /**
     * @return {?}
     */
    ThfPageLoginComponent.prototype.changeLoginModel = /**
     * @return {?}
     */
    function () {
        if (this.authenticationUrl) {
            this.loginErrors = [];
        }
        else {
            this.setLoginErrors(this.loginErrors);
            this.loginChange.emit(this.login);
        }
    };
    /**
     * @return {?}
     */
    ThfPageLoginComponent.prototype.changePasswordModel = /**
     * @return {?}
     */
    function () {
        if (this.authenticationUrl) {
            this.passwordErrors = [];
        }
        else {
            this.setPasswordErrors(this.passwordErrors);
            this.passwordChange.emit(this.password);
        }
    };
    /**
     * @private
     * @param {?} object
     * @param {?} property
     * @return {?}
     */
    ThfPageLoginComponent.prototype.checkingForMetadataProperty = /**
     * @private
     * @param {?} object
     * @param {?} property
     * @return {?}
     */
    function (object, property) {
        if (Object.prototype.hasOwnProperty.call(object, property)) {
            return object[property];
        }
    };
    /**
     * @private
     * @param {?} data
     * @return {?}
     */
    ThfPageLoginComponent.prototype.checkingForRouteMetadata = /**
     * @private
     * @param {?} data
     * @return {?}
     */
    function (data) {
        if (Object.keys(data).length !== 0) {
            this.authenticationUrl = this.checkingForMetadataProperty(data, 'serviceApi') || this.authenticationUrl;
            this.authenticationType = this.checkingForMetadataProperty(data, 'authenticationType') || this.authenticationType;
            this.environment = this.checkingForMetadataProperty(data, 'environment') || this.environment;
            this.recovery = this.checkingForMetadataProperty(data, 'recovery') || this.recovery;
            this.registerUrl = this.checkingForMetadataProperty(data, 'registerUrl') || this.registerUrl;
        }
    };
    /**
     * @private
     * @param {?} thfPageLoginRecovery
     * @return {?}
     */
    ThfPageLoginComponent.prototype.createModalPasswordRecoveryComponent = /**
     * @private
     * @param {?} thfPageLoginRecovery
     * @return {?}
     */
    function (thfPageLoginRecovery) {
        var _this = this;
        if (this.componentRef) {
            this.thfComponentInjector.destroyComponentInApplication(this.componentRef);
        }
        this.componentRef = this.thfComponentInjector.createComponentInApplication(ThfModalPasswordRecoveryComponent);
        this.componentRef.instance.urlRecovery = thfPageLoginRecovery.url;
        this.componentRef.instance.contactEmail = thfPageLoginRecovery.contactMail;
        this.componentRef.instance.phoneMask = thfPageLoginRecovery.phoneMask;
        this.componentRef.instance.type = thfPageLoginRecovery.type || ThfModalPasswordRecoveryType.Email;
        this.componentRef.changeDetectorRef.detectChanges();
        setTimeout((/**
         * @return {?}
         */
        function () {
            _this.componentRef.instance.open();
        }));
    };
    /**
     * @private
     * @return {?}
     */
    ThfPageLoginComponent.prototype.generateLoginError = /**
     * @private
     * @return {?}
     */
    function () {
        if (this.loginErrors && this.loginErrors.length) {
            this.setLoginErrors(this.loginErrors);
        }
        else {
            /** @type {?} */
            var control = this.loginForm.form.controls['login'];
            if (control) {
                this.resetControl(control);
            }
        }
    };
    /**
     * @private
     * @return {?}
     */
    ThfPageLoginComponent.prototype.generatePasswordError = /**
     * @private
     * @return {?}
     */
    function () {
        if (this.passwordErrors && this.passwordErrors.length) {
            this.setPasswordErrors(this.passwordErrors);
        }
        else {
            /** @type {?} */
            var control = this.loginForm.form.controls['password'];
            if (control) {
                this.resetControl(control);
            }
        }
    };
    /**
     * @private
     * @param {?} control
     * @return {?}
     */
    ThfPageLoginComponent.prototype.resetControl = /**
     * @private
     * @param {?} control
     * @return {?}
     */
    function (control) {
        control.markAsPristine();
        control.markAsUntouched();
        control.updateValueAndValidity();
    };
    /**
     * @private
     * @param {?} allErrors
     * @param {?} control
     * @param {?} errors
     * @param {?} patternError
     * @return {?}
     */
    ThfPageLoginComponent.prototype.setControlErrors = /**
     * @private
     * @param {?} allErrors
     * @param {?} control
     * @param {?} errors
     * @param {?} patternError
     * @return {?}
     */
    function (allErrors, control, errors, patternError) {
        if (control) {
            this[allErrors] = control.hasError('pattern') ? tslib_1.__spread(errors, [patternError]) : tslib_1.__spread(errors);
            if (errors && errors.length && (control.valid || control.pristine)) {
                control.markAsTouched();
                control.markAsDirty();
                control.setErrors(this.customPasswordError);
            }
        }
    };
    /**
     * @private
     * @param {?} url
     * @return {?}
     */
    ThfPageLoginComponent.prototype.setUrlRedirect = /**
     * @private
     * @param {?} url
     * @return {?}
     */
    function (url) {
        isExternalLink(url) ? window.open(url, '_blank') : this.router.navigate([url]);
    };
    /**
     * @private
     * @param {?} differ
     * @param {?} array
     * @return {?}
     */
    ThfPageLoginComponent.prototype.validateArrayChanges = /**
     * @private
     * @param {?} differ
     * @param {?} array
     * @return {?}
     */
    function (differ, array) {
        var _this = this;
        array.forEach((/**
         * @param {?} element
         * @return {?}
         */
        function (element) {
            /** @type {?} */
            var changes = differ.diff(element.array);
            if (changes) {
                element.callback();
                _this.changeDetector.detectChanges();
            }
        }));
    };
    /**
     * @protected
     * @param {?} recovery
     * @return {?}
     */
    ThfPageLoginComponent.prototype.openUrl = /**
     * @protected
     * @param {?} recovery
     * @return {?}
     */
    function (recovery) {
        switch (typeof recovery) {
            case 'string': {
                this.setUrlRedirect(recovery);
                break;
            }
            case 'function': {
                recovery();
                break;
            }
            case 'object': {
                this.createModalPasswordRecoveryComponent(recovery);
                break;
            }
        }
    };
    /**
     * @protected
     * @param {?} errors
     * @return {?}
     */
    ThfPageLoginComponent.prototype.setLoginErrors = /**
     * @protected
     * @param {?} errors
     * @return {?}
     */
    function (errors) {
        /** @type {?} */
        var control = this.loginForm.form.controls['login'];
        this.setControlErrors('allLoginErrors', control, errors, this.literals.loginErrorPattern);
    };
    /**
     * @protected
     * @param {?} errors
     * @return {?}
     */
    ThfPageLoginComponent.prototype.setPasswordErrors = /**
     * @protected
     * @param {?} errors
     * @return {?}
     */
    function (errors) {
        /** @type {?} */
        var control = this.loginForm.form.controls['password'];
        this.setControlErrors('allPasswordErrors', control, errors, this.literals.passwordErrorPattern);
    };
    /**
     * @param {?} language
     * @return {?}
     */
    ThfPageLoginComponent.prototype.onSelectedLanguage = /**
     * @param {?} language
     * @return {?}
     */
    function (language) {
        this.selectedLanguage = language;
        this.getLiterals(language, this.containsCustomLiterals ? this.customizedDefaultLiterals : undefined);
        this.setTitleLiteral(language, this.productName);
        this.setLoginHintLiteral(language, this.contactEmail);
    };
    ThfPageLoginComponent.decorators = [
        { type: Component, args: [{
                    selector: 'thf-page-login',
                    template: "<thf-page-background #pageLogin\n  t-show-select-language\n  [t-background]=\"background\"\n  [t-highlight-info]=\"literals.highlightInfo\"\n  [t-logo]=\"logo\"\n  [t-secondary-logo]=\"secondaryLogo\"\n  (t-selected-language)=\"onSelectedLanguage($event)\">\n\n  <div class=\"thf-page-login-header thf-mb-2 thf-mb-sm-1 thf-pt-sm-1\">\n    <div class=\"thf-page-login-header-title thf-mb-sm-1\">{{ literals.title }}</div>\n    <thf-tag\n      *ngIf=\"environment\"\n      t-type=\"warning\"\n      [t-value]=\"environment\">\n    </thf-tag>\n  </div>\n\n  <form #loginForm=\"ngForm\" class=\"thf-page-login-form\">\n    <div class=\"thf-row\">\n      <div class=\"thf-lg-12\">\n        <div class=\"thf-page-login-hint thf-page-login-info-container\">\n          <thf-login\n            class=\"thf-page-login-info-field\"\n            name=\"login\"\n            [(ngModel)]=\"login\"\n            t-focus\n            t-required\n            [t-label]=\"literals.loginLabel\"\n            [t-pattern]=\"loginPattern\"\n            [t-placeholder]=\"literals.loginPlaceholder\"\n            (click)=\"closePopover()\"\n            (keyup.enter)=\"loginForm.valid && onLoginSubmit()\"\n            (t-change-model)=\"changeLoginModel()\">\n          </thf-login>\n\n          <div class=\"thf-page-login-info-icon-container\">\n            <span\n              class=\"thf-icon thf-field-icon thf-icon-info\"\n              t-tooltip-position=\"right\"\n              [t-tooltip]=\"literals.loginHint\">\n            </span>\n          </div>\n        </div>\n\n        <div class=\"thf-field-container-bottom thf-field-container-error-container\">\n          <div *ngFor=\"let error of allLoginErrors\u001C\" class=\"thf-field-container-bottom-text-error thf-field-container-error-item\">\n            <span class=\"thf-icon thf-icon-exclamation\"></span>\n            <span class=\"thf-field-container-error-text\">{{error}}</span>\n          </div>\n        </div>\n      </div>\n\n      <div class=\"thf-lg-12\">\n        <div class=\"thf-page-login-password-container\">\n          <thf-password\n            class=\"thf-page-login-field-size thf-page-login-password-item\"\n            name=\"password\"\n            [(ngModel)]=\"password\"\n            t-required\n            [t-label]=\"literals.passwordLabel\"\n            [t-pattern]=\"passwordPattern\"\n            [t-placeholder]=\"literals.passwordPlaceholder\"\n            (click)=\"closePopover()\"\n            (keyup.enter)=\"loginForm.valid && onLoginSubmit()\"\n            (t-change-model)=\"changePasswordModel()\">\n          </thf-password>\n          <div class=\"thf-page-login-password-item thf-page-login-password-popover-container\">\n            <thf-page-login-popover *ngIf=\"showExceededAttemptsWarning && exceededAttemptsWarning\"\n              [t-recovery]=\"recovery\"\n              [t-remaining-attempts]=\"exceededAttemptsWarning\"\n              [t-selected-language]=\"selectedLanguage\"\n              (t-forgot-password) = \"openUrl($event)\">\n            </thf-page-login-popover>\n          </div>\n        </div>\n        <div class=\"thf-field-container-bottom thf-field-container-error-container\">\n          <div *ngFor=\"let error of allPasswordErrors\" class=\"thf-field-container-bottom-text-error thf-field-container-error-item\">\n            <span class=\"thf-icon thf-icon-exclamation\"></span>\n            <span class=\"thf-field-container-error-text\">{{error}}</span>\n          </div>\n        </div>\n      </div>\n\n      <thf-input *ngIf=\"customField && customFieldType === 'input'\"\n        class=\"thf-page-login-field-size thf-lg-12\"\n        name=\"customFieldInput\"\n        [(ngModel)]=\"customFieldObject.value\"\n        t-required\n        [t-error-pattern]=\"customFieldObject.errorPattern || literals.customFieldErrorPattern\"\n        [t-pattern]=\"customFieldObject.pattern\"\n        [t-placeholder]=\"customFieldObject.placeholder || literals.customFieldPlaceholder\"\n        (keyup.enter)=\"loginForm.valid && onLoginSubmit()\">\n      </thf-input>\n\n      <thf-combo *ngIf=\"customField && customFieldType === 'combo'\"\n        class=\"thf-page-login-field-size thf-lg-12\"\n        name=\"customFieldCombo\"\n        [(ngModel)]=\"customFieldObject.value\"\n        t-required\n        [t-field-value]=\"customFieldObject.fieldValue\"\n        [t-filter-service]=\"customFieldObject.url\"\n        [t-placeholder]=\"customFieldObject.placeholder || literals.customFieldPlaceholder\">\n      </thf-combo>\n\n      <thf-select *ngIf=\"customField && customFieldType === 'select'\"\n        class=\"thf-page-login-field-size thf-lg-12\"\n        name=\"customFieldSelect\"\n        [(ngModel)]=\"customFieldObject.value\"\n        t-required\n        [t-placeholder]=\"customFieldObject.placeholder || literals.customFieldPlaceholder\"\n        [t-options]=\"customFieldObject.options\">\n      </thf-select>\n\n      <div class=\"thf-lg-12\">\n        <div *ngIf=\"!hideRememberUser\" class=\"thf-page-login-hint thf-page-login-info-container\">\n          <thf-switch\n            class=\"thf-page-login-info-field thf-lg-7 thf-offset-lg-5 thf-offset-xl-5\"\n            name=\"rememberUser\"\n            [(ngModel)]=\"rememberUser\"\n            t-label-position=\"1\"\n            [t-label-off]=\"literals.rememberUser\"\n            [t-label-on]=\"literals.rememberUser\"\n            (keyup.enter)=\"loginForm.valid && onLoginSubmit()\">\n          </thf-switch>\n\n          <div class=\"thf-page-login-info-icon-container thf-page-login-info-icon-remember-user\">\n            <span\n              class=\"thf-icon thf-field-icon thf-icon-info\"\n              t-tooltip-position=\"right\"\n              [t-tooltip]=\"literals.rememberUserHint\">\n            </span>\n          </div>\n        </div>\n      </div>\n\n\n      <thf-button\n        class=\"thf-lg-12 thf-page-login-button thf-page-login-field-size\"\n        t-type=\"primary\"\n        [t-disabled]=\"loginForm.invalid\"\n        [t-label]=\"loading ? literals.submittedLabel : literals.submitLabel\"\n        [t-loading]=\"loading\"\n        (t-click)=\"onLoginSubmit()\">\n      </thf-button>\n\n      <div *ngIf=\"recovery\"\n        class=\"thf-page-login-recovery-link\">\n        <a class=\"thf-font-text-large-bold\" (click)=\"openUrl(recovery)\">{{ literals.forgotPassword }}</a>\n      </div>\n\n      <div *ngIf=\"registerUrl\"\n        class=\"thf-page-login-register-link\">\n        <a class=\"thf-font-text-large-bold\" (click)=\"openUrl(registerUrl)\">{{ literals.registerUrl }}</a>\n      </div>\n    </div>\n  </form>\n</thf-page-background>\n"
                }] }
    ];
    /** @nocollapse */
    ThfPageLoginComponent.ctorParameters = function () { return [
        { type: ChangeDetectorRef },
        { type: ActivatedRoute },
        { type: ThfComponentInjectorService },
        { type: IterableDiffers },
        { type: ThfPageLoginService },
        { type: Router }
    ]; };
    ThfPageLoginComponent.propDecorators = {
        loginForm: [{ type: ViewChild, args: ['loginForm', { read: NgForm, static: true },] }],
        pageLogin: [{ type: ViewChild, args: ['pageLogin', { read: ViewContainerRef, static: true },] }]
    };
    return ThfPageLoginComponent;
}(ThfPageLoginBaseComponent));
export { ThfPageLoginComponent };
if (false) {
    /**
     * @type {?}
     * @private
     */
    ThfPageLoginComponent.prototype.componentRef;
    /**
     * @type {?}
     * @private
     */
    ThfPageLoginComponent.prototype.differ;
    /**
     * @type {?}
     * @private
     */
    ThfPageLoginComponent.prototype.customPasswordError;
    /** @type {?} */
    ThfPageLoginComponent.prototype.loginForm;
    /** @type {?} */
    ThfPageLoginComponent.prototype.pageLogin;
    /** @type {?} */
    ThfPageLoginComponent.prototype.changeDetector;
    /**
     * @type {?}
     * @private
     */
    ThfPageLoginComponent.prototype.activatedRoute;
    /**
     * @type {?}
     * @private
     */
    ThfPageLoginComponent.prototype.thfComponentInjector;
}
