// @totvs/thf-ui/esm5/lib/components/thf-slide/thf-slide.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { RouterModule } from '@angular/router';
import { ThfContainerModule } from '../thf-container/thf-container.module';
import { ThfSlideCirclesComponent } from './thf-slide-circles/thf-slide-circles.component';
import { ThfSlideComponent } from './thf-slide.component';
import { ThfSlideContentTemplateDirective } from './directives/thf-slide-content-template.directive';
import { ThfSlideControlComponent } from './thf-slide-control/thf-slide-control.component';
import { ThfSlideItemComponent } from './thf-slide-item/thf-slide-item.component';
/**
 * \@description
 *
 * Módulo do componente `thf-slide`.
 *
 * > Para o correto funcionamento do componente `thf-slide`, deve ser importado o módulo `BrowserAnimationsModule` no
 * > módulo principal da sua aplicação.
 *
 * Módulo da aplicação:
 * ```
 * import { BrowserAnimationsModule } from '\@angular/platform-browser/animations';
 * import { ThfModule } from '\@totvs/thf-ui';
 * ...
 *
 * \@NgModule({
 *   imports: [
 *     BrowserModule,
 *     BrowserAnimationsModule,
 *     ...
 *     ThfModule
 *   ],
 *   declarations: [
 *     AppComponent,
 *     ...
 *   ],
 *   providers: [],
 *   bootstrap: [AppComponent]
 * })
 * export class AppModule { }
 * ```
 */
var ThfSlideModule = /** @class */ (function () {
    function ThfSlideModule() {
    }
    ThfSlideModule.decorators = [
        { type: NgModule, args: [{
                    imports: [
                        CommonModule,
                        RouterModule,
                        ThfContainerModule
                    ],
                    declarations: [
                        ThfSlideCirclesComponent,
                        ThfSlideComponent,
                        ThfSlideControlComponent,
                        ThfSlideContentTemplateDirective,
                        ThfSlideItemComponent,
                    ],
                    exports: [
                        ThfSlideComponent,
                        ThfSlideContentTemplateDirective
                    ],
                    providers: [],
                    schemas: []
                },] }
    ];
    return ThfSlideModule;
}());
export { ThfSlideModule };
