// @totvs/thf-templates/esm5/lib/components/thf-page-change-password/thf-page-change-password.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { CommonModule } from '@angular/common';
import { FormsModule } from '@angular/forms';
import { NgModule } from '@angular/core';
import { RouterModule } from '@angular/router';
import { ThfModule } from '@totvs/thf-ui';
import { ThfModalPasswordRecoveryComponent } from '../thf-modal-password-recovery/thf-modal-password-recovery.component';
import { ThfPageBackgroundModule } from '../thf-page-background/index';
import { ThfPageChangePasswordComponent } from './thf-page-change-password.component';
import { ThfPageChangePasswordService } from './thf-page-change-password.service';
/**
 * \@description
 *
 * Módulo do template do thf-page-change-password.
 */
var ThfPageChangePasswordModule = /** @class */ (function () {
    function ThfPageChangePasswordModule() {
    }
    ThfPageChangePasswordModule.decorators = [
        { type: NgModule, args: [{
                    imports: [
                        CommonModule,
                        FormsModule,
                        RouterModule,
                        ThfModule,
                        ThfPageBackgroundModule
                    ],
                    declarations: [
                        ThfPageChangePasswordComponent
                    ],
                    entryComponents: [
                        ThfModalPasswordRecoveryComponent
                    ],
                    providers: [
                        ThfPageChangePasswordService
                    ],
                    exports: [
                        ThfPageChangePasswordComponent
                    ]
                },] }
    ];
    return ThfPageChangePasswordModule;
}());
export { ThfPageChangePasswordModule };
