// @totvs/thf-ui/esm5/lib/components/thf-stepper/thf-stepper-circle/thf-stepper-circle.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input } from '@angular/core';
import { ThfStepperStatus } from '../enums/thf-stepper-status.enum';
/** @type {?} */
var thfMediumStepSize = 32;
/** @type {?} */
var thfLargeStepSize = 48;
/**
 * \@docsPrivate
 *
 * \@description
 *
 * Componente responsável por exibir os círculos dos *steps*.
 */
var ThfStepperCircleComponent = /** @class */ (function () {
    function ThfStepperCircleComponent() {
    }
    Object.defineProperty(ThfStepperCircleComponent.prototype, "isActive", {
        get: /**
         * @return {?}
         */
        function () {
            return this.status === ThfStepperStatus.Active;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ThfStepperCircleComponent.prototype, "isDefault", {
        get: /**
         * @return {?}
         */
        function () {
            return this.status === ThfStepperStatus.Default;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ThfStepperCircleComponent.prototype, "isDisabled", {
        get: /**
         * @return {?}
         */
        function () {
            return this.status === ThfStepperStatus.Disabled;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ThfStepperCircleComponent.prototype, "isDone", {
        get: /**
         * @return {?}
         */
        function () {
            return this.status === ThfStepperStatus.Done;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ThfStepperCircleComponent.prototype, "isError", {
        get: /**
         * @return {?}
         */
        function () {
            return this.status === ThfStepperStatus.Error;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ThfStepperCircleComponent.prototype, "isLargeStep", {
        get: /**
         * @return {?}
         */
        function () {
            return this.size >= thfLargeStepSize;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ThfStepperCircleComponent.prototype, "isMediumStep", {
        get: /**
         * @return {?}
         */
        function () {
            return this.size >= thfMediumStepSize && !this.isLargeStep;
        },
        enumerable: true,
        configurable: true
    });
    ThfStepperCircleComponent.decorators = [
        { type: Component, args: [{
                    selector: 'thf-stepper-circle',
                    template: "<div class=\"thf-stepper-circle\"\n  [class.thf-stepper-circle-with-icon]=\"icons\"\n  [style.height.px]=\"size\"\n  [style.width.px]=\"size\"\n  [tabindex]=\"isDisabled ? -1 : 0\">\n\n  <span class=\"thf-stepper-circle-content\"\n    [class.thf-icon]=\"icons\"\n    [class.thf-icon-exclamation]=\"icons && isError\"\n    [class.thf-icon-info]=\"icons && (isActive || isDefault || isDisabled)\"\n    [class.thf-icon-ok]=\"icons && isDone\"\n    [class.thf-stepper-circle-content-lg]=\"isLargeStep\"\n    [class.thf-stepper-circle-content-md]=\"isMediumStep\">\n\n    {{ !icons ? content: '' }}\n  </span>\n</div>\n"
                }] }
    ];
    ThfStepperCircleComponent.propDecorators = {
        content: [{ type: Input, args: ['t-content',] }],
        icons: [{ type: Input, args: ['t-icons',] }],
        size: [{ type: Input, args: ['t-size',] }],
        status: [{ type: Input, args: ['t-status',] }]
    };
    return ThfStepperCircleComponent;
}());
export { ThfStepperCircleComponent };
if (false) {
    /** @type {?} */
    ThfStepperCircleComponent.prototype.content;
    /** @type {?} */
    ThfStepperCircleComponent.prototype.icons;
    /** @type {?} */
    ThfStepperCircleComponent.prototype.size;
    /** @type {?} */
    ThfStepperCircleComponent.prototype.status;
}
