// @totvs/thf-ui/esm5/lib/components/thf-dynamic/thf-dynamic-view/thf-dynamic-view.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component } from '@angular/core';
import { CurrencyPipe, DatePipe, DecimalPipe, TitleCasePipe } from '@angular/common';
import { ThfTimePipe } from '../../../pipes/thf-time/thf-time.pipe';
import { ThfDynamicViewBaseComponent } from './thf-dynamic-view-base.component';
/**
 * \@docsExtends ThfDynamicViewBaseComponent
 *
 * \@example
 *
 * <example name="thf-dynamic-view-basic" title="Totvs Dynamic View Basic">
 *  <file name="sample-thf-dynamic-view-basic/sample-thf-dynamic-view-basic.component.html"> </file>
 *  <file name="sample-thf-dynamic-view-basic/sample-thf-dynamic-view-basic.component.ts"> </file>
 * </example>
 *
 * <example name="thf-dynamic-view-employee" title="Totvs Dynamic View - Employee">
 *  <file name="sample-thf-dynamic-view-employee/sample-thf-dynamic-view-employee.component.html"> </file>
 *  <file name="sample-thf-dynamic-view-employee/sample-thf-dynamic-view-employee.component.ts"> </file>
 * </example>
 */
var ThfDynamicViewComponent = /** @class */ (function (_super) {
    tslib_1.__extends(ThfDynamicViewComponent, _super);
    function ThfDynamicViewComponent(currencyPipe, datePipe, decimalPipe, timePipe, titleCasePipe) {
        return _super.call(this, currencyPipe, datePipe, decimalPipe, timePipe, titleCasePipe) || this;
    }
    /**
     * @param {?} changes
     * @return {?}
     */
    ThfDynamicViewComponent.prototype.ngOnChanges = /**
     * @param {?} changes
     * @return {?}
     */
    function (changes) {
        if (changes.fields || changes.value || changes.showAllValue) {
            this.visibleFields = this.getVisibleFields();
        }
    };
    /**
     * @private
     * @return {?}
     */
    ThfDynamicViewComponent.prototype.getVisibleFields = /**
     * @private
     * @return {?}
     */
    function () {
        if (this.showAllValue) {
            return this.getMergedFields();
        }
        return this.value && this.fields.length ? this.getConfiguredFields() : this.getValueFields();
    };
    ThfDynamicViewComponent.decorators = [
        { type: Component, args: [{
                    selector: 'thf-dynamic-view',
                    template: "<div class=\"thf-row\" *ngIf=\"visibleFields.length\">\n\n  <ng-template ngFor let-field [ngForOf]=\"visibleFields\">\n\n    <thf-divider *ngIf=\"field?.divider?.trim()\" class=\"thf-sm-12\" [t-label]=\"field.divider\">\n    </thf-divider>\n\n    <ng-container *ngTemplateOutlet=\"!field.tag ? thfInfo : thfTag; context: { $implicit: field }\">\n    </ng-container>\n\n  </ng-template>\n\n</div>\n\n<ng-template #thfInfo let-field>\n  <thf-info [ngClass]=\"field.cssClass\"\n    [t-label]=\"field.label\"\n    [t-value]=\"field.value\">\n  </thf-info>\n</ng-template>\n\n<ng-template #thfTag let-field>\n  <thf-tag [ngClass]=\"field.cssClass\"\n    [t-label]=\"field.label\"\n    [t-value]=\"field.value\">\n  </thf-tag>\n</ng-template>\n"
                }] }
    ];
    /** @nocollapse */
    ThfDynamicViewComponent.ctorParameters = function () { return [
        { type: CurrencyPipe },
        { type: DatePipe },
        { type: DecimalPipe },
        { type: ThfTimePipe },
        { type: TitleCasePipe }
    ]; };
    return ThfDynamicViewComponent;
}(ThfDynamicViewBaseComponent));
export { ThfDynamicViewComponent };
