// @totvs/thf-ui/esm2015/lib/components/thf-slide/thf-slide-item/thf-slide-item.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ElementRef, Input, ViewChild } from '@angular/core';
import { isIEOrEdge } from '../../../utils/util';
import { ThfSlideContentTemplateDirective } from '../directives/thf-slide-content-template.directive';
/**
 * \@docsPrivate
 *
 * \@description
 *
 * Componente para cada item do thf-slide.
 */
export class ThfSlideItemComponent {
    constructor() {
        this.isIEOrEdge = isIEOrEdge();
    }
    /**
     * @return {?}
     */
    setLinkType() {
        if (!this.template && this.link) {
            return this.link.startsWith('http') ? 'externalLink' : 'internalLink';
        }
        return 'noLink';
    }
}
ThfSlideItemComponent.decorators = [
    { type: Component, args: [{
                selector: 'thf-slide-item',
                template: "<div #slideItem\n  class=\"thf-slide-item\"\n  [class.thf-slide-item-background-image]=\"!template && isIEOrEdge\"\n  [style.background-image]=\"!template && isIEOrEdge ? 'url(' + image + ')' : ''\"\n  [ngSwitch]=\"setLinkType()\">\n\n  <!-- slide com link interno -->\n  <a\n    *ngSwitchCase=\"'internalLink'\" \n    class=\"thf-slide-item-link\" \n    [routerLink]=\"link\">\n    <ng-container \n      [ngTemplateOutlet]=\"slideItemImage\" \n      [ngTemplateOutletContext]=\"{ item: { alt: alt, image: image }}\">\n    </ng-container>\n  </a>\n  <!-- slide com link externo -->\n  <a \n    *ngSwitchCase=\"'externalLink'\"\n    class=\"thf-slide-item-link\" \n    [href]=\"link\">\n    <ng-container \n      [ngTemplateOutlet]=\"slideItemImage\" \n      [ngTemplateOutletContext]=\"{ item: { alt: alt, image: image }}\">\n    </ng-container>\n  </a>\n  <!-- slide sem link -->\n  <a \n    *ngSwitchCase=\"'noLink'\"\n    [ngClass]=\"action ? 'thf-slide-item-link' : 'thf-slide-item-no-link'\"\n    (click)=\"action ? action(data) : undefined\">\n    <ng-container \n      [ngTemplateOutlet]=\"template ? slideItemTemplate : slideItemImage\"\n      [ngTemplateOutletContext]=\"{ item: { alt: alt, image: image, data: data }}\">\n    </ng-container>\n  </a>\n\n</div>\n\n<ng-template #slideItemTemplate let-item='item'>\n  <div\n    *ngIf=\"template\"\n    class=\"thf-slide-item-content\">\n    <ng-template\n      [ngTemplateOutlet]=\"template.templateRef\"\n      [ngTemplateOutletContext]=\"{ $implicit: item.data, index: index }\">\n    </ng-template>\n  </div>\n</ng-template>\n\n<ng-template #slideItemImage let-item='item'>\n  <img\n    *ngIf=\"item.image && !isIEOrEdge\"\n    class=\"thf-slide-image\"\n    [alt]=\"item.alt\"\n    [src]=\"item.image\"\n    [style.height.px]=\"imageHeight\">\n</ng-template>\n"
            }] }
];
ThfSlideItemComponent.propDecorators = {
    itemElement: [{ type: ViewChild, args: ['slideItem', { static: true },] }],
    action: [{ type: Input, args: ['t-action',] }],
    alt: [{ type: Input, args: ['t-alt',] }],
    data: [{ type: Input, args: ['t-data',] }],
    image: [{ type: Input, args: ['t-image',] }],
    imageHeight: [{ type: Input, args: ['t-image-height',] }],
    template: [{ type: Input, args: ['t-template',] }],
    link: [{ type: Input, args: ['t-link',] }]
};
if (false) {
    /** @type {?} */
    ThfSlideItemComponent.prototype.itemElement;
    /** @type {?} */
    ThfSlideItemComponent.prototype.isIEOrEdge;
    /**
     * Ação executada ao clicar em uma imagem
     * @type {?}
     */
    ThfSlideItemComponent.prototype.action;
    /**
     * Texto alternativo quando a imagem não é encontrada
     * @type {?}
     */
    ThfSlideItemComponent.prototype.alt;
    /**
     * Dados para o template customizado
     * @type {?}
     */
    ThfSlideItemComponent.prototype.data;
    /**
     * Caminho da imagem
     * @type {?}
     */
    ThfSlideItemComponent.prototype.image;
    /**
     * Altura da imagem
     * @type {?}
     */
    ThfSlideItemComponent.prototype.imageHeight;
    /**
     * Template customizado
     * @type {?}
     */
    ThfSlideItemComponent.prototype.template;
    /**
     * Link executado ao clicar em uma imagem
     * @type {?}
     */
    ThfSlideItemComponent.prototype.link;
}
