// @totvs/thf-ui/esm5/lib/services/thf-notification/thf-notification.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { ThfToasterComponent } from './thf-toaster/thf-toaster.component';
var ThfNotificationModule = /** @class */ (function () {
    function ThfNotificationModule() {
    }
    ThfNotificationModule.decorators = [
        { type: NgModule, args: [{
                    declarations: [
                        ThfToasterComponent
                    ],
                    imports: [
                        CommonModule
                    ],
                    exports: [],
                    entryComponents: [
                        ThfToasterComponent
                    ],
                    providers: [],
                    bootstrap: []
                },] }
    ];
    return ThfNotificationModule;
}());
export { ThfNotificationModule };
