// @totvs/thf-ui/esm5/lib/components/thf-menu/thf-menu-filter/thf-menu-filter.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ElementRef, EventEmitter, Input, Output, ViewChild } from '@angular/core';
/**
 * \@docsPrivate
 *
 * \@description
 *
 * Componente que apresenta o campo de pesquisa no thf-menu.
 */
var ThfMenuFilterComponent = /** @class */ (function () {
    function ThfMenuFilterComponent() {
        // Variável necessária para o thf-clean identificar que deve ser criado.
        this.clean = true;
        this.filter = new EventEmitter();
    }
    /**
     * @param {?} search
     * @return {?}
     */
    ThfMenuFilterComponent.prototype.filterItems = /**
     * @param {?} search
     * @return {?}
     */
    function (search) {
        this.filter.emit(search);
    };
    ThfMenuFilterComponent.decorators = [
        { type: Component, args: [{
                    selector: 'thf-menu-filter',
                    template: "<div class=\"thf-menu-filter-container\">\n\n  <input #inputFilter\n    type=\"text\"\n    class=\"thf-menu-filter\"\n    placeholder=\"Pesquisar\"\n    (keyup)=\"filterItems(inputFilter.value)\">\n\n  <div class=\"thf-menu-filter-search-icon-container\">\n    <span *ngIf=\"!loading\" class=\"thf-icon thf-menu-filter-icon thf-icon-search\"></span>\n    <span *ngIf=\"loading\" class=\"thf-icon thf-menu-filter-icon thf-loading-icon\"></span>\n  </div>\n  <div class=\"thf-menu-filter-close-icon-container\">\n    <thf-clean [t-element-ref]=\"inputFilterElement\" (t-change-event)=\"filterItems(inputFilter.value)\"></thf-clean>\n  </div>\n\n</div>\n"
                }] }
    ];
    ThfMenuFilterComponent.propDecorators = {
        loading: [{ type: Input, args: ['t-loading',] }],
        inputFilterElement: [{ type: ViewChild, args: ['inputFilter', { read: ElementRef, static: true },] }],
        filter: [{ type: Output, args: ['t-filter',] }]
    };
    return ThfMenuFilterComponent;
}());
export { ThfMenuFilterComponent };
if (false) {
    /** @type {?} */
    ThfMenuFilterComponent.prototype.clean;
    /** @type {?} */
    ThfMenuFilterComponent.prototype.loading;
    /** @type {?} */
    ThfMenuFilterComponent.prototype.inputFilterElement;
    /** @type {?} */
    ThfMenuFilterComponent.prototype.filter;
}
