// @totvs/thf-templates/esm2015/lib/components/thf-page-job-scheduler/thf-page-job-scheduler-internal.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
export class ThfPageJobSchedulerInternal {
    constructor() {
        this.periodicity = 'single';
        this.firstExecution = new Date();
        this.firstExecutionHour = this.getCurrentHour(this.firstExecution);
        this.recurrent = true;
    }
    /**
     * @private
     * @param {?} date
     * @return {?}
     */
    getCurrentHour(date) {
        /** @type {?} */
        const hours = date.getHours() < 10 ? '0' + date.getHours() : date.getHours();
        /** @type {?} */
        const minutes = date.getMinutes() < 10 ? '0' + date.getMinutes() : date.getMinutes();
        return `${hours}:${minutes}`;
    }
}
if (false) {
    /** @type {?} */
    ThfPageJobSchedulerInternal.prototype.periodicity;
    /** @type {?} */
    ThfPageJobSchedulerInternal.prototype.firstExecution;
    /** @type {?} */
    ThfPageJobSchedulerInternal.prototype.firstExecutionHour;
    /** @type {?} */
    ThfPageJobSchedulerInternal.prototype.recurrent;
}
