// @totvs/thf-ui/esm5/lib/components/thf-navbar/thf-navbar-base.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Input } from '@angular/core';
import { browserLanguage, convertToBoolean, thfLocaleDefault } from '../../utils/util';
/** @type {?} */
export var thfNavbarLiteralsDefault = {
    en: (/** @type {?} */ ({
        navbarLinks: 'Navbar links'
    })),
    es: (/** @type {?} */ ({
        navbarLinks: 'Navbar links'
    })),
    pt: (/** @type {?} */ ({
        navbarLinks: 'Navbar links'
    }))
};
/**
 * \@description
 *
 * O componente `thf-navbar` é um cabeçalho fixo que permite apresentar uma lista de links para facilitar a navegação pelas
 * páginas da aplicação. Também possui ícones com ações.
 */
var ThfNavbarBaseComponent = /** @class */ (function () {
    function ThfNavbarBaseComponent() {
        this._iconActions = [];
        this._items = [];
        this._shadow = false;
    }
    Object.defineProperty(ThfNavbarBaseComponent.prototype, "iconActions", {
        get: /**
         * @return {?}
         */
        function () {
            return this._iconActions;
        },
        /**
         * @optional
         *
         * @description
         *
         * Define uma lista de ações apresentadas em ícones no lado direito do `thf-navbar`.
         */
        set: /**
         * \@optional
         *
         * \@description
         *
         * Define uma lista de ações apresentadas em ícones no lado direito do `thf-navbar`.
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._iconActions = value || [];
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ThfNavbarBaseComponent.prototype, "items", {
        get: /**
         * @return {?}
         */
        function () {
            return this._items;
        },
        /**
         * @optional
         *
         * @description
         *
         * Define uma lista de items do `thf-navbar`.
         */
        set: /**
         * \@optional
         *
         * \@description
         *
         * Define uma lista de items do `thf-navbar`.
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._items = value || [];
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ThfNavbarBaseComponent.prototype, "literals", {
        get: /**
         * @return {?}
         */
        function () {
            return this._literals || thfNavbarLiteralsDefault[browserLanguage()];
        },
        /**
         * @optional
         *
         * @description
         *
         * Objeto com a literal usada na propriedade `t-literals`.
         *
         * Para customizar a literal, basta declarar um objeto do tipo `ThfNavbarLiterals` conforme exemplo abaixo:
         *
         * ```
         *  const customLiterals: ThfNavbarLiterals = {
         *    navbarLinks: 'Itens de navegação'
         *  };
         * ```
         *
         * E para carregar as literais customizadas, basta apenas passar o objeto para o componente.
         *
         * ```
         * <thf-navbar
         *   [t-literals]="customLiterals">
         * </thf-navbar>
         * ```
         *
         *  > O objeto padrão de literais será traduzido de acordo com o idioma do *browser* (pt, en, es).
         */
        set: /**
         * \@optional
         *
         * \@description
         *
         * Objeto com a literal usada na propriedade `t-literals`.
         *
         * Para customizar a literal, basta declarar um objeto do tipo `ThfNavbarLiterals` conforme exemplo abaixo:
         *
         * ```
         *  const customLiterals: ThfNavbarLiterals = {
         *    navbarLinks: 'Itens de navegação'
         *  };
         * ```
         *
         * E para carregar as literais customizadas, basta apenas passar o objeto para o componente.
         *
         * ```
         * <thf-navbar
         *   [t-literals]="customLiterals">
         * </thf-navbar>
         * ```
         *
         *  > O objeto padrão de literais será traduzido de acordo com o idioma do *browser* (pt, en, es).
         * @param {?} value
         * @return {?}
         */
        function (value) {
            if (value instanceof Object && !(value instanceof Array)) {
                this._literals = tslib_1.__assign({}, thfNavbarLiteralsDefault[thfLocaleDefault], thfNavbarLiteralsDefault[browserLanguage()], value);
            }
            else {
                this._literals = thfNavbarLiteralsDefault[browserLanguage()];
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ThfNavbarBaseComponent.prototype, "shadow", {
        get: /**
         * @return {?}
         */
        function () {
            return this._shadow;
        },
        /**
         * @optional
         *
         * @description
         *
         * Aplica uma sombra na parte inferior do `thf-navbar`.
         *
         * @default `false`
         */
        set: /**
         * \@optional
         *
         * \@description
         *
         * Aplica uma sombra na parte inferior do `thf-navbar`.
         *
         * \@default `false`
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._shadow = convertToBoolean(value);
        },
        enumerable: true,
        configurable: true
    });
    ThfNavbarBaseComponent.propDecorators = {
        iconActions: [{ type: Input, args: ['t-icon-actions',] }],
        items: [{ type: Input, args: ['t-items',] }],
        literals: [{ type: Input, args: ['t-literals',] }],
        logo: [{ type: Input, args: ['t-logo',] }],
        menu: [{ type: Input, args: ['t-menu',] }],
        shadow: [{ type: Input, args: ['t-shadow',] }]
    };
    return ThfNavbarBaseComponent;
}());
export { ThfNavbarBaseComponent };
if (false) {
    /**
     * @type {?}
     * @private
     */
    ThfNavbarBaseComponent.prototype._iconActions;
    /**
     * @type {?}
     * @private
     */
    ThfNavbarBaseComponent.prototype._items;
    /**
     * @type {?}
     * @private
     */
    ThfNavbarBaseComponent.prototype._literals;
    /**
     * @type {?}
     * @private
     */
    ThfNavbarBaseComponent.prototype._shadow;
    /**
     * \@optional
     *
     * \@description
     *
     * Define a logo apresentada `thf-navbar`.
     * @type {?}
     */
    ThfNavbarBaseComponent.prototype.logo;
    /**
     * \@optional
     *
     * \@description
     *
     * Caso já possua um menu na aplicação o mesmo deve ser repassado para essa propriedade para que quando entre em modo
     * responsivo os items do `thf-navbar` possam ser adicionados no primeiro item do menu definido.
     *
     * Exemplo:
     *
     * ```
     * <thf-navbar
     *  [t-items]="items"
     *  [t-icon-actions]="iconActions"
     *  [t-menu]="userMenu">
     * </thf-navbar>
     * <div class="thf-wrapper">
     *  <thf-menu #userMenu
     *   [t-menus]="[{ label: 'Item 1', link: '/' }]">
     *  </thf-menu>
     * </div>
     * ```
     * @type {?}
     */
    ThfNavbarBaseComponent.prototype.menu;
}
