// @totvs/thf-ui/esm2015/lib/components/thf-loading/thf-loading-overlay/thf-loading-overlay-base.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Input } from '@angular/core';
import { convertToBoolean } from './../../../utils/util';
/** @type {?} */
export const thfLoadingOverlayLiteralsDefault = {
    en: (/** @type {?} */ ({
        loading: 'Loading',
    })),
    es: (/** @type {?} */ ({
        loading: 'Cargando',
    })),
    pt: (/** @type {?} */ ({
        loading: 'Carregando',
    }))
};
/**
 *
 * \@description
 *
 * Este componente mostra ao usuário uma imagem de _loading_ e bloqueia a página inteira ou o container escolhido,
 * enquanto aguarda a resposta de alguma requisição.
 */
export class ThfLoadingOverlayBaseComponent {
    /**
     * @param {?} languageService
     */
    constructor(languageService) {
        this.languageService = languageService;
        this._screenLock = false;
        if (!this.text) {
            this.text = this.getTextDefault();
        }
    }
    /**
     * \@optional
     *
     * \@description
     *
     * Define se o _overlay_ será aplicado a um container ou a página inteira.
     *
     * \@default `false`
     * @param {?} screenLock
     * @return {?}
     */
    set screenLock(screenLock) {
        this._screenLock = convertToBoolean(screenLock);
    }
    /**
     * @return {?}
     */
    get screenLock() {
        return this._screenLock;
    }
    /**
     * \@optional
     *
     * \@description
     *
     * Texto a ser exibido no componente.
     *
     * > O valor padrão será traduzido acordo com o idioma configurado no [**ThfI18n**](/documentation/thf-i18n) ou navegador.
     *
     * \@default `Carregando`
     * @param {?} value
     * @return {?}
     */
    set text(value) {
        this._text = value || this.getTextDefault();
    }
    /**
     * @return {?}
     */
    get text() {
        return this._text;
    }
    /**
     * @private
     * @return {?}
     */
    getTextDefault() {
        /** @type {?} */
        const language = this.languageService.getShortLanguage();
        return thfLoadingOverlayLiteralsDefault[language].loading;
    }
}
ThfLoadingOverlayBaseComponent.propDecorators = {
    screenLock: [{ type: Input, args: ['t-screen-lock',] }],
    text: [{ type: Input, args: ['t-text',] }]
};
if (false) {
    /**
     * @type {?}
     * @private
     */
    ThfLoadingOverlayBaseComponent.prototype._screenLock;
    /**
     * @type {?}
     * @private
     */
    ThfLoadingOverlayBaseComponent.prototype._text;
    /**
     * @type {?}
     * @private
     */
    ThfLoadingOverlayBaseComponent.prototype.languageService;
}
