// @totvs/thf-ui/esm2015/lib/components/thf-toolbar/thf-toolbar.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ViewContainerRef } from '@angular/core';
import { Title } from '@angular/platform-browser';
import { ThfToolbarBaseComponent } from './thf-toolbar-base.component';
/**
 * \@docsExtends ThfToolbarBaseComponent
 *
 * \@example
 *
 * <example name="thf-toolbar-basic" title="Totvs Toolbar Basic">
 *  <file name="sample-thf-toolbar-basic/sample-thf-toolbar-basic.component.html"> </file>
 *  <file name="sample-thf-toolbar-basic/sample-thf-toolbar-basic.component.ts"> </file>
 * </example>
 *
 * <example name="thf-toolbar-labs" title="Totvs Toolbar Labs">
 *  <file name="sample-thf-toolbar-labs/sample-thf-toolbar-labs.component.html"> </file>
 *  <file name="sample-thf-toolbar-labs/sample-thf-toolbar-labs.component.ts"> </file>
 * </example>
 *
 * <example name="thf-toolbar-logged" title="Totvs Toolbar - Logged">
 *  <file name="sample-thf-toolbar-logged/sample-thf-toolbar-logged.component.html"> </file>
 *  <file name="sample-thf-toolbar-logged/sample-thf-toolbar-logged.component.ts"> </file>
 * </example>
 */
export class ThfToolbarComponent extends ThfToolbarBaseComponent {
    /**
     * @param {?} titleService
     * @param {?} viewRef
     */
    constructor(titleService, viewRef) {
        super();
        this.titleService = titleService;
        this.parentRef = viewRef['_view']['component'];
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this.setTitle(this.title);
    }
    /**
     * @private
     * @param {?} newTitle
     * @return {?}
     */
    setTitle(newTitle) {
        this.titleService.setTitle(newTitle);
    }
}
ThfToolbarComponent.decorators = [
    { type: Component, args: [{
                selector: 'thf-toolbar',
                template: "<div class=\"thf-toolbar\">\n  <div class=\"thf-toolbar-title\"> {{title}} </div>\n\n  <div class=\"thf-toolbar-group-icon\">\n    <thf-toolbar-actions *ngIf=\"actions && actions.length\"\n      [t-actions]=\"actions\"\n      [t-actions-icon]=\"actionsIcon\">\n    </thf-toolbar-actions>\n\n    <thf-toolbar-notification *ngIf=\"showNotification && notificationActions\"\n      [t-notification-actions]=\"notificationActions\"\n      [t-notification-number]=\"notificationNumber\">\n    </thf-toolbar-notification>\n\n    <thf-toolbar-profile *ngIf=\"isShowProfile\"\n      [t-profile]=\"profile || { avatar: userSrc, title: userName }\"\n      [t-profile-actions]=\"profileActions || userActions\">\n    </thf-toolbar-profile>\n  </div>\n</div>\n"
            }] }
];
/** @nocollapse */
ThfToolbarComponent.ctorParameters = () => [
    { type: Title },
    { type: ViewContainerRef }
];
if (false) {
    /** @type {?} */
    ThfToolbarComponent.prototype.parentRef;
    /**
     * @type {?}
     * @private
     */
    ThfToolbarComponent.prototype.titleService;
}
