// @totvs/thf-ui/esm2015/lib/components/thf-navbar/thf-navbar-actions/thf-navbar-action/thf-navbar-action.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, ViewContainerRef } from '@angular/core';
import { Router } from '@angular/router';
import { callFunction, isExternalLink, openExternalLink } from '../../../../utils/util';
export class ThfNavbarActionComponent {
    /**
     * @param {?} viewContainerRef
     * @param {?} router
     */
    constructor(viewContainerRef, router) {
        this.router = router;
        this.parentRef = viewContainerRef['_view']['component'];
    }
    /**
     * @return {?}
     */
    onActionClick() {
        if (this.action) {
            return callFunction(this.action, this.parentRef, this.param || this);
        }
        if (this.link) {
            return this.openUrl(this.link);
        }
    }
    /**
     * @private
     * @param {?} url
     * @return {?}
     */
    openUrl(url) {
        if (isExternalLink(url)) {
            return openExternalLink(url);
        }
        if (url) {
            return this.router.navigate([url]);
        }
    }
}
ThfNavbarActionComponent.decorators = [
    { type: Component, args: [{
                selector: 'thf-navbar-action',
                template: "<div tabindex=\"0\"\n  class=\"thf-navbar-action-content thf-clickable\"\n  (click)=\"onActionClick()\">\n\n  <span\n  class=\"thf-icon {{ icon }}\"\n  [t-tooltip]=\"tooltip\">\n  </span>\n\n</div>\n"
            }] }
];
/** @nocollapse */
ThfNavbarActionComponent.ctorParameters = () => [
    { type: ViewContainerRef },
    { type: Router }
];
ThfNavbarActionComponent.propDecorators = {
    action: [{ type: Input, args: ['t-action',] }],
    icon: [{ type: Input, args: ['t-icon',] }],
    label: [{ type: Input, args: ['t-label',] }],
    link: [{ type: Input, args: ['t-link',] }],
    tooltip: [{ type: Input, args: ['t-tooltip-text',] }]
};
if (false) {
    /**
     * @type {?}
     * @private
     */
    ThfNavbarActionComponent.prototype.param;
    /**
     * @type {?}
     * @private
     */
    ThfNavbarActionComponent.prototype.parentRef;
    /** @type {?} */
    ThfNavbarActionComponent.prototype.action;
    /** @type {?} */
    ThfNavbarActionComponent.prototype.icon;
    /** @type {?} */
    ThfNavbarActionComponent.prototype.label;
    /** @type {?} */
    ThfNavbarActionComponent.prototype.link;
    /** @type {?} */
    ThfNavbarActionComponent.prototype.tooltip;
    /**
     * @type {?}
     * @private
     */
    ThfNavbarActionComponent.prototype.router;
}
