// @totvs/thf-templates/esm5/lib/components/thf-page-dynamic-search/thf-page-dynamic-search-base.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { EventEmitter, Input, Output } from '@angular/core';
import { browserLanguage, thfLocaleDefault } from '../../utils/util';
/** @type {?} */
export var thfPageDynamicSearchLiteralsDefault = {
    en: {
        disclaimerGroupTitle: 'Displaying results filtered by:',
        filterSettingsPlaceholder: 'Search',
        quickSearchLabel: 'Quick search:'
    },
    es: {
        disclaimerGroupTitle: 'Presentando resultados filtrados por:',
        filterSettingsPlaceholder: 'Buscar',
        quickSearchLabel: 'Búsqueda rápida:'
    },
    pt: {
        disclaimerGroupTitle: 'Apresentando resultados filtrados por:',
        filterSettingsPlaceholder: 'Pesquisar',
        quickSearchLabel: 'Pesquisa rápida:'
    }
};
/**
 * \@description
 *
 * Componente com as ações de pesquisa já definidas, bastando que o desenvolvedor implemente apenas a chamada para as APIs
 * e exiba as informações.
 */
var ThfPageDynamicSearchBaseComponent = /** @class */ (function () {
    function ThfPageDynamicSearchBaseComponent() {
        this._filters = [];
        this.literals = tslib_1.__assign({}, thfPageDynamicSearchLiteralsDefault[thfLocaleDefault], thfPageDynamicSearchLiteralsDefault[browserLanguage()]);
        /**
         * Nesta propriedade deve ser definido um array de objetos que implementam a interface `ThfPageAction`.
         */
        this.actions = [];
        /**
         * Objeto com propriedades do breadcrumb.
         */
        this.breadcrumb = { items: [] };
        /**
         * \@description
         *
         * Evento disparado ao executar a pesquisa avançada, o mesmo irá repassar um objeto com os valores preenchidos no modal de pesquisa.
         *
         * > Campos não preenchidos não irão aparecer no objeto passado por parâmetro.
         */
        this.advancedSearch = new EventEmitter();
        /**
         * Evento disparado ao remover um ou todos os disclaimers pelo usuário.
         */
        this.changeDisclaimers = new EventEmitter();
        /**
         * Evento disparado ao realizar uma busca pelo campo de pesquisa rápida, o mesmo será chamado repassando o valor digitado.
         */
        this.quickSearch = new EventEmitter();
    }
    Object.defineProperty(ThfPageDynamicSearchBaseComponent.prototype, "filters", {
        get: /**
         * @return {?}
         */
        function () {
            return this._filters;
        },
        /**
         * @optional
         *
         * @description
         *
         * Lista dos campos usados na busca avançada. Caso o mesmo não seja passado a busca avançada não será exibida.
         */
        set: /**
         * \@optional
         *
         * \@description
         *
         * Lista dos campos usados na busca avançada. Caso o mesmo não seja passado a busca avançada não será exibida.
         * @param {?} filters
         * @return {?}
         */
        function (filters) {
            this._filters = Array.isArray(filters) ? tslib_1.__spread(filters) : [];
        },
        enumerable: true,
        configurable: true
    });
    ThfPageDynamicSearchBaseComponent.propDecorators = {
        actions: [{ type: Input, args: ['t-actions',] }],
        breadcrumb: [{ type: Input, args: ['t-breadcrumb',] }],
        filters: [{ type: Input, args: ['t-filters',] }],
        title: [{ type: Input, args: ['t-title',] }],
        advancedSearch: [{ type: Output, args: ['t-advanced-search',] }],
        changeDisclaimers: [{ type: Output, args: ['t-change-disclaimers',] }],
        quickSearch: [{ type: Output, args: ['t-quick-search',] }]
    };
    return ThfPageDynamicSearchBaseComponent;
}());
export { ThfPageDynamicSearchBaseComponent };
if (false) {
    /**
     * @type {?}
     * @private
     */
    ThfPageDynamicSearchBaseComponent.prototype._filters;
    /** @type {?} */
    ThfPageDynamicSearchBaseComponent.prototype.literals;
    /**
     * Nesta propriedade deve ser definido um array de objetos que implementam a interface `ThfPageAction`.
     * @type {?}
     */
    ThfPageDynamicSearchBaseComponent.prototype.actions;
    /**
     * Objeto com propriedades do breadcrumb.
     * @type {?}
     */
    ThfPageDynamicSearchBaseComponent.prototype.breadcrumb;
    /**
     * Título da página.
     * @type {?}
     */
    ThfPageDynamicSearchBaseComponent.prototype.title;
    /**
     * \@description
     *
     * Evento disparado ao executar a pesquisa avançada, o mesmo irá repassar um objeto com os valores preenchidos no modal de pesquisa.
     *
     * > Campos não preenchidos não irão aparecer no objeto passado por parâmetro.
     * @type {?}
     */
    ThfPageDynamicSearchBaseComponent.prototype.advancedSearch;
    /**
     * Evento disparado ao remover um ou todos os disclaimers pelo usuário.
     * @type {?}
     */
    ThfPageDynamicSearchBaseComponent.prototype.changeDisclaimers;
    /**
     * Evento disparado ao realizar uma busca pelo campo de pesquisa rápida, o mesmo será chamado repassando o valor digitado.
     * @type {?}
     */
    ThfPageDynamicSearchBaseComponent.prototype.quickSearch;
}
