// @totvs/thf-ui/esm2015/lib/components/thf-toolbar/thf-toolbar-actions/thf-toolbar-actions.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input } from '@angular/core';
import { isTypeof } from '../../../utils/util';
import { ThfControlPositionService } from '../../../services/thf-control-position/thf-control-position.service';
/** @type {?} */
const thfToolbarActionsIconDefault = 'thf-icon-more';
/**
 * \@docsPrivate
 *
 * \@usedBy ThfToolbarComponent
 *
 * \@description
 *
 * O componente `thf-toolbar-actions` tem como objetivo receber uma lista de ações e um ícone que podem ser personalizados.
 */
export class ThfToolbarActionsComponent {
    constructor() {
        this._actionsIcon = thfToolbarActionsIconDefault;
    }
    /**
     * Define o ícone das ações.
     * @param {?} icon
     * @return {?}
     */
    set actionsIcon(icon) {
        this._actionsIcon = isTypeof(icon, 'string') ? icon : thfToolbarActionsIconDefault;
    }
    /**
     * @return {?}
     */
    get actionsIcon() {
        return this._actionsIcon;
    }
}
ThfToolbarActionsComponent.decorators = [
    { type: Component, args: [{
                selector: 'thf-toolbar-actions',
                template: "<div class=\"thf-toolbar-actions thf-clickable\" (click)=\"popup.toggle()\">\n  <span #toolbarActions class=\"thf-icon {{ actionsIcon }} thf-toolbar-icon\"></span>\n</div>\n\n<thf-popup #popup\n  [t-actions]=\"actions\"\n  [t-target]=\"toolbarActions\">\n</thf-popup>\n",
                providers: [ThfControlPositionService]
            }] }
];
ThfToolbarActionsComponent.propDecorators = {
    actions: [{ type: Input, args: ['t-actions',] }],
    actionsIcon: [{ type: Input, args: ['t-actions-icon',] }]
};
if (false) {
    /**
     * @type {?}
     * @private
     */
    ThfToolbarActionsComponent.prototype._actionsIcon;
    /**
     * Define uma lista de ações.
     * @type {?}
     */
    ThfToolbarActionsComponent.prototype.actions;
}
