// @totvs/thf-ui/esm5/lib/components/thf-field/thf-upload/thf-upload-file.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ThfUploadStatus } from './thf-upload-status.enum';
// Classe responsável pelo arquivo do THF Upload
var 
// Classe responsável pelo arquivo do THF Upload
ThfUploadFile = /** @class */ (function () {
    function ThfUploadFile(file) {
        if (file) {
            this.name = file.name;
            this.extension = this.getExtension(file.name);
            this.size = file.size;
            this.rawFile = file;
            this.uid = this.generateUUID();
            this.status = ThfUploadStatus.None;
        }
    }
    /**
     * @private
     * @param {?} value
     * @return {?}
     */
    ThfUploadFile.prototype.getExtension = /**
     * @private
     * @param {?} value
     * @return {?}
     */
    function (value) {
        if (value) {
            /** @type {?} */
            var extension = value.substr(value.lastIndexOf('.'));
            return extension.toLowerCase();
        }
        return '';
    };
    /**
     * @private
     * @return {?}
     */
    ThfUploadFile.prototype.generateUUID = /**
     * @private
     * @return {?}
     */
    function () {
        /**
         * @return {?}
         */
        function s4() {
            return Math.floor((1 + Math.random()) * 0x10000)
                .toString(16)
                .substring(1);
        }
        return s4() + s4() + '-' + s4() + '-' + s4() + '-' +
            s4() + '-' + s4() + s4() + s4();
    };
    return ThfUploadFile;
}());
// Classe responsável pelo arquivo do THF Upload
export { ThfUploadFile };
if (false) {
    /** @type {?} */
    ThfUploadFile.prototype.name;
    /** @type {?} */
    ThfUploadFile.prototype.rawFile;
    /** @type {?} */
    ThfUploadFile.prototype.status;
    /** @type {?} */
    ThfUploadFile.prototype.uid;
    /** @type {?} */
    ThfUploadFile.prototype.extension;
    /** @type {?} */
    ThfUploadFile.prototype.size;
}
