// @totvs/thf-ui/esm2015/lib/components/thf-field/thf-field-container/thf-field-container.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input } from '@angular/core';
import { convertToBoolean, getShortBrowserLanguage } from '../../../utils/util';
/**
 * \@docsPrivate
 *
 * Componente de uso interno, responsável por atribuir uma label para o campo
 */
export class ThfFieldContainerComponent {
    constructor() {
        this._optional = false;
    }
    /**
     * Indica se o campo será opcional.
     * @param {?} value
     * @return {?}
     */
    set optional(value) {
        this._optional = convertToBoolean(value);
    }
    /**
     * @return {?}
     */
    get optional() {
        return this._optional;
    }
    /**
     * @return {?}
     */
    getOptionalText() {
        /** @type {?} */
        const browserLanguage = getShortBrowserLanguage();
        /** @type {?} */
        const optional = {
            pt: '(Opcional)',
            en: '(Optional)',
            es: '(Opcional)'
        };
        return optional[browserLanguage];
    }
}
ThfFieldContainerComponent.decorators = [
    { type: Component, args: [{
                selector: 'thf-field-container',
                template: "<div class=\"thf-field-container\">\n  <div class=\"thf-field-container-title\">\n    <span class=\"thf-field-title\">{{ label }}</span>\n    <span class=\"thf-field-optional\" *ngIf=\"optional && (label || help)\">{{ getOptionalText() }}</span>\n  </div>\n\n  <div class=\"thf-field-help\" *ngIf=\"help\">{{ help }}</div>\n\n  <ng-content></ng-content>\n</div>\n"
            }] }
];
ThfFieldContainerComponent.propDecorators = {
    label: [{ type: Input, args: ['t-label',] }],
    help: [{ type: Input, args: ['t-help',] }],
    optional: [{ type: Input, args: ['t-optional',] }]
};
if (false) {
    /**
     * @type {?}
     * @private
     */
    ThfFieldContainerComponent.prototype._optional;
    /**
     * Label do campo.
     * @type {?}
     */
    ThfFieldContainerComponent.prototype.label;
    /**
     * Texto de apoio do campo.
     * @type {?}
     */
    ThfFieldContainerComponent.prototype.help;
}
