// @totvs/thf-ui/esm2015/lib/components/thf-field/thf-switch/thf-switch.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ChangeDetectorRef, Component, forwardRef } from '@angular/core';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
import { ThfSwitchBaseComponent } from './thf-switch-base.component';
import { ThfSwitchLabelPosition } from './thf-switch-label-position.enum';
/**
 * \@docsExtends ThfSwitchBaseComponent
 *
 * \@example
 *
 * <example name="thf-switch-basic" title="Totvs Switch Basic">
 *   <file name="sample-thf-switch-basic/sample-thf-switch-basic.component.html"> </file>
 *   <file name="sample-thf-switch-basic/sample-thf-switch-basic.component.ts"> </file>
 * </example>
 *
 * <example name="thf-switch-labs" title="Totvs Switch Labs">
 *   <file name="sample-thf-switch-labs/sample-thf-switch-labs.component.html"> </file>
 *   <file name="sample-thf-switch-labs/sample-thf-switch-labs.component.ts"> </file>
 *   <file name="sample-thf-switch-labs/sample-thf-switch-labs.component.e2e-spec.ts"> </file>
 *   <file name="sample-thf-switch-labs/sample-thf-switch-labs.component.po.ts"> </file>
 * </example>
 *
 * <example name="thf-switch-order" title="Totvs Switch - Order Summary">
 *   <file name="sample-thf-switch-order/sample-thf-switch-order.component.html"> </file>
 *   <file name="sample-thf-switch-order/sample-thf-switch-order.component.ts"> </file>
 * </example>
 *
 * <example name="thf-switch-order-reactive-form" title="Totvs Switch - Order Summary Reactive Form">
 *   <file name="sample-thf-switch-order-reactive-form/sample-thf-switch-order-reactive-form.component.html"> </file>
 *   <file name="sample-thf-switch-order-reactive-form/sample-thf-switch-order-reactive-form.component.ts"> </file>
 * </example>
 */
export class ThfSwitchComponent extends ThfSwitchBaseComponent {
    /**
     * @param {?} changeDetector
     */
    constructor(changeDetector) {
        super();
        this.changeDetector = changeDetector;
    }
    /**
     * @return {?}
     */
    ngAfterViewChecked() {
        this.changeDetector.detectChanges();
    }
    /**
     * @return {?}
     */
    getLabelPosition() {
        switch (this.labelPosition) {
            case ThfSwitchLabelPosition.Left: return 'left';
            case ThfSwitchLabelPosition.Right: return 'right';
            default: return 'right';
        }
    }
    /**
     * @return {?}
     */
    getSwitchPosition() {
        switch (this.labelPosition) {
            case ThfSwitchLabelPosition.Left: return 'right';
            case ThfSwitchLabelPosition.Right: return 'left';
            default: return 'left';
        }
    }
    /**
     * @param {?} event
     * @return {?}
     */
    onKeyDown(event) {
        if (event.which === 32 || event.keyCode === 32) {
            event.preventDefault();
            this.eventClick();
        }
    }
}
ThfSwitchComponent.decorators = [
    { type: Component, args: [{
                selector: 'thf-switch',
                template: "<thf-field-container\n  [t-label]=\"label\"\n  [t-help]=\"help\">\n\n  <div class=\"thf-field-container-content thf-switch-content\" [attr.name]=\"name\">\n    <div class=\"thf-switch-container thf-clickable\"\n      [class.thf-switch-container-disabled]=\"disabled\"\n      [class.thf-switch-container-off]=\"switchValue === false && !disabled\"\n      [class.thf-switch-container-on]=\"switchValue === true && !disabled\"\n      [class.thf-switch-container-right]=\"getSwitchPosition() === 'right'\"\n      [tabindex]=\"disabled ? -1 : 0\"\n      (click)=\"eventClick()\"\n      (keydown)=\"onKeyDown($event)\">\n\n      <div class=\"thf-switch-button\"\n        [class.thf-switch-button-on]=\"switchValue === true\"\n        [class.thf-switch-button-off]=\"switchValue === false\"\n        [class.thf-switch-button-disabled]=\"disabled\">\n\n        <span class=\"thf-icon thf-icon-{{ switchValue === true ? 'ok' : 'close' }}\"></span>\n      </div>\n    </div>\n\n    <span class=\"thf-switch-label thf-clickable thf-switch-label-{{ getLabelPosition() }}\"\n      (click)=\"eventClick()\">\n        {{ switchValue === true ? labelOn : labelOff }}\n    </span>\n  </div>\n\n  <thf-field-container-bottom></thf-field-container-bottom>\n</thf-field-container>\n",
                providers: [
                    {
                        provide: NG_VALUE_ACCESSOR,
                        useExisting: forwardRef((/**
                         * @return {?}
                         */
                        () => ThfSwitchComponent)),
                        multi: true,
                    }
                ]
            }] }
];
/** @nocollapse */
ThfSwitchComponent.ctorParameters = () => [
    { type: ChangeDetectorRef }
];
if (false) {
    /**
     * @type {?}
     * @private
     */
    ThfSwitchComponent.prototype.changeDetector;
}
