// @totvs/thf-ui/esm2015/lib/components/thf-breadcrumb/thf-breadcrumb-favorite/thf-breadcrumb-favorite.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
/**
 * \@docsPrivate
 *
 * \@description
 *
 * Serviço utilizado para favoritar/desfavoritar uma url no componente thf-breadcrumb.
 * A API deve retornar um objeto no formato { isFavorite: booleano, url: string }. Em ambos os métodos, GET e POST este
 * objeto deve ser retornado.
 * Ao alterar o status de favorito, a API deve retornar o objeto acima, com o novo status atualizado.
 * No método POST, existe a possibilidade de ser enviado parâmetros junto com a requisição. Neste caso, o objeto enviado
 * para a API terá o formato: { isFavorite: booleano, url: string, params: {} }, onde params é o objeto que pode ser enviado
 * junto com a requisição.
 *
 */
export class ThfBreadcrumbFavoriteService {
    /**
     * @param {?} http
     */
    constructor(http) {
        this.http = http;
    }
    /**
     * @return {?}
     */
    get url() { return this._url; }
    /**
     * @return {?}
     */
    get bodyParams() { return this._bodyParams; }
    /**
     * @param {?} url
     * @param {?} params
     * @param {?} item
     * @return {?}
     */
    configService(url, params, item) {
        this._url = url;
        this._bodyParams = {
            isFavorite: false,
            url: item.link,
            params: params
        };
    }
    /**
     * @param {?} status
     * @return {?}
     */
    sendStatusFavorite(status) {
        this.bodyParams['isFavorite'] = status;
        return this.http.post(this.url, this.bodyParams);
    }
    /**
     * @return {?}
     */
    getFavorite() {
        return this.http.get(this.url, { responseType: 'json', params: { url: this.bodyParams['url'] } });
    }
}
ThfBreadcrumbFavoriteService.decorators = [
    { type: Injectable }
];
/** @nocollapse */
ThfBreadcrumbFavoriteService.ctorParameters = () => [
    { type: HttpClient }
];
if (false) {
    /**
     * @type {?}
     * @private
     */
    ThfBreadcrumbFavoriteService.prototype._url;
    /**
     * @type {?}
     * @private
     */
    ThfBreadcrumbFavoriteService.prototype._bodyParams;
    /**
     * @type {?}
     * @private
     */
    ThfBreadcrumbFavoriteService.prototype.http;
}
