// @totvs/thf-ui/esm5/lib/services/thf-dialog/thf-dialog.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { ThfModalModule } from '../../components/thf-modal/thf-modal.module';
import { ThfDialogComponent } from './thf-dialog.component';
import { ThfDialogService } from './thf-dialog.service';
var ThfDialogModule = /** @class */ (function () {
    function ThfDialogModule() {
    }
    ThfDialogModule.decorators = [
        { type: NgModule, args: [{
                    declarations: [
                        ThfDialogComponent
                    ],
                    imports: [
                        ThfModalModule
                    ],
                    exports: [
                        ThfDialogComponent
                    ],
                    entryComponents: [
                        ThfDialogComponent
                    ],
                    providers: [
                        ThfDialogService
                    ]
                },] }
    ];
    return ThfDialogModule;
}());
export { ThfDialogModule };
