// @totvs/thf-ui/esm2015/lib/components/thf-page/thf-page-default/thf-page-default-base.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Input, ViewChild } from '@angular/core';
import { ThfPageContentComponent } from '../thf-page-content/thf-page-content.component';
/**
 * \@description
 *
 * O componente `thf-page-default` é utilizado como o container principal para as telas sem um template definido.
 * @abstract
 */
export class ThfPageDefaultBaseComponent {
    constructor() {
        this._actions = [];
    }
    /**
     * \@optional
     *
     * \@description
     *
     * Nesta propriedade deve ser definido um array de objetos que implementam a interface `ThfPageAction`.
     * @param {?} actions
     * @return {?}
     */
    set actions(actions) {
        this._actions = Array.isArray(actions) ? actions : [];
        this.setDropdownActions();
    }
    /**
     * @return {?}
     */
    get actions() {
        return this._actions;
    }
    /**
     * Título da página.
     * @param {?} title
     * @return {?}
     */
    set title(title) {
        this._title = title;
        this.thfPageContent.recalculateHeaderSize();
    }
    /**
     * @return {?}
     */
    get title() {
        return this._title;
    }
}
ThfPageDefaultBaseComponent.propDecorators = {
    thfPageContent: [{ type: ViewChild, args: [ThfPageContentComponent, { static: true },] }],
    actions: [{ type: Input, args: ['t-actions',] }],
    breadcrumb: [{ type: Input, args: ['t-breadcrumb',] }],
    title: [{ type: Input, args: ['t-title',] }]
};
if (false) {
    /**
     * @type {?}
     * @private
     */
    ThfPageDefaultBaseComponent.prototype._actions;
    /**
     * @type {?}
     * @private
     */
    ThfPageDefaultBaseComponent.prototype._title;
    /** @type {?} */
    ThfPageDefaultBaseComponent.prototype.thfPageContent;
    /**
     * Objeto com propriedades do breadcrumb.
     * @type {?}
     */
    ThfPageDefaultBaseComponent.prototype.breadcrumb;
    /**
     * @abstract
     * @return {?}
     */
    ThfPageDefaultBaseComponent.prototype.setDropdownActions = function () { };
}
