// @totvs/thf-ui/esm2015/lib/components/thf-field/thf-multiselect/thf-multiselect-search/thf-multiselect-search.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ElementRef, EventEmitter, Input, Output, ViewChild } from '@angular/core';
import { isTypeof } from '../../../../utils/util';
/**
 * \@docsPrivate
 *
 * \@description
 *
 * Componente de pesquisa que será criado dentro do dropdown do `thf-multiselect`.
 */
export class ThfMultiselectSearchComponent {
    constructor() {
        /**
         * Evento que será disparado a cada tecla digitada no campo de busca.
         */
        this.change = new EventEmitter();
    }
    /**
     * \@optional
     *
     * \@description
     *
     * Placeholder do campo de pesquisa.
     *
     * > Caso o mesmo não seja informado, o valor padrão será traduzido com base no idioma do navegador (pt, es e en).
     *
     * \@default `Buscar`
     * @param {?} placeholder
     * @return {?}
     */
    set placeholder(placeholder) {
        this._placeholder = placeholder && isTypeof(placeholder, 'string') ? placeholder : this.literals.placeholderSearch;
    }
    /**
     * @return {?}
     */
    get placeholder() {
        return this._placeholder || this.literals.placeholderSearch;
    }
    /**
     * @param {?} event
     * @return {?}
     */
    onChange(event) {
        this.change.emit({ event: event, value: this.inputElement.nativeElement.value });
    }
    /**
     * @return {?}
     */
    setFocus() {
        this.inputElement.nativeElement.focus();
    }
    /**
     * @return {?}
     */
    clean() {
        this.inputElement.nativeElement.value = '';
    }
}
ThfMultiselectSearchComponent.decorators = [
    { type: Component, args: [{
                selector: 'thf-multiselect-search',
                template: "<div class=\"thf-field-container-content thf-multiselect-container-search\">\n\n  <div class=\"thf-field-icon-container-left\">\n    <span class=\"thf-icon thf-icon-search thf-field-icon\"></span>\n  </div>\n\n  <input #inputElement\n    class=\"thf-input thf-input-icon-left\"\n    type=\"text\"\n    [placeholder]=\"placeholder\"\n    (keyup)=\"onChange($event)\">\n\n</div>\n"
            }] }
];
ThfMultiselectSearchComponent.propDecorators = {
    literals: [{ type: Input, args: ['t-literals',] }],
    placeholder: [{ type: Input, args: ['t-placeholder',] }],
    change: [{ type: Output, args: ['t-change',] }],
    inputElement: [{ type: ViewChild, args: ['inputElement', { read: ElementRef, static: true },] }]
};
if (false) {
    /**
     * @type {?}
     * @private
     */
    ThfMultiselectSearchComponent.prototype._placeholder;
    /**
     * Propriedade que recebe as literais definidas no `thf-multiselect`.
     * @type {?}
     */
    ThfMultiselectSearchComponent.prototype.literals;
    /**
     * Evento que será disparado a cada tecla digitada no campo de busca.
     * @type {?}
     */
    ThfMultiselectSearchComponent.prototype.change;
    /** @type {?} */
    ThfMultiselectSearchComponent.prototype.inputElement;
}
