// @totvs/thf-templates/esm2015/lib/components/thf-page-dynamic-search/thf-advanced-filter/thf-advanced-filter.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ViewChild } from '@angular/core';
import { ThfDynamicFormComponent, ThfModalComponent } from '@totvs/thf-ui';
import { ThfAdvancedFilterBaseComponent } from './thf-advanced-filter-base.component';
/**
 * \@docsPrivate
 *
 * \@docsExtends ThfAdvancedFilterBaseComponent
 *
 * \@examplePrivate
 *
 * <example-private name="thf-advanced-filter" title="Totvs Busca Avançada">
 *   <file name="sample-thf-advanced-filter.component.html"> </file>
 *   <file name="sample-thf-advanced-filter.component.ts"> </file>
 * </example-private>
 */
export class ThfAdvancedFilterComponent extends ThfAdvancedFilterBaseComponent {
    constructor() {
        super(...arguments);
        this.filter = {};
        this.primaryAction = {
            action: (/**
             * @return {?}
             */
            () => {
                /** @type {?} */
                const models = this.getValuesFromForm();
                this.searchEvent.emit(models);
                this.thfModal.close();
            }),
            label: this.literals.primaryActionLabel
        };
        this.secondaryAction = {
            action: (/**
             * @return {?}
             */
            () => {
                this.thfModal.close();
            }),
            label: this.literals.secondaryActionLabel
        };
    }
    // Retorna os models dos campos preenchidos
    /**
     * @return {?}
     */
    getValuesFromForm() {
        Object.keys(this.filter).forEach((/**
         * @param {?} property
         * @return {?}
         */
        property => {
            if (this.filter[property] === undefined || this.filter[property] === '') {
                delete this.filter[property];
            }
        }));
        return this.filter;
    }
    /**
     * @return {?}
     */
    open() {
        this.filter = {};
        this.thfModal.open();
    }
}
ThfAdvancedFilterComponent.decorators = [
    { type: Component, args: [{
                selector: 'thf-advanced-filter',
                template: "<thf-modal\n  t-hide-close\n  [t-primary-action]=\"primaryAction\"\n  [t-secondary-action]=\"secondaryAction\"\n  [t-title]=\"literals.title\">\n\n  <thf-dynamic-form\n    [t-fields]=\"filters\"\n    [t-value]=\"filter\">\n  </thf-dynamic-form>\n\n</thf-modal>\n"
            }] }
];
ThfAdvancedFilterComponent.propDecorators = {
    thfModal: [{ type: ViewChild, args: [ThfModalComponent, { static: true },] }],
    thfDynamicForm: [{ type: ViewChild, args: [ThfDynamicFormComponent, { static: true },] }]
};
if (false) {
    /** @type {?} */
    ThfAdvancedFilterComponent.prototype.filter;
    /** @type {?} */
    ThfAdvancedFilterComponent.prototype.thfModal;
    /** @type {?} */
    ThfAdvancedFilterComponent.prototype.thfDynamicForm;
    /** @type {?} */
    ThfAdvancedFilterComponent.prototype.primaryAction;
    /** @type {?} */
    ThfAdvancedFilterComponent.prototype.secondaryAction;
}
