// @totvs/thf-templates/esm2015/lib/components/thf-page-dynamic-search/thf-advanced-filter/thf-advanced-filter-base.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { EventEmitter, Input, Output } from '@angular/core';
import { browserLanguage, thfLocaleDefault } from '../../../utils/util';
/** @type {?} */
export const thfAdvancedFiltersLiteralsDefault = {
    en: {
        title: 'Advanced search',
        primaryActionLabel: 'Apply filters',
        secondaryActionLabel: 'Cancel'
    },
    es: {
        title: 'Búsqueda avanzada',
        primaryActionLabel: 'Aplicar filtros',
        secondaryActionLabel: 'Cancelar'
    },
    pt: {
        title: 'Busca avançada',
        primaryActionLabel: 'Aplicar filtros',
        secondaryActionLabel: 'Cancelar'
    }
};
/**
 * \@docsPrivate
 *
 * \@description
 *
 * Filtro de busca avançada criado a partir de um formulário dinâmico.
 * Componente de uso interno.
 */
export class ThfAdvancedFilterBaseComponent {
    constructor() {
        this._filters = [];
        this.literals = Object.assign({}, thfAdvancedFiltersLiteralsDefault[thfLocaleDefault], thfAdvancedFiltersLiteralsDefault[browserLanguage()]);
        /**
         * Função que será disparada e receberá os valores do formulário ao ser clicado no botão buscar.
         */
        this.searchEvent = new EventEmitter();
    }
    /**
     * Coleção de objetos que implementam a interface ThfDynamicFormField, para definição dos campos que serão criados
     * dinamicamente.
     * @param {?} filters
     * @return {?}
     */
    set filters(filters) {
        this._filters = Array.isArray(filters) ? [...filters] : [];
    }
    /**
     * @return {?}
     */
    get filters() {
        return this._filters;
    }
}
ThfAdvancedFilterBaseComponent.propDecorators = {
    filters: [{ type: Input, args: ['t-filters',] }],
    searchEvent: [{ type: Output, args: ['t-search-event',] }]
};
if (false) {
    /**
     * @type {?}
     * @private
     */
    ThfAdvancedFilterBaseComponent.prototype._filters;
    /** @type {?} */
    ThfAdvancedFilterBaseComponent.prototype.literals;
    /**
     * Função que será disparada e receberá os valores do formulário ao ser clicado no botão buscar.
     * @type {?}
     */
    ThfAdvancedFilterBaseComponent.prototype.searchEvent;
}
