// @totvs/thf-ui/esm5/lib/services/services.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { ThfColorPaletteModule } from './thf-color-palette/thf-color-palette.module';
import { ThfComponentInjectorModule } from './thf-component-injector/thf-component-injector.module';
import { ThfControlPositionModule } from './thf-control-position/thf-control-position.module';
import { ThfDateTimeModule } from './thf-date/thf-date.module';
import { ThfDialogModule } from './thf-dialog/thf-dialog.module';
import { ThfI18nPipe } from './thf-i18n/thf-i18n.pipe';
import { ThfLanguageModule } from './thf-language/thf-language.module';
import { ThfNotificationModule } from './thf-notification/thf-notification.module';
var ThfServicesModule = /** @class */ (function () {
    function ThfServicesModule() {
    }
    ThfServicesModule.decorators = [
        { type: NgModule, args: [{
                    declarations: [
                        ThfI18nPipe
                    ],
                    imports: [
                        ThfColorPaletteModule,
                        ThfComponentInjectorModule,
                        ThfControlPositionModule,
                        ThfDateTimeModule,
                        ThfDialogModule,
                        ThfLanguageModule,
                        ThfNotificationModule
                    ],
                    exports: [
                        ThfColorPaletteModule,
                        ThfComponentInjectorModule,
                        ThfControlPositionModule,
                        ThfDateTimeModule,
                        ThfDialogModule,
                        ThfI18nPipe,
                        ThfNotificationModule
                    ],
                    providers: [],
                    bootstrap: []
                },] }
    ];
    return ThfServicesModule;
}());
export { ThfServicesModule };
