// @totvs/thf-ui/esm5/lib/components/thf-field/thf-lookup/thf-lookup-modal/thf-lookup-modal.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, ElementRef, ViewChild } from '@angular/core';
import { fromEvent } from 'rxjs';
import { debounceTime, filter } from 'rxjs/operators';
import { ThfLookupModalBaseComponent } from '../thf-lookup-modal/thf-lookup-modal-base.component';
/**
 * \@docsPrivate
 *
 * \@docsExtends ThfLookupModalBaseComponent
 */
var ThfLookupModalComponent = /** @class */ (function (_super) {
    tslib_1.__extends(ThfLookupModalComponent, _super);
    function ThfLookupModalComponent() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.keyUpObservable = null;
        _this.containerHeight = 375;
        _this.tableHeight = 370;
        return _this;
    }
    /**
     * @return {?}
     */
    ThfLookupModalComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        _super.prototype.ngOnInit.call(this);
        this.initializeEventInput();
        this.setTableHeight();
    };
    /**
     * @return {?}
     */
    ThfLookupModalComponent.prototype.initializeEventInput = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this.keyUpObservable = fromEvent(this.inputSearchEl.nativeElement, 'keyup')
            .pipe(filter((/**
         * @param {?} e
         * @return {?}
         */
        function (e) { return _this.validateEnterPressed(e); })), debounceTime(400));
        this.keyUpObservable.subscribe((/**
         * @return {?}
         */
        function () {
            _this.search();
        }));
    };
    /**
     * @return {?}
     */
    ThfLookupModalComponent.prototype.openModal = /**
     * @return {?}
     */
    function () {
        this.thfModal.open();
    };
    /**
     * @private
     * @return {?}
     */
    ThfLookupModalComponent.prototype.setTableHeight = /**
     * @private
     * @return {?}
     */
    function () {
        if (window.innerHeight < 615) {
            this.tableHeight -= 50;
            this.containerHeight -= 50;
        }
    };
    /**
     * @private
     * @param {?} e
     * @return {?}
     */
    ThfLookupModalComponent.prototype.validateEnterPressed = /**
     * @private
     * @param {?} e
     * @return {?}
     */
    function (e) {
        return e.keyCode === 13;
    };
    ThfLookupModalComponent.decorators = [
        { type: Component, args: [{
                    selector: 'thf-lookup-modal',
                    template: "<thf-modal\n  t-click-out=\"false\"\n  t-hide-close=\"false\"\n  t-size=\"lg\"\n  [t-primary-action]=\"primaryAction\"\n  [t-secondary-action]=\"secondaryAction\"\n  [t-title]=\"title\">\n\n  <thf-field-container\n    class=\"thf-lookup-header thf-md-6 thf-pull-right\"\n    [t-optional]=\"false\">\n\n    <div class=\"thf-field-container-content\">\n      <input #inpsearch\n        class=\"thf-input thf-input-icon-right\"\n        name=\"contentSearch\"\n        [(ngModel)]=\"searchValue\"\n        [placeholder]=\"literals.modalPlaceholder\"\n        type=\"text\">\n\n      <div class=\"thf-field-icon-container-right\">\n        <span #iconLookup\n          class=\"thf-icon thf-field-icon thf-icon-search\"\n          (click)=\"search()\">\n        </span>\n      </div>\n    </div>\n\n  </thf-field-container>\n\n  <div class=\"thf-row thf-lookup-container-table\" [style.height.px]=\"containerHeight\">\n\n    <thf-table\n      class=\"thf-md-12\"\n      t-checkbox=\"true\"\n      t-hide-detail=\"true\"\n      t-single-select=\"true\"\n      t-sort=\"true\"\n      [t-columns]=\"columns\"\n      [t-height]=\"tableHeight\"\n      [t-items]=\"items\"\n      [t-literals]=\"tableLiterals\"\n      [t-loading]=\"isLoading\"\n      [t-show-more-disabled]=\"!hasNext\"\n      (t-show-more)=\"showMoreEvent()\">\n    </thf-table>\n\n  </div>\n\n</thf-modal>\n"
                }] }
    ];
    ThfLookupModalComponent.propDecorators = {
        inputSearchEl: [{ type: ViewChild, args: ['inpsearch', { static: true },] }]
    };
    return ThfLookupModalComponent;
}(ThfLookupModalBaseComponent));
export { ThfLookupModalComponent };
if (false) {
    /** @type {?} */
    ThfLookupModalComponent.prototype.inputSearchEl;
    /** @type {?} */
    ThfLookupModalComponent.prototype.keyUpObservable;
    /** @type {?} */
    ThfLookupModalComponent.prototype.containerHeight;
    /** @type {?} */
    ThfLookupModalComponent.prototype.tableHeight;
}
