// @totvs/thf-ui/esm5/lib/components/thf-field/thf-lookup/services/thf-lookup-filter.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { isTypeof } from '../../../../utils/util';
/**
 * \@docsPrivate
 *
 * \@description
 *
 * Serviço padrão utilizado para filtrar os dados do componente thf-lookup.
 */
var ThfLookupFilterService = /** @class */ (function () {
    function ThfLookupFilterService(httpClient) {
        this.httpClient = httpClient;
    }
    /**
     * @param {?} filter
     * @param {?} page
     * @param {?=} pageSize
     * @param {?=} filterParams
     * @return {?}
     */
    ThfLookupFilterService.prototype.getFilteredData = /**
     * @param {?} filter
     * @param {?} page
     * @param {?=} pageSize
     * @param {?=} filterParams
     * @return {?}
     */
    function (filter, page, pageSize, filterParams) {
        /** @type {?} */
        var validatedFilterParams = this.validateParams(filterParams);
        return this.httpClient.get(this.url, { params: tslib_1.__assign({ page: page.toString(), pageSize: pageSize.toString() }, validatedFilterParams, { filter: filter }) });
    };
    /**
     * @param {?} value
     * @param {?=} filterParams
     * @return {?}
     */
    ThfLookupFilterService.prototype.getObjectByValue = /**
     * @param {?} value
     * @param {?=} filterParams
     * @return {?}
     */
    function (value, filterParams) {
        /** @type {?} */
        var validatedFilterParams = this.validateParams(filterParams);
        return this.httpClient.get(this.url + "/" + value, { params: validatedFilterParams });
    };
    /**
     * @param {?} url
     * @return {?}
     */
    ThfLookupFilterService.prototype.setUrl = /**
     * @param {?} url
     * @return {?}
     */
    function (url) {
        this.url = url;
    };
    /**
     * @private
     * @param {?} params
     * @return {?}
     */
    ThfLookupFilterService.prototype.validateParams = /**
     * @private
     * @param {?} params
     * @return {?}
     */
    function (params) {
        return isTypeof(params, 'object') && !Array.isArray(params) ? params : undefined;
    };
    ThfLookupFilterService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    ThfLookupFilterService.ctorParameters = function () { return [
        { type: HttpClient }
    ]; };
    return ThfLookupFilterService;
}());
export { ThfLookupFilterService };
if (false) {
    /**
     * @type {?}
     * @private
     */
    ThfLookupFilterService.prototype.url;
    /**
     * @type {?}
     * @private
     */
    ThfLookupFilterService.prototype.httpClient;
}
