// @totvs/thf-ui/esm5/lib/components/thf-field/thf-login/thf-login.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, ElementRef, forwardRef } from '@angular/core';
import { NG_VALUE_ACCESSOR, NG_VALIDATORS } from '@angular/forms';
import { ThfInputGeneric } from '../thf-input-generic/thf-input-generic';
/**
 * \@docsExtends ThfInputBaseComponent
 *
 * \@description
 * O thf-login é um input específico para login. Já possui tipo, estilo e ícone predefinidos.
 *
 * \@example
 *
 * <example name="thf-login-basic" title="Totvs Login Basic">
 *  <file name="sample-thf-login-basic/sample-thf-login-basic.component.html"> </file>
 *  <file name="sample-thf-login-basic/sample-thf-login-basic.component.ts"> </file>
 * </example>
 *
 * <example name="thf-login-labs" title="Totvs Login Labs">
 *  <file name="sample-thf-login-labs/sample-thf-login-labs.component.html"> </file>
 *  <file name="sample-thf-login-labs/sample-thf-login-labs.component.ts"> </file>
 * </example>
 *
 * <example name="thf-login-confirm" title="Totvs Login - Confirm Identity">
 *  <file name="sample-thf-login-confirm/sample-thf-login-confirm.component.html"> </file>
 *  <file name="sample-thf-login-confirm/sample-thf-login-confirm.component.ts"> </file>
 * </example>
 *
 */
var ThfLoginComponent = /** @class */ (function (_super) {
    tslib_1.__extends(ThfLoginComponent, _super);
    function ThfLoginComponent(el) {
        var _this = _super.call(this, el) || this;
        _this.type = 'text';
        return _this;
    }
    /**
     * @param {?} c
     * @return {?}
     */
    ThfLoginComponent.prototype.extraValidation = /**
     * @param {?} c
     * @return {?}
     */
    function (c) {
        return null;
    };
    ThfLoginComponent.decorators = [
        { type: Component, args: [{
                    selector: 'thf-login',
                    template: "<thf-field-container\n  [t-label]=\"label\"\n  [t-help]=\"help\"\n  [t-optional]=\"!required && optional\">\n\n  <div class=\"thf-field-container-content\">\n    <div class=\"thf-field-icon-container-left\">\n      <span class=\"thf-icon thf-field-icon thf-icon-user\" [class.thf-field-icon-disabled]=\"disabled\"></span>\n    </div>\n\n    <input #inp\n      class=\"thf-input thf-input-icon-left\"\n      [attr.name]=\"name\"\n      [class.thf-input-icon-right]=\"clean\"\n      [disabled]=\"disabled\"\n      [placeholder]=\"placeholder\"\n      [readonly]=\"readonly\"\n      [required]=\"required\"\n      [type]=\"type\"\n      (blur)=\"eventOnBlur($event)\"\n      (click)=\"eventOnClick($event)\"\n      (focus)=\"eventOnFocus($event)\"\n      (input)=\"eventOnInput($event)\">\n\n    <div class=\"thf-field-icon-container-right\">\n      <thf-clean [t-element-ref]=\"inputEl\" (t-change-event)=\"clear($event)\"></thf-clean>\n    </div>\n  </div>\n\n  <thf-field-container-bottom\n    [t-error-pattern]=\"getErrorPattern()\">\n  </thf-field-container-bottom>\n\n</thf-field-container>\n",
                    providers: [
                        {
                            provide: NG_VALUE_ACCESSOR,
                            useExisting: forwardRef((/**
                             * @return {?}
                             */
                            function () { return ThfLoginComponent; })),
                            multi: true,
                        },
                        {
                            provide: NG_VALIDATORS,
                            useExisting: forwardRef((/**
                             * @return {?}
                             */
                            function () { return ThfLoginComponent; })),
                            multi: true,
                        }
                    ]
                }] }
    ];
    /** @nocollapse */
    ThfLoginComponent.ctorParameters = function () { return [
        { type: ElementRef }
    ]; };
    return ThfLoginComponent;
}(ThfInputGeneric));
export { ThfLoginComponent };
if (false) {
    /** @type {?} */
    ThfLoginComponent.prototype.type;
}
