// @totvs/thf-ui/esm2015/lib/interceptors/thf-http-request/thf-http-request.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { HTTP_INTERCEPTORS } from '@angular/common/http';
import { NgModule } from '@angular/core';
import { ThfComponentInjectorService } from '../../services/thf-component-injector/thf-component-injector.service';
import { ThfHttpRequesControltService } from './thf-http-request-control-service';
import { ThfHttpRequestInterceptorService } from './thf-http-request-interceptor.service';
import { ThfLoadingComponent } from '../../components/thf-loading/thf-loading.component';
import { ThfLoadingOverlayComponent } from '../../components/thf-loading/thf-loading-overlay/thf-loading-overlay.component';
export class ThfHttpRequestModule {
}
ThfHttpRequestModule.decorators = [
    { type: NgModule, args: [{
                providers: [
                    ThfHttpRequesControltService,
                    ThfHttpRequestInterceptorService,
                    ThfComponentInjectorService,
                    {
                        provide: HTTP_INTERCEPTORS,
                        useClass: ThfHttpRequestInterceptorService,
                        multi: true
                    }
                ],
                entryComponents: [ThfLoadingOverlayComponent, ThfLoadingComponent]
            },] }
];
