// @totvs/thf-ui/esm2015/lib/components/thf-tabs/thf-tab/thf-tab.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ElementRef } from '@angular/core';
import { ThfTabBaseComponent } from './thf-tab-base.component';
/**
 * \@docsExtends ThfTabBaseComponent
 */
export class ThfTabComponent extends ThfTabBaseComponent {
    /**
     * @param {?} elementRef
     */
    constructor(elementRef) {
        super();
        this.elementRef = elementRef;
    }
    /**
     * @return {?}
     */
    ngAfterContentInit() {
        this.setDisplayOnActive();
    }
    /**
     * @protected
     * @return {?}
     */
    setDisplayOnActive() {
        this.elementRef.nativeElement.style.display = this.active ? '' : 'none';
    }
}
ThfTabComponent.decorators = [
    { type: Component, args: [{
                selector: 'thf-tab',
                template: "<ng-content></ng-content>\n"
            }] }
];
/** @nocollapse */
ThfTabComponent.ctorParameters = () => [
    { type: ElementRef }
];
if (false) {
    /**
     * @type {?}
     * @private
     */
    ThfTabComponent.prototype.elementRef;
}
