// @totvs/thf-ui/esm2015/lib/components/thf-menu/services/thf-menu-items.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { Subject } from 'rxjs';
/**
 * \@docsPrivate
 *
 * \@description
 *
 * Serviço que implementa a comunicação entre os items do thf-menu.
 */
export class ThfMenuItemsService {
    constructor() {
        this.subjectParent = new Subject();
        this.subjectChild = new Subject();
    }
    // Envia informações do click do thf-menu-item para o thf-menu
    /**
     * @param {?} menu
     * @return {?}
     */
    sendToParentMenuClicked(menu) {
        this.subjectChild.next(menu);
    }
    // Recebe do thf-menu-item sua informação de click.
    /**
     * @return {?}
     */
    receiveFromChildMenuClicked() {
        return this.subjectChild.asObservable();
    }
    // Envia para os thf-menu-item a resposta do processamento de click de um thf-menu-item.
    /**
     * @param {?} menu
     * @return {?}
     */
    sendToChildMenuClicked(menu) {
        this.subjectParent.next(menu);
    }
    // Recebe do thf-menu as informações processadas do click de um thf-menu-item.
    /**
     * @return {?}
     */
    receiveFromParentMenuClicked() {
        return this.subjectParent.asObservable();
    }
}
ThfMenuItemsService.decorators = [
    { type: Injectable }
];
if (false) {
    /**
     * @type {?}
     * @private
     */
    ThfMenuItemsService.prototype.subjectParent;
    /**
     * @type {?}
     * @private
     */
    ThfMenuItemsService.prototype.subjectChild;
}
