// @totvs/thf-ui/esm2015/lib/components/thf-field/thf-field.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { FormsModule } from '@angular/forms';
import { HttpClientModule } from '@angular/common/http';
import { ThfButtonGroupModule } from '../thf-button-group/index';
import { ThfButtonModule } from '../../components/thf-button/index';
import { ThfDisclaimerModule } from './../thf-disclaimer/thf-disclaimer.module';
import { ThfLoadingModule } from '../../components/thf-loading/index';
import { ThfModalModule } from '../../components/thf-modal/thf-modal.module';
import { ThfTableModule } from '../../components/thf-table/thf-table.module';
import { ThfCalendarComponent } from './thf-datepicker/thf-calendar/thf-calendar.component';
import { ThfCheckboxGroupComponent } from './thf-checkbox-group/thf-checkbox-group.component';
import { ThfCleanComponent } from './thf-clean/thf-clean.component';
import { ThfComboComponent } from './thf-combo/thf-combo.component';
import { ThfDatepickerComponent } from './thf-datepicker/thf-datepicker.component';
import { ThfDatepickerRangeComponent } from './thf-datepicker-range/thf-datepicker-range.component';
import { ThfDecimalComponent } from './thf-decimal/thf-decimal.component';
import { ThfEmailComponent } from './thf-email/thf-email.component';
import { ThfFieldContainerComponent } from './thf-field-container/thf-field-container.component';
import { ThfFieldContainerBottomComponent } from './thf-field-container/thf-field-container-bottom/thf-field-container-bottom.component';
import { ThfLoginComponent } from './thf-login/thf-login.component';
import { ThfLookupComponent } from './thf-lookup/thf-lookup.component';
import { ThfLookupModalComponent } from './thf-lookup/thf-lookup-modal/thf-lookup-modal.component';
import { ThfMultiselectDropdownComponent } from './thf-multiselect/thf-multiselect-dropdown/thf-multiselect-dropdown.component';
import { ThfMultiselectComponent } from './thf-multiselect/thf-multiselect.component';
import { ThfMultiselectItemComponent } from './thf-multiselect/thf-multiselect-item/thf-multiselect-item.component';
import { ThfMultiselectSearchComponent } from './thf-multiselect/thf-multiselect-search/thf-multiselect-search.component';
import { ThfRichTextBodyComponent } from './thf-rich-text/thf-rich-text-body/thf-rich-text-body.component';
import { ThfRichTextComponent } from './thf-rich-text/thf-rich-text.component';
import { ThfRichTextToolbarComponent } from './thf-rich-text/thf-rich-text-toolbar/thf-rich-text-toolbar.component';
import { ThfInputComponent } from './thf-input/thf-input.component';
import { ThfNumberComponent } from './thf-number/thf-number.component';
import { ThfPasswordComponent } from './thf-password/thf-password.component';
import { ThfRadioGroupComponent } from './thf-radio-group/thf-radio-group.component';
import { ThfSelectComponent } from './thf-select/thf-select.component';
import { ThfSelectOptionTemplateDirective } from './thf-select/thf-select-option-template/thf-select-option-template.directive';
import { ThfSwitchComponent } from './thf-switch/thf-switch.component';
import { ThfTextareaComponent } from './thf-textarea/thf-textarea.component';
import { ThfUploadComponent } from './thf-upload/thf-upload.component';
import { ThfUrlComponent } from './thf-url/thf-url.component';
/**
 * \@description
 *
 * Módulo dos componentes thf-combo, thf-checkbox-group, thf-datepicker, thf-datepicker-range, thf-email, thf-input, thf-lookup,
 * thf-number, thf-multiselect, thf-password, thf-radio-group, thf-select, thf-switch, thf-textarea, thf-upload e thf-url.
 */
export class ThfFieldModule {
}
ThfFieldModule.decorators = [
    { type: NgModule, args: [{
                imports: [
                    CommonModule,
                    FormsModule,
                    HttpClientModule,
                    ThfButtonModule,
                    ThfDisclaimerModule,
                    ThfLoadingModule,
                    ThfModalModule,
                    ThfTableModule,
                    ThfButtonGroupModule
                ],
                exports: [
                    ThfCheckboxGroupComponent,
                    ThfCleanComponent,
                    ThfComboComponent,
                    ThfDecimalComponent,
                    ThfDatepickerComponent,
                    ThfDatepickerRangeComponent,
                    ThfEmailComponent,
                    ThfInputComponent,
                    ThfLoginComponent,
                    ThfLookupComponent,
                    ThfLookupModalComponent,
                    ThfMultiselectComponent,
                    ThfNumberComponent,
                    ThfPasswordComponent,
                    ThfRadioGroupComponent,
                    ThfRichTextComponent,
                    ThfSelectComponent,
                    ThfSelectOptionTemplateDirective,
                    ThfSwitchComponent,
                    ThfTextareaComponent,
                    ThfUploadComponent,
                    ThfUrlComponent
                ],
                declarations: [
                    ThfCalendarComponent,
                    ThfCheckboxGroupComponent,
                    ThfCleanComponent,
                    ThfComboComponent,
                    ThfDecimalComponent,
                    ThfDatepickerComponent,
                    ThfDatepickerRangeComponent,
                    ThfEmailComponent,
                    ThfFieldContainerComponent,
                    ThfFieldContainerBottomComponent,
                    ThfInputComponent,
                    ThfLoginComponent,
                    ThfLookupComponent,
                    ThfLookupModalComponent,
                    ThfMultiselectComponent,
                    ThfMultiselectDropdownComponent,
                    ThfMultiselectItemComponent,
                    ThfMultiselectSearchComponent,
                    ThfNumberComponent,
                    ThfPasswordComponent,
                    ThfRadioGroupComponent,
                    ThfRichTextBodyComponent,
                    ThfRichTextComponent,
                    ThfRichTextToolbarComponent,
                    ThfSelectComponent,
                    ThfSelectOptionTemplateDirective,
                    ThfSwitchComponent,
                    ThfTextareaComponent,
                    ThfUploadComponent,
                    ThfUrlComponent
                ],
                providers: [],
                entryComponents: [
                    ThfCalendarComponent,
                    ThfLookupModalComponent
                ]
            },] }
];
