// @totvs/thf-ui/esm2015/lib/components/thf-page/thf-page-list/thf-page-list-base.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Input } from '@angular/core';
import { browserLanguage, thfLocaleDefault } from './../../../utils/util';
import { ThfPageDefaultBaseComponent } from '../thf-page-default/thf-page-default-base.component';
/** @type {?} */
export const thfPageListLiteralsDefault = {
    en: (/** @type {?} */ ({
        otherActions: 'Other actions'
    })),
    es: (/** @type {?} */ ({
        otherActions: 'Otras acciones'
    })),
    pt: (/** @type {?} */ ({
        otherActions: 'Outras ações'
    }))
};
/**
 * \@description
 *
 * O componente `thf-page-list` é utilizado como o container principal para as telas de listagem de dados,
 * podendo ser apresentado como lista ou tabela.
 *
 * Este componente possibilita realizar filtro dos dados, no qual permite que seja atribuido uma função que será executada no momento
 * da filtragem. Este comportamento pode ser acionado tanto ao *click* do ícone [thf-icon-search](/guides/icons)
 * quanto ao pressionar da tecla *ENTER* quando o foco estiver no campo de pesquisa.
 *
 * Para facilitar a manipulação e visualização dos filtros aplicados, é possível também utilizar o componente
 * [`thf-disclaimer-group`](/documentation/thf-disclaimer-group).
 * @abstract
 */
export class ThfPageListBaseComponent extends ThfPageDefaultBaseComponent {
    /**
     * \@optional
     *
     * \@description
     *
     * Objeto que implementa as propriedades da interface `ThfDisclaimerGroup`.
     * @param {?} value
     * @return {?}
     */
    set disclaimerGroup(value) {
        if (!value) {
            value = (/** @type {?} */ ({}));
        }
        this._disclaimerGroup = value;
    }
    /**
     * @return {?}
     */
    get disclaimerGroup() {
        return this._disclaimerGroup;
    }
    /**
     * \@optional
     *
     * \@description
     *
     * Objeto com as literais usadas no `thf-page-list`.
     *
     * Existem duas maneiras de customizar o componente, passando um objeto com todas as literais disponíveis:
     *
     * ```
     *  const customLiterals: ThfPageListLiterals = {
     *    otherActions: 'Mais ações'
     *  };
     * ```
     *
     * Ou passando apenas as literais que deseja customizar:
     *
     * ```
     *  const customLiterals: ThfPageListLiterals = {
     *    otherActions: 'Ações da página'
     *  };
     * ```
     *
     * E para carregar as literais customizadas, basta apenas passar o objeto para o componente.
     *
     * ```
     * <thf-page-list
     *   [t-literals]="customLiterals">
     * </thf-page-list>
     * ```
     *
     *  > O objeto padrão de literais será traduzido de acordo com o idioma do browser (pt, en, es).
     * @param {?} value
     * @return {?}
     */
    set literals(value) {
        if (value instanceof Object && !(value instanceof Array)) {
            this._literals = Object.assign({}, thfPageListLiteralsDefault[thfLocaleDefault], thfPageListLiteralsDefault[browserLanguage()], value);
        }
        else {
            this._literals = thfPageListLiteralsDefault[browserLanguage()];
        }
    }
    /**
     * @return {?}
     */
    get literals() {
        return this._literals || thfPageListLiteralsDefault[browserLanguage()];
    }
}
ThfPageListBaseComponent.propDecorators = {
    breadcrumb: [{ type: Input, args: ['t-breadcrumb',] }],
    disclaimerGroup: [{ type: Input, args: ['t-disclaimer-group',] }],
    filter: [{ type: Input, args: ['t-filter',] }],
    literals: [{ type: Input, args: ['t-literals',] }]
};
if (false) {
    /**
     * @type {?}
     * @private
     */
    ThfPageListBaseComponent.prototype._disclaimerGroup;
    /**
     * @type {?}
     * @private
     */
    ThfPageListBaseComponent.prototype._literals;
    /**
     * \@optional
     *
     * \@description
     *
     * Objeto que implementa as propriedades da interface `ThfBreadcrumb`.
     * @type {?}
     */
    ThfPageListBaseComponent.prototype.breadcrumb;
    /**
     * \@description
     *
     * Objeto que implementa as propriedades da interface `ThfPageFilter`.
     * @type {?}
     */
    ThfPageListBaseComponent.prototype.filter;
    /**
     * @type {?}
     * @protected
     */
    ThfPageListBaseComponent.prototype.resizeListener;
}
