// @totvs/thf-templates/esm2015/lib/components/thf-page-blocked-user/thf-page-blocked-user-reason/thf-page-blocked-user-reason.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ChangeDetectorRef, Component, Input } from '@angular/core';
import { browserLanguage, thfLocaleDefault } from '../../../utils/util';
import { thfPageBlockedUserLiterals } from './../literals/i18n/thf-page-blocked-user-literals';
import { ThfPageBlockedUserReason } from '../enums/thf-page-blocked-user-reason.enum';
export class ThfPageBlockedUserReasonComponent {
    /**
     * @param {?} changeDetector
     */
    constructor(changeDetector) {
        this.changeDetector = changeDetector;
    }
    /**
     * @param {?} changes
     * @return {?}
     */
    ngOnChanges(changes) {
        if (changes.reason || changes.params) {
            this.getLiterals();
        }
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this.getLiterals();
    }
    /**
     * @return {?}
     */
    getImageByReasonType() {
        /** @type {?} */
        let reasonImage;
        switch (this.reason) {
            case 'none': {
                reasonImage = 'big-lock';
                break;
            }
            case 'exceededAttempts': {
                reasonImage = 'blocked-user';
                break;
            }
            case 'expiredPassword': {
                reasonImage = 'expired';
                break;
            }
        }
        return `./assets/images/${reasonImage}.svg`;
    }
    /**
     * @return {?}
     */
    getParams() {
        this.literalParams =
            this.reason === 'expiredPassword' ? [this.params.days, this.params.days] : [this.params.attempts, this.params.hours];
    }
    /**
     * @private
     * @return {?}
     */
    getLiterals() {
        this.getParams();
        this.literals = Object.assign({}, thfPageBlockedUserLiterals[this.reason][thfLocaleDefault], thfPageBlockedUserLiterals[this.reason][browserLanguage()]);
        this.changeDetector.detectChanges();
    }
}
ThfPageBlockedUserReasonComponent.decorators = [
    { type: Component, args: [{
                selector: 'thf-page-blocked-user-reason',
                template: "<div class=\"thf-page-blocked-user-reason-content\">\n  <img \n    class=\"thf-page-blocked-user-image thf-mb-3\" \n    [src]=\"getImageByReasonType()\"> \n  <div class=\"thf-page-blocked-user-header thf-mb-md-3 thf-row\">\n    <div class=\"thf-md-12\">{{ literals?.title }}</div>\n  </div>\n  <div class=\"thf-page-blocked-user-text thf-font-text thf-row\">\n    <div class=\"thf-mb-sm-2 thf-mb-md-3 thf-md-12 thf-page-blocked-user-text-bold\">{{ literals?.firstPhrase | thfI18n:literalParams }}</div>\n    <div class=\"thf-mb-sm-2 thf-mb-md-3 thf-md-12\">{{ literals?.secondPhrase }}</div>\n    <div class=\"thf-mb-sm-2 thf-mb-md-3 thf-md-12\" *ngIf=\"literals?.thirdPhrase\">{{ literals?.thirdPhrase }}</div>\n  </div>\n</div>"
            }] }
];
/** @nocollapse */
ThfPageBlockedUserReasonComponent.ctorParameters = () => [
    { type: ChangeDetectorRef }
];
ThfPageBlockedUserReasonComponent.propDecorators = {
    params: [{ type: Input, args: ['t-params',] }],
    reason: [{ type: Input, args: ['t-reason',] }]
};
if (false) {
    /** @type {?} */
    ThfPageBlockedUserReasonComponent.prototype.literalParams;
    /** @type {?} */
    ThfPageBlockedUserReasonComponent.prototype.literals;
    /** @type {?} */
    ThfPageBlockedUserReasonComponent.prototype.params;
    /** @type {?} */
    ThfPageBlockedUserReasonComponent.prototype.reason;
    /**
     * @type {?}
     * @private
     */
    ThfPageBlockedUserReasonComponent.prototype.changeDetector;
}
