// @totvs/thf-ui/esm5/lib/components/thf-field/thf-rich-text/thf-rich-text-body/thf-rich-text-body.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ElementRef, EventEmitter, Input, Output, ViewChild } from '@angular/core';
/** @type {?} */
var thfRichTextBodyCommands = [
    'bold', 'italic', 'underline', 'justifyleft', 'justifycenter', 'justifyright', 'justifyfull', 'insertUnorderedList'
];
var ThfRichTextBodyComponent = /** @class */ (function () {
    function ThfRichTextBodyComponent() {
        this.commands = new EventEmitter();
        this.value = new EventEmitter();
    }
    /**
     * @return {?}
     */
    ThfRichTextBodyComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this.bodyElement.nativeElement.designMode = 'on';
        // timeout necessário para setar o valor vindo do writeValue do componente principal.
        setTimeout((/**
         * @return {?}
         */
        function () { _this.updateValueWithModelValue(); }));
    };
    /**
     * @param {?} command
     * @return {?}
     */
    ThfRichTextBodyComponent.prototype.executeCommand = /**
     * @param {?} command
     * @return {?}
     */
    function (command) {
        this.bodyElement.nativeElement.focus();
        document.execCommand(command, false, null);
        this.updateModel();
        this.value.emit(this.modelValue);
    };
    /**
     * @return {?}
     */
    ThfRichTextBodyComponent.prototype.onClick = /**
     * @return {?}
     */
    function () {
        this.emitSelectionCommands();
    };
    /**
     * @return {?}
     */
    ThfRichTextBodyComponent.prototype.onKeyUp = /**
     * @return {?}
     */
    function () {
        // Tratamento necessário para eliminar a tag <br> criada no firefox quando o body for limpo.
        if (!this.bodyElement.nativeElement.innerText.trim() && this.bodyElement.nativeElement.firstChild) {
            this.bodyElement.nativeElement.firstChild.remove();
        }
        this.updateModel();
        this.emitSelectionCommands();
    };
    /**
     * @return {?}
     */
    ThfRichTextBodyComponent.prototype.update = /**
     * @return {?}
     */
    function () {
        var _this = this;
        setTimeout((/**
         * @return {?}
         */
        function () { _this.updateModel(); }));
        setTimeout((/**
         * @return {?}
         */
        function () { _this.onKeyUp(); }));
    };
    /**
     * @private
     * @return {?}
     */
    ThfRichTextBodyComponent.prototype.emitSelectionCommands = /**
     * @private
     * @return {?}
     */
    function () {
        /** @type {?} */
        var commands = thfRichTextBodyCommands.filter((/**
         * @param {?} command
         * @return {?}
         */
        function (command) {
            return document.queryCommandState(command);
        }));
        this.commands.emit(commands);
    };
    /**
     * @private
     * @return {?}
     */
    ThfRichTextBodyComponent.prototype.updateModel = /**
     * @private
     * @return {?}
     */
    function () {
        this.modelValue = this.bodyElement.nativeElement.innerHTML;
        this.value.emit(this.modelValue);
    };
    /**
     * @private
     * @return {?}
     */
    ThfRichTextBodyComponent.prototype.updateValueWithModelValue = /**
     * @private
     * @return {?}
     */
    function () {
        if (this.modelValue) {
            this.bodyElement.nativeElement.insertAdjacentHTML('afterbegin', this.modelValue);
        }
    };
    ThfRichTextBodyComponent.decorators = [
        { type: Component, args: [{
                    selector: 'thf-rich-text-body',
                    template: "<div #bodyElement\n  class=\"thf-rich-text-body\"\n  [attr.contenteditable]=\"!readonly\"\n  [attr.data-placeholder]=\"placeholder\"\n  [style.height.px]=\"height\"\n  (click)=\"onClick()\"\n  (cut)=\"update()\"\n  (keyup)=\"onKeyUp()\"\n  (paste)=\"update()\">\n</div>\n"
                }] }
    ];
    ThfRichTextBodyComponent.propDecorators = {
        bodyElement: [{ type: ViewChild, args: ['bodyElement', { static: true },] }],
        height: [{ type: Input, args: ['t-height',] }],
        modelValue: [{ type: Input, args: ['t-model-value',] }],
        placeholder: [{ type: Input, args: ['t-placeholder',] }],
        readonly: [{ type: Input, args: ['t-readonly',] }],
        commands: [{ type: Output, args: ['t-commands',] }],
        value: [{ type: Output, args: ['t-value',] }]
    };
    return ThfRichTextBodyComponent;
}());
export { ThfRichTextBodyComponent };
if (false) {
    /** @type {?} */
    ThfRichTextBodyComponent.prototype.bodyElement;
    /** @type {?} */
    ThfRichTextBodyComponent.prototype.height;
    /** @type {?} */
    ThfRichTextBodyComponent.prototype.modelValue;
    /** @type {?} */
    ThfRichTextBodyComponent.prototype.placeholder;
    /** @type {?} */
    ThfRichTextBodyComponent.prototype.readonly;
    /** @type {?} */
    ThfRichTextBodyComponent.prototype.commands;
    /** @type {?} */
    ThfRichTextBodyComponent.prototype.value;
}
