// @totvs/thf-ui/esm2015/lib/components/thf-menu-panel/thf-menu-panel.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ViewContainerRef } from '@angular/core';
import { Location } from '@angular/common';
import { NavigationEnd, Router } from '@angular/router';
import { callFunction, getFormattedLink, openExternalLink } from '../../utils/util';
import { ThfMenuPanelBaseComponent } from './thf-menu-panel-base.component';
import { ThfMenuPanelItemsService } from './services/thf-menu-panel-items.service';
/**
 * \@docsExtends ThfMenuPanelBaseComponent
 *
 * \@description
 *
 * Para o menu funcionar corretamente é necessário importar o `RouterModule` e `Routes` do módulo principal de
 *  sua aplicação:
 *
 * ````
 * import { RouterModule, Routes } from '\@angular/router';
 *
 * ...
 *
 * \@NgModule({
 *   imports: [
 *     RouterModule,
 *     Routes,
 *     ...
 *     ThfModule,
 *     ...
 *   ],
 *   declarations: [
 *     AppComponent
 *   ],
 *   providers: [],
 *   bootstrap: [AppComponent]
 * })
 * export class AppModule { }
 * ```
 *
 * Além disso é necessário criar um módulo configurando as rotas da aplicação.
 *
 * ```
 * import { NgModule } from '\@angular/core';
 *
 * import { RouterModule, Routes } from '\@angular/router';
 *
 * import { HelloWorldComponent } from './hello-world/hello-world.component';
 *
 * const routes: Routes = [
 *   {path: 'hello-world', component: HelloWorldComponent}
 * ];
 *
 * \@NgModule({
 *   imports: [RouterModule.forRoot(routes, {useHash: true})],
 *   exports: [RouterModule]
 * })
 * export class AppRoutingModule {}
 * ```
 *
 * \@example
 *
 * <example name="thf-menu-panel-basic" title="Totvs Menu Panel Basic">
 *   <file name="sample-thf-menu-panel-basic/sample-thf-menu-panel-basic.component.html"> </file>
 *   <file name="sample-thf-menu-panel-basic/sample-thf-menu-panel-basic.component.ts"> </file>
 * </example>
 *
 * <example name="thf-menu-panel-labs" title="Totvs Menu Panel Labs">
 *   <file name="sample-thf-menu-panel-labs/sample-thf-menu-panel-labs.component.html"> </file>
 *   <file name="sample-thf-menu-panel-labs/sample-thf-menu-panel-labs.component.ts"> </file>
 * </example>
 *
 * <example name="thf-menu-panel-customer" title="Totvs Menu Panel - Customers">
 *   <file name="sample-thf-menu-panel-customer/sample-thf-menu-panel-customer.component.html"> </file>
 *   <file name="sample-thf-menu-panel-customer/sample-thf-menu-panel-customer.component.ts"> </file>
 * </example>
 */
export class ThfMenuPanelComponent extends ThfMenuPanelBaseComponent {
    /**
     * @param {?} viewRef
     * @param {?} location
     * @param {?} menuItemsService
     * @param {?} router
     */
    constructor(viewRef, location, menuItemsService, router) {
        super();
        this.location = location;
        this.menuItemsService = menuItemsService;
        this.router = router;
        this.parentRef = viewRef['_view']['component'];
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        this.itemSubscription.unsubscribe();
        this.routeSubscription.unsubscribe();
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this.subscribeToMenuItem();
        this.subscribeToRoute();
    }
    /**
     * @private
     * @param {?} urlPath
     * @param {?} menus
     * @return {?}
     */
    activateMenuByUrl(urlPath, menus) {
        if (menus) {
            return menus.some((/**
             * @param {?} menu
             * @return {?}
             */
            menu => {
                if (getFormattedLink(menu.link) === urlPath) {
                    this.activateMenuItem(menu);
                    return true;
                }
            }));
        }
    }
    /**
     * @private
     * @param {?} menu
     * @return {?}
     */
    activateMenuItem(menu) {
        this.activeMenuItem = menu;
        this.linkActive = getFormattedLink(menu.link);
        this.menuItemsService.sendToChildMenuClicked({ active: this.activeMenuItem, activatedByRoute: true });
    }
    /**
     * @private
     * @param {?} urlPath
     * @return {?}
     */
    checkActiveMenuByUrl(urlPath) {
        if (!this.linkActive || this.linkActive !== urlPath) {
            this.activateMenuByUrl(urlPath, this.menus);
        }
    }
    /**
     * @private
     * @param {?} menu
     * @return {?}
     */
    clickMenuItem(menu) {
        if (menu.action) {
            callFunction(menu.action, this.parentRef, menu);
        }
        if (menu.type === 'externalLink') {
            openExternalLink(menu.link);
        }
        else if (menu.type === 'internalLink') {
            this.activateMenuItem(menu);
        }
    }
    /**
     * @private
     * @return {?}
     */
    subscribeToRoute() {
        this.routeSubscription = this.router.events.subscribe((/**
         * @param {?} rounterEvent
         * @return {?}
         */
        rounterEvent => {
            if (rounterEvent instanceof NavigationEnd) {
                this.checkActiveMenuByUrl(this.location.path());
            }
        }));
    }
    /**
     * @private
     * @return {?}
     */
    subscribeToMenuItem() {
        this.itemSubscription =
            this.menuItemsService.receiveFromChildMenuClicked().subscribe((/**
             * @param {?} menu
             * @return {?}
             */
            (menu) => this.clickMenuItem(menu)));
    }
}
ThfMenuPanelComponent.decorators = [
    { type: Component, args: [{
                selector: 'thf-menu-panel',
                template: "<div class=\"thf-menu-panel\">\n  <div class=\"thf-menu-panel-logo-container\">\n    <a href=\"./\">\n      <div class=\"thf-menu-panel-logo\"></div>\n    </a>\n  </div>\n\n  <nav class=\"thf-menu-panel-container\">\n    <div class=\"thf-menu-panel-inner\">\n      <div *ngFor=\"let menu of menus, let menuIndex = index;\" class=\"thf-menu-panel-item-wrapper\">\n        <thf-menu-panel-item\n          [class.thf-menu-panel-item-first]=\"menuIndex === 0\"\n          [t-menu-item-internal]=\"menu\">\n        </thf-menu-panel-item>\n      </div>\n    </div>\n  </nav>\n</div>\n"
            }] }
];
/** @nocollapse */
ThfMenuPanelComponent.ctorParameters = () => [
    { type: ViewContainerRef },
    { type: Location },
    { type: ThfMenuPanelItemsService },
    { type: Router }
];
if (false) {
    /** @type {?} */
    ThfMenuPanelComponent.prototype.activeMenuItem;
    /** @type {?} */
    ThfMenuPanelComponent.prototype.linkActive;
    /** @type {?} */
    ThfMenuPanelComponent.prototype.parentRef;
    /**
     * @type {?}
     * @private
     */
    ThfMenuPanelComponent.prototype.routeSubscription;
    /**
     * @type {?}
     * @private
     */
    ThfMenuPanelComponent.prototype.itemSubscription;
    /**
     * @type {?}
     * @private
     */
    ThfMenuPanelComponent.prototype.location;
    /**
     * @type {?}
     * @private
     */
    ThfMenuPanelComponent.prototype.menuItemsService;
    /**
     * @type {?}
     * @private
     */
    ThfMenuPanelComponent.prototype.router;
}
