// @totvs/thf-templates/esm2015/lib/components/thf-page-dynamic-edit/thf-page-dynamic-edit.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, ViewChild } from '@angular/core';
import { ActivatedRoute, Router } from '@angular/router';
import { ThfDialogService, ThfDynamicFormComponent, ThfGridComponent, ThfNotificationService } from '@totvs/thf-ui';
import * as util from './../../utils/util';
import { ThfPageDynamicService } from './thf-page-dynamic.service';
/** @type {?} */
export const thfPageDynamicEditLiteralsDefault = {
    en: {
        cancelConfirmMessage: 'Are you sure you want to cancel this operation?',
        detailActionNew: 'New',
        pageActionCancel: 'Cancel',
        pageActionSave: 'Save',
        pageActionSaveNew: 'Save and new',
        registerNotFound: 'Register not found.',
        saveNewNotificationSuccessSave: 'Resource successfully saved.',
        saveNewNotificationSuccessUpdate: 'Resource successfully updated.',
        saveNewNotificationWarning: 'Form must be filled out correctly.',
        saveNotificationSuccessSave: 'Resource successfully saved.',
        saveNotificationSuccessUpdate: 'Resource successfully updated.',
        saveNotificationWarning: 'Form must be filled out correctly.',
    },
    es: {
        cancelConfirmMessage: 'Está seguro de que desea cancelar esta operación?',
        detailActionNew: 'Nuevo',
        pageActionCancel: 'Cancelar',
        pageActionSave: 'Guardar',
        pageActionSaveNew: 'Guardar y nuevo',
        registerNotFound: 'Registro no encontrado.',
        saveNewNotificationSuccessSave: 'Recurso salvo con éxito.',
        saveNewNotificationSuccessUpdate: 'Recurso actualizado con éxito.',
        saveNewNotificationWarning: 'El formulario debe llenarse correctamente.',
        saveNotificationSuccessSave: 'Recurso salvo con éxito.',
        saveNotificationSuccessUpdate: 'Recurso actualizado con éxito.',
        saveNotificationWarning: 'El formulario debe llenarse correctamente.',
    },
    pt: {
        cancelConfirmMessage: 'Tem certeza que deseja cancelar esta operação?',
        detailActionNew: 'Novo',
        pageActionCancel: 'Cancelar',
        pageActionSave: 'Salvar',
        pageActionSaveNew: 'Salvar e novo',
        registerNotFound: 'Registro não encontrado.',
        saveNewNotificationSuccessSave: 'Recurso salvo com sucesso.',
        saveNewNotificationSuccessUpdate: 'Recurso atualizado com sucesso.',
        saveNewNotificationWarning: 'Formulário precisa ser preenchido corretamente.',
        saveNotificationSuccessSave: 'Recurso salvo com sucesso.',
        saveNotificationSuccessUpdate: 'Recurso atualizado com sucesso.',
        saveNotificationWarning: 'Formulário precisa ser preenchido corretamente.',
    }
};
/**
 * \@description
 *
 * O `thf-page-dynamic-edit` é uma página que pode servir para editar ou criar novos registros,
 * o mesmo também suporta metadados conforme especificado na documentação.
 *
 * \@example
 *
 * <example name="thf-page-dynamic-edit-basic" title="Totvs Page Dynamic Edit Basic">
 *  <file name="sample-thf-page-dynamic-edit-basic/sample-thf-page-dynamic-edit-basic.component.html"> </file>
 *  <file name="sample-thf-page-dynamic-edit-basic/sample-thf-page-dynamic-edit-basic.component.ts"> </file>
 * </example>
 *
 * <example name="thf-page-dynamic-edit-user" title="Totvs Page Dynamic Edit - User">
 *  <file name="sample-thf-page-dynamic-edit-user/sample-thf-page-dynamic-edit-user.component.html"> </file>
 *  <file name="sample-thf-page-dynamic-edit-user/sample-thf-page-dynamic-edit-user.component.ts"> </file>
 * </example>
 *
 * <example name="thf-page-dynamic-edit-master-detail" title="Totvs Page Dynamic Edit - Master Detail">
 *  <file name="sample-thf-page-dynamic-edit-master-detail/sample-thf-page-dynamic-edit-master-detail.component.html"> </file>
 *  <file name="sample-thf-page-dynamic-edit-master-detail/sample-thf-page-dynamic-edit-master-detail.component.ts"> </file>
 * </example>
 */
export class ThfPageDynamicEditComponent {
    /**
     * @param {?} router
     * @param {?} activatedRoute
     * @param {?} thfNotification
     * @param {?} thfDialogService
     * @param {?} thfPageDynamicService
     */
    constructor(router, activatedRoute, thfNotification, thfDialogService, thfPageDynamicService) {
        this.router = router;
        this.activatedRoute = activatedRoute;
        this.thfNotification = thfNotification;
        this.thfDialogService = thfDialogService;
        this.thfPageDynamicService = thfPageDynamicService;
        this._actions = {};
        this._autoRouter = false;
        this._controlFields = [];
        this._detailFields = [];
        this._duplicates = [];
        this._fields = [];
        this._keys = [];
        this._pageActions = [];
        this.literals = Object.assign({}, thfPageDynamicEditLiteralsDefault[util.thfLocaleDefault], thfPageDynamicEditLiteralsDefault[util.browserLanguage()]);
        this.model = {};
        // beforeSave: return boolean
        // afterSave
        // beforeRemove: return boolean
        // afterRemove
        // beforeInsert: : return boolean
        this.detailActions = {};
        /**
         * Objeto com propriedades do breadcrumb.
         */
        this.breadcrumb = { items: [] };
    }
    /**
     * \@optional
     *
     * \@description
     *
     * Ações da página.
     * @param {?} value
     * @return {?}
     */
    set actions(value) {
        this._actions = this.isObject(value) ? value : {};
        this._pageActions = this.getPageActions(this._actions);
    }
    /**
     * \@todo Validar rotas na mão pois se existir uma rota '**' o catch do navigation não funciona.
     *
     * \@optional
     *
     * \@description
     *
     * Cria automaticamente as rotas de edição (novo/duplicate) e detalhes caso as ações
     * estejam definidas nas ações.
     *
     * > Para o correto funcionamento não pode haver nenhum rota coringa (`**`) especificada.
     *
     * \@default false
     * @param {?} value
     * @return {?}
     */
    set autoRouter(value) {
        this._autoRouter = util.convertToBoolean(value);
    }
    /**
     * @return {?}
     */
    get autoRouter() {
        return this._autoRouter;
    }
    /**
     * Lista dos campos usados na tabela e busca avançada.
     * @param {?} value
     * @return {?}
     */
    set fields(value) {
        this._fields = Array.isArray(value) ? [...value] : [];
        this._keys = this.getKeysByFields(this._fields);
        this._duplicates = this.getDuplicatesByFields(this._fields);
        this._controlFields = this.getControlFields(this._fields);
        this._detailFields = this.getDetailFields(this._fields);
    }
    /**
     * @return {?}
     */
    get fields() {
        return this._fields;
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        /** @type {?} */
        const paramId = this.activatedRoute.snapshot.params['id'];
        /** @type {?} */
        const duplicate = this.activatedRoute.snapshot.queryParams['duplicate'];
        if (this.activatedRoute.snapshot.data.serviceApi) {
            this.serviceApi = this.activatedRoute.snapshot.data.serviceApi;
            this.thfPageDynamicService.configServiceApi({ endpoint: this.serviceApi });
            this.loadMetadata(paramId, duplicate);
        }
        else {
            this.thfPageDynamicService.configServiceApi({ endpoint: this.serviceApi });
            this.loadData(paramId, duplicate);
        }
    }
    /**
     * @return {?}
     */
    get duplicates() {
        return [...this._duplicates];
    }
    /**
     * @return {?}
     */
    get keys() {
        return [...this._keys];
    }
    /**
     * @return {?}
     */
    get pageActions() {
        return [...this._pageActions];
    }
    /**
     * @return {?}
     */
    get controlFields() {
        return this._controlFields;
    }
    /**
     * @return {?}
     */
    get detailFields() {
        return this._detailFields;
    }
    /**
     * @private
     * @param {?} path
     * @return {?}
     */
    cancel(path) {
        if (this.dynamicForm && this.dynamicForm.form.dirty) {
            this.thfDialogService.confirm({
                message: this.literals.cancelConfirmMessage,
                title: this.literals.pageActionCancel,
                confirm: this.goBack.bind(this, path)
            });
        }
        else {
            this.goBack(path);
        }
    }
    /**
     * @private
     * @param {?} item
     * @return {?}
     */
    formatUniqueKey(item) {
        /** @type {?} */
        const keys = util.mapObjectByProperties(item, this.keys);
        return util.valuesFromObject(keys).join('|');
    }
    /**
     * @private
     * @param {?} path
     * @return {?}
     */
    goBack(path) {
        if (path) {
            this.router.navigate([path]);
        }
        else {
            window.history.back();
        }
    }
    /**
     * @private
     * @param {?} id
     * @param {?=} duplicate
     * @return {?}
     */
    loadData(id, duplicate) {
        if (!id) {
            try {
                this.model = JSON.parse(duplicate) || {};
            }
            catch (_a) {
                this.model = {};
            }
            return;
        }
        this.thfPageDynamicService.getResource(id).toPromise().then((/**
         * @param {?} response
         * @return {?}
         */
        response => {
            this.model = response;
        })).catch((/**
         * @return {?}
         */
        () => {
            this.model = undefined;
            this.actions = undefined;
            this._pageActions = [];
        }));
    }
    /**
     * @private
     * @param {?} paramId
     * @param {?} duplicate
     * @return {?}
     */
    loadMetadata(paramId, duplicate) {
        /** @type {?} */
        const typeMetadata = paramId ? 'edit' : 'create';
        this.thfPageDynamicService.getMetadata(typeMetadata).toPromise().then((/**
         * @param {?} response
         * @return {?}
         */
        response => {
            this.autoRouter = response.autoRouter;
            this.actions = response.actions || {};
            this.breadcrumb = response.breadcrumb || { items: [] };
            this.fields = response.fields || [];
            this.title = response.title;
            this.loadData(paramId, duplicate);
        }));
    }
    /**
     * @private
     * @param {?} path
     * @return {?}
     */
    navigateTo(path) {
        if (path) {
            /** @type {?} */
            const url = this.resolveUrl(this.model, path);
            this.router.navigate([url]);
        }
        else {
            window.history.back();
        }
    }
    /**
     * @private
     * @return {?}
     */
    detailActionNew() {
        this.gridDetail.insertRow();
    }
    /**
     * @private
     * @param {?} item
     * @param {?} path
     * @return {?}
     */
    resolveUrl(item, path) {
        /** @type {?} */
        const uniqueKey = this.formatUniqueKey(item);
        return path.replace(/:id/g, uniqueKey);
    }
    /**
     * @private
     * @param {?} path
     * @return {?}
     */
    save(path) {
        if (this.dynamicForm.form.invalid) {
            this.thfNotification.warning(this.literals.saveNotificationWarning);
            return;
        }
        /** @type {?} */
        const paramId = this.activatedRoute.snapshot.params['id'];
        /** @type {?} */
        const saveOperation = paramId
            ? this.thfPageDynamicService.updateResource(paramId, this.model)
            : this.thfPageDynamicService.createResource(this.model);
        /** @type {?} */
        const msgSucess = paramId ? this.literals.saveNotificationSuccessUpdate : this.literals.saveNotificationSuccessSave;
        saveOperation.toPromise().then((/**
         * @return {?}
         */
        () => {
            this.thfNotification.success(msgSucess);
            this.navigateTo(path);
        }));
    }
    /**
     * @private
     * @param {?} path
     * @return {?}
     */
    saveNew(path) {
        if (this.dynamicForm.form.invalid) {
            this.thfNotification.warning(this.literals.saveNewNotificationWarning);
            return;
        }
        /** @type {?} */
        const paramId = this.activatedRoute.snapshot.params['id'];
        if (paramId) {
            this.thfPageDynamicService.updateResource(paramId, this.model)
                .toPromise().then((/**
             * @return {?}
             */
            () => {
                this.thfNotification.success(this.literals.saveNewNotificationSuccessUpdate);
                this.navigateTo(path);
            }));
        }
        else {
            this.thfPageDynamicService.createResource(this.model)
                .toPromise().then((/**
             * @return {?}
             */
            () => {
                this.thfNotification.success(this.literals.saveNewNotificationSuccessSave);
                this.model = {};
                this.dynamicForm.form.reset();
            }));
        }
    }
    /**
     * @private
     * @param {?=} fields
     * @return {?}
     */
    getKeysByFields(fields = []) {
        return fields.filter((/**
         * @param {?} field
         * @return {?}
         */
        field => field.key === true)).map((/**
         * @param {?} field
         * @return {?}
         */
        field => field.property));
    }
    /**
     * @private
     * @param {?=} fields
     * @return {?}
     */
    getControlFields(fields = []) {
        return fields.filter((/**
         * @param {?} field
         * @return {?}
         */
        field => field.type !== 'detail'));
    }
    /**
     * @private
     * @param {?=} fields
     * @return {?}
     */
    getDetailFields(fields = []) {
        return fields.filter((/**
         * @param {?} field
         * @return {?}
         */
        field => field.type === 'detail'));
    }
    /**
     * @private
     * @param {?=} fields
     * @return {?}
     */
    getDuplicatesByFields(fields = []) {
        return fields.filter((/**
         * @param {?} field
         * @return {?}
         */
        field => field.duplicate === true)).map((/**
         * @param {?} field
         * @return {?}
         */
        field => field.property));
    }
    /**
     * @private
     * @param {?=} actions
     * @return {?}
     */
    getPageActions(actions = {}) {
        /** @type {?} */
        const pageActions = [{ label: this.literals.pageActionSave, action: this.save.bind(this, actions.save) }];
        if (actions.saveNew) {
            pageActions.push({ label: this.literals.pageActionSaveNew, action: this.saveNew.bind(this, actions.saveNew) });
        }
        if (actions.cancel === undefined || actions.cancel) {
            pageActions.push({ label: this.literals.pageActionCancel, action: this.cancel.bind(this, actions.cancel) });
        }
        return pageActions;
    }
    /**
     * @private
     * @param {?} value
     * @return {?}
     */
    isObject(value) {
        return !!value && typeof value === 'object' && !Array.isArray(value);
    }
}
ThfPageDynamicEditComponent.decorators = [
    { type: Component, args: [{
                selector: 'thf-page-dynamic-edit',
                template: "<thf-page-default\n  [t-actions]=\"pageActions\"\n  [t-breadcrumb]=\"breadcrumb\"\n  [t-title]=\"title\">\n\n  <ng-container *ngIf=\"model; then formFieldsTemplate; else registerNotFoundTemplate\">\n  </ng-container>\n\n</thf-page-default>\n\n<ng-template #registerNotFoundTemplate>\n\n  <thf-widget>\n    <span class=\"thf-icon thf-icon-info\"></span>\n    <span class=\"thf-font-text-large\">\n      {{ literals.registerNotFound }}\n    </span>\n  </thf-widget>\n\n</ng-template>\n\n<ng-template #formFieldsTemplate>\n\n  <thf-dynamic-form #dynamicForm [t-fields]=\"controlFields\" [t-value]=\"model\">\n  </thf-dynamic-form>\n\n  <div *ngIf=\"detailFields.length > 0\" class=\"thf-sm-12\">\n    <thf-divider [t-label]=\"detailFields[0].divider\"></thf-divider>\n\n    <div class=\"thf-row thf-mb-2\">\n      <thf-button [t-label]=\"literals.detailActionNew\" (t-click)=\"detailActionNew()\" t-small></thf-button>\n    </div>\n\n    <thf-grid #gridDetail\n      [t-row-actions]=\"detailActions\"\n      [t-columns]=\"detailFields[0].detail.columns\"\n      [t-data]=\"model[detailFields[0].property]\">\n    </thf-grid>\n  </div>\n\n</ng-template>\n",
                providers: [ThfPageDynamicService]
            }] }
];
/** @nocollapse */
ThfPageDynamicEditComponent.ctorParameters = () => [
    { type: Router },
    { type: ActivatedRoute },
    { type: ThfNotificationService },
    { type: ThfDialogService },
    { type: ThfPageDynamicService }
];
ThfPageDynamicEditComponent.propDecorators = {
    actions: [{ type: Input, args: ['t-actions',] }],
    autoRouter: [{ type: Input, args: ['t-auto-router',] }],
    breadcrumb: [{ type: Input, args: ['t-breadcrumb',] }],
    fields: [{ type: Input, args: ['t-fields',] }],
    serviceApi: [{ type: Input, args: ['t-service-api',] }],
    title: [{ type: Input, args: ['t-title',] }],
    dynamicForm: [{ type: ViewChild, args: ['dynamicForm', { static: true },] }],
    gridDetail: [{ type: ViewChild, args: ['gridDetail', { static: true },] }]
};
if (false) {
    /**
     * @type {?}
     * @private
     */
    ThfPageDynamicEditComponent.prototype._actions;
    /**
     * @type {?}
     * @private
     */
    ThfPageDynamicEditComponent.prototype._autoRouter;
    /**
     * @type {?}
     * @private
     */
    ThfPageDynamicEditComponent.prototype._controlFields;
    /**
     * @type {?}
     * @private
     */
    ThfPageDynamicEditComponent.prototype._detailFields;
    /**
     * @type {?}
     * @private
     */
    ThfPageDynamicEditComponent.prototype._duplicates;
    /**
     * @type {?}
     * @private
     */
    ThfPageDynamicEditComponent.prototype._fields;
    /**
     * @type {?}
     * @private
     */
    ThfPageDynamicEditComponent.prototype._keys;
    /**
     * @type {?}
     * @private
     */
    ThfPageDynamicEditComponent.prototype._pageActions;
    /** @type {?} */
    ThfPageDynamicEditComponent.prototype.literals;
    /** @type {?} */
    ThfPageDynamicEditComponent.prototype.model;
    /** @type {?} */
    ThfPageDynamicEditComponent.prototype.detailActions;
    /**
     * Objeto com propriedades do breadcrumb.
     * @type {?}
     */
    ThfPageDynamicEditComponent.prototype.breadcrumb;
    /**
     * \@description
     *
     * Endpoint usado pelo template para requisição do recurso que serão exibido para edição.
     *
     * Para as ações de `save` e `saveNew`, será feito uma requisição de criação nesse mesmo endpoint passando os valores
     * preenchidos pelo usuário via payload.
     *
     * > `POST {end-point}`
     *
     * ```
     *  <thf-page-dynamic-edit
     *    [t-actions]="{ save: '/', saveNew: 'new' }"
     *    [t-fields]="[ { property: 'name' }, { property: 'city' } ]"
     *    t-service="/api/thf-samples/v1/people"
     *    ...>
     *  </thf-page-dynamic-edit>
     * ```
     *
     * Resquisição disparada, onde a propriedade `name` e `city` foram preenchidas:
     *
     * ```
     *  POST /api/thf-samples/v1/people HTTP/1.1
     *  Host: localhost:4000
     *  Connection: keep-alive
     *  Accept: application/json, text/plain
     *  ...
     * ```
     *
     * Request payload:
     *
     * ```
     * { "name": "Fulano", "city": "Smallville" }
     * ```
     * > Caso esteja usando metadados com o template, será disparado uma requisição na inicialização do template para buscar
     * > os metadados da página passando o tipo do metadado esperado e a versão cacheada pelo browser.
     * >
     * > `GET {end-point}/metadata?type=create&version={version}`
     *
     * Caso queira que o template carregue um recurso já existente, deve-se ser incluído um parametro na rota chamado `id`.
     *
     * Exemplo de configuração de rota:
     *
     * ```
     *  RouterModule.forRoot([
     *    ...
     *    { path: 'edit/:id', component: PersonEditComponent },
     *    ...
     *  ],
     * ```
     *
     * Baseado nisso, na inicialização do template, será disparado uma requisição para buscar o recurso que será editado.
     *
     * > `GET {end-point}/{id}`
     *
     * Nos métodos de `save` e `saveNew`, ao invés de um `POST`, será disparado um `PUT`.
     *
     * Resquisição disparada, onde a propriedade `name` e `city` foram preenchidas / atualizadas, e o `id` da url é 2:
     *
     * ```
     *  PUT /api/thf-samples/v1/people/2 HTTP/1.1
     *  Host: localhost:4000
     *  Connection: keep-alive
     *  Accept: application/json, text/plain
     *  ...
     * ```
     *
     * Request payload:
     *
     * ```
     * { "name": "Fulano", "city": "Metropolis" }
     * ```
     *
     * > Caso esteja usando metadados com o template, será disparado uma requisição na inicialização do template para buscar
     * > os metadados da página passando o tipo do metadado esperado e a versão cacheada pelo browser.
     * >
     * > `GET {end-point}/metadata?type=edit&version={version}`
     * @type {?}
     */
    ThfPageDynamicEditComponent.prototype.serviceApi;
    /**
     * Título da página.
     * @type {?}
     */
    ThfPageDynamicEditComponent.prototype.title;
    /** @type {?} */
    ThfPageDynamicEditComponent.prototype.dynamicForm;
    /** @type {?} */
    ThfPageDynamicEditComponent.prototype.gridDetail;
    /**
     * @type {?}
     * @private
     */
    ThfPageDynamicEditComponent.prototype.router;
    /**
     * @type {?}
     * @private
     */
    ThfPageDynamicEditComponent.prototype.activatedRoute;
    /**
     * @type {?}
     * @private
     */
    ThfPageDynamicEditComponent.prototype.thfNotification;
    /**
     * @type {?}
     * @private
     */
    ThfPageDynamicEditComponent.prototype.thfDialogService;
    /**
     * @type {?}
     * @private
     */
    ThfPageDynamicEditComponent.prototype.thfPageDynamicService;
}
