// @totvs/thf-templates/esm5/lib/components/thf-page-job-scheduler/thf-page-job-scheduler.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { ActivatedRoute, Router } from '@angular/router';
import { Component, ViewChild, ViewEncapsulation } from '@angular/core';
import { ThfDialogService, ThfNotificationService, ThfStepperStatus } from '@totvs/thf-ui';
import * as util from './../../utils/util';
import { ThfPageJobSchedulerInternal } from './thf-page-job-scheduler-internal';
import { ThfPageJobSchedulerBaseComponent } from './thf-page-job-scheduler-base.component';
import { thfPageJobSchedulerLiteralsDefault } from './thf-page-job-scheduler-literals';
import { ThfPageJobSchedulerLookupService } from './thf-page-job-scheduler-lookup.service';
import { ThfPageJobSchedulerService } from './thf-page-job-scheduler.service';
/**
 * \@docsExtends ThfPageJobSchedulerBaseComponent
 *
 * \@example
 *
 * <example name="thf-page-job-scheduler-basic" title="Totvs Page Job Scheduler Basic">
 *  <file name="sample-thf-page-job-scheduler-basic/sample-thf-page-job-scheduler-basic.component.html"> </file>
 *  <file name="sample-thf-page-job-scheduler-basic/sample-thf-page-job-scheduler-basic.component.ts"> </file>
 * </example>
 *
 */
var ThfPageJobSchedulerComponent = /** @class */ (function (_super) {
    tslib_1.__extends(ThfPageJobSchedulerComponent, _super);
    function ThfPageJobSchedulerComponent(thfPageDynamicLookupService, activatedRoute, thfDialogService, thfNotification, router, thfPageJobSchedulerService) {
        var _this = _super.call(this, thfPageJobSchedulerService) || this;
        _this.thfPageDynamicLookupService = thfPageDynamicLookupService;
        _this.activatedRoute = activatedRoute;
        _this.thfDialogService = thfDialogService;
        _this.thfNotification = thfNotification;
        _this.router = router;
        _this.isEdit = false;
        _this.literals = tslib_1.__assign({}, thfPageJobSchedulerLiteralsDefault[util.thfLocaleDefault], thfPageJobSchedulerLiteralsDefault[util.browserLanguage()]);
        _this.parameters = [];
        _this.step = 1;
        _this.backPageAction = {
            label: _this.literals.back,
            action: _this.nextStepOperation.bind(_this, 'back'),
            disabled: _this.isDisabledBack.bind(_this)
        };
        _this.concludePageActions = [
            {
                label: _this.literals.conclude,
                action: _this.confirmJobScheduler.bind(_this)
            },
            tslib_1.__assign({}, _this.backPageAction)
        ];
        _this.nextPageActions = [
            {
                label: _this.literals.next,
                action: _this.nextStepOperation.bind(_this, 'next'),
                disabled: _this.isDisabledAdvance.bind(_this)
            },
            tslib_1.__assign({}, _this.backPageAction)
        ];
        _this.jobSchedulerActions = tslib_1.__spread(_this.nextPageActions);
        _this.steps = [
            { label: _this.literals.scheduling },
            { label: _this.literals.parameterization },
            { label: _this.literals.conclude }
        ];
        return _this;
    }
    Object.defineProperty(ThfPageJobSchedulerComponent.prototype, "stepperOrientation", {
        get: /**
         * @return {?}
         */
        function () {
            return window.innerWidth > 481 && window.innerWidth < 960 ? 'horizontal' : 'vertical';
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    ThfPageJobSchedulerComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var paramId = this.activatedRoute.snapshot.params['id'];
        this.isEdit = !!paramId;
        this.thfPageJobSchedulerService.configServiceApi({ endpoint: this.serviceApi });
        this.loadData(paramId);
    };
    /**
     * @param {?} currentStep
     * @param {?} nextStep
     * @return {?}
     */
    ThfPageJobSchedulerComponent.prototype.changePageActionsBySteps = /**
     * @param {?} currentStep
     * @param {?} nextStep
     * @return {?}
     */
    function (currentStep, nextStep) {
        /** @type {?} */
        var stepsLength = this.steps.length;
        if (nextStep === stepsLength) {
            this.jobSchedulerActions = tslib_1.__spread(this.concludePageActions);
        }
        else if (currentStep === stepsLength && nextStep < currentStep) {
            this.jobSchedulerActions = tslib_1.__spread(this.nextPageActions);
        }
    };
    /**
     * @param {?} stepNumber
     * @return {?}
     */
    ThfPageJobSchedulerComponent.prototype.nextStep = /**
     * @param {?} stepNumber
     * @return {?}
     */
    function (stepNumber) {
        if (stepNumber > 1 && this.schedulerExecution.form.invalid) {
            this.markAsDirtyInvalidControls(this.schedulerExecution.form.controls);
            return;
        }
        if (stepNumber > 2 && (this.schedulerParameters && this.schedulerParameters.form && this.schedulerParameters.form.invalid)) {
            this.markAsDirtyInvalidControls(this.schedulerParameters.form.controls);
            return;
        }
        this.changePageActionsBySteps(this.step, stepNumber);
        /** @type {?} */
        var steps = this.steps[this.step - 1];
        this.step = stepNumber;
        if (steps) {
            steps.status = ThfStepperStatus.Done;
        }
    };
    /**
     * @param {?} process
     * @return {?}
     */
    ThfPageJobSchedulerComponent.prototype.onChangeProcess = /**
     * @param {?} process
     * @return {?}
     */
    function (process) {
        if (process.existAPI && process.processId) {
            this.getParametersByProcess(process.processId);
            if (!this.isEdit) {
                this.model.executionParameter = {};
            }
            return;
        }
    };
    /**
     * @private
     * @return {?}
     */
    ThfPageJobSchedulerComponent.prototype.confirmJobScheduler = /**
     * @private
     * @return {?}
     */
    function () {
        var _this = this;
        /** @type {?} */
        var paramId = this.activatedRoute.snapshot.params['id'];
        /** @type {?} */
        var confirmMessage = paramId ? this.literals.confirmUpdateMessage : this.literals.confirmSaveMessage;
        this.thfDialogService.confirm({
            title: this.literals.confirmation,
            message: confirmMessage,
            confirm: (/**
             * @return {?}
             */
            function () {
                /** @type {?} */
                var model = Object.assign({}, _this.model);
                _this.save(model, paramId);
            })
        });
    };
    /**
     * @private
     * @param {?} msgSuccess
     * @param {?} saveOperation
     * @return {?}
     */
    ThfPageJobSchedulerComponent.prototype.emitSuccessMessage = /**
     * @private
     * @param {?} msgSuccess
     * @param {?} saveOperation
     * @return {?}
     */
    function (msgSuccess, saveOperation) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, saveOperation.toPromise()];
                    case 1:
                        _a.sent();
                        this.thfNotification.success(msgSuccess);
                        this.resetJobSchedulerForm();
                        return [2 /*return*/];
                }
            });
        });
    };
    /**
     * @private
     * @param {?} process
     * @return {?}
     */
    ThfPageJobSchedulerComponent.prototype.getParametersByProcess = /**
     * @private
     * @param {?} process
     * @return {?}
     */
    function (process) {
        var _this = this;
        this.thfPageJobSchedulerService.getParametersByProcess(process).subscribe((/**
         * @param {?} parameters
         * @return {?}
         */
        function (parameters) {
            _this.parameters = parameters;
        }));
    };
    /**
     * @private
     * @return {?}
     */
    ThfPageJobSchedulerComponent.prototype.isDisabledAdvance = /**
     * @private
     * @return {?}
     */
    function () {
        return this.schedulerExecution ? this.schedulerExecution.form.invalid : false;
    };
    /**
     * @private
     * @return {?}
     */
    ThfPageJobSchedulerComponent.prototype.isDisabledBack = /**
     * @private
     * @return {?}
     */
    function () {
        return this.step === 1;
    };
    /**
     * @private
     * @param {?=} operation
     * @return {?}
     */
    ThfPageJobSchedulerComponent.prototype.nextStepOperation = /**
     * @private
     * @param {?=} operation
     * @return {?}
     */
    function (operation) {
        /** @type {?} */
        var stepNumber = operation === 'back' ? this.step - 1 : this.step + 1;
        this.nextStep(stepNumber);
    };
    /**
     * @private
     * @return {?}
     */
    ThfPageJobSchedulerComponent.prototype.resetJobSchedulerForm = /**
     * @private
     * @return {?}
     */
    function () {
        var _this = this;
        this.schedulerExecution.form.reset();
        // radiogroup não estava atribuindo novo valor, fica vermelho sem o timetout.
        setTimeout((/**
         * @return {?}
         */
        function () {
            _this.model = new ThfPageJobSchedulerInternal();
            _this.step = 1;
            _this.steps.forEach((/**
             * @param {?} step
             * @return {?}
             */
            function (step) {
                step.status = ThfStepperStatus.Default;
            }));
            _this.jobSchedulerActions = tslib_1.__spread(_this.nextPageActions);
        }));
    };
    /**
     * @private
     * @param {?} model
     * @param {?} paramId
     * @return {?}
     */
    ThfPageJobSchedulerComponent.prototype.save = /**
     * @private
     * @param {?} model
     * @param {?} paramId
     * @return {?}
     */
    function (model, paramId) {
        /** @type {?} */
        var saveOperation = paramId
            ? this.thfPageJobSchedulerService.updateResource(paramId, model)
            : this.thfPageJobSchedulerService.createResource(model);
        /** @type {?} */
        var msgSuccess = paramId ? this.literals.saveNotificationSuccessUpdate : this.literals.saveNotificationSuccessSave;
        this.emitSuccessMessage(msgSuccess, saveOperation);
    };
    ThfPageJobSchedulerComponent.decorators = [
        { type: Component, args: [{
                    selector: 'thf-page-job-scheduler',
                    template: "<thf-page-default\n  [t-actions]=\"jobSchedulerActions\"\n  [t-breadcrumb]=\"breadcrumb\"\n  [t-title]=\"title\">\n\n  <div class=\"thf-row\">\n\n    <thf-stepper\n      class=\"thf-lg-3 thf-xl-2\"\n      t-sequential=\"true\"\n      [t-orientation]=\"stepperOrientation\"\n      [t-step]=\"step\"\n      [t-steps]=\"steps\"\n      (t-change-step)=\"nextStep($event)\">\n    </thf-stepper>\n\n    <thf-container class=\"thf-lg-8 thf-xl-6\">\n      <form #formScheduler=\"ngForm\">\n\n        <thf-page-job-scheduler-execution [hidden]=\"step !== 1\"\n          #schedulerExecution\n          class=\"thf-md-12\"\n          [t-is-edit]=\"isEdit\"\n          [t-literals]=\"literals\"\n          [t-value]=\"model\"\n          (t-change-process)=\"onChangeProcess($event)\">\n        </thf-page-job-scheduler-execution>\n\n        <thf-page-job-scheduler-parameters *ngIf=\"step === 2\"\n          #schedulerParameters\n          class=\"thf-md-12\"\n          [t-literals]=\"literals\"\n          [t-parameters]=\"parameters || []\"\n          [(t-value)]=\"model.executionParameter\">\n        </thf-page-job-scheduler-parameters>\n\n        <thf-page-job-scheduler-summary *ngIf=\"step === 3\"\n          class=\"thf-md-12\"\n          [t-literals]=\"literals\"\n          [t-parameters]=\"parameters\"\n          [t-value]=\"model\">\n        </thf-page-job-scheduler-summary>\n      </form>\n    </thf-container>\n  </div>\n\n</thf-page-default>\n",
                    encapsulation: ViewEncapsulation.None,
                    styles: ["\n    thf-container .thf-container {\n      overflow-y: unset;\n    }\n  "]
                }] }
    ];
    /** @nocollapse */
    ThfPageJobSchedulerComponent.ctorParameters = function () { return [
        { type: ThfPageJobSchedulerLookupService },
        { type: ActivatedRoute },
        { type: ThfDialogService },
        { type: ThfNotificationService },
        { type: Router },
        { type: ThfPageJobSchedulerService }
    ]; };
    ThfPageJobSchedulerComponent.propDecorators = {
        schedulerExecution: [{ type: ViewChild, args: ['schedulerExecution', { static: true },] }],
        schedulerParameters: [{ type: ViewChild, args: ['schedulerParameters', { static: true },] }]
    };
    return ThfPageJobSchedulerComponent;
}(ThfPageJobSchedulerBaseComponent));
export { ThfPageJobSchedulerComponent };
if (false) {
    /** @type {?} */
    ThfPageJobSchedulerComponent.prototype.isEdit;
    /** @type {?} */
    ThfPageJobSchedulerComponent.prototype.literals;
    /** @type {?} */
    ThfPageJobSchedulerComponent.prototype.parameters;
    /** @type {?} */
    ThfPageJobSchedulerComponent.prototype.saveOperation;
    /** @type {?} */
    ThfPageJobSchedulerComponent.prototype.step;
    /**
     * @type {?}
     * @private
     */
    ThfPageJobSchedulerComponent.prototype.backPageAction;
    /**
     * @type {?}
     * @private
     */
    ThfPageJobSchedulerComponent.prototype.concludePageActions;
    /**
     * @type {?}
     * @private
     */
    ThfPageJobSchedulerComponent.prototype.nextPageActions;
    /** @type {?} */
    ThfPageJobSchedulerComponent.prototype.jobSchedulerActions;
    /** @type {?} */
    ThfPageJobSchedulerComponent.prototype.steps;
    /** @type {?} */
    ThfPageJobSchedulerComponent.prototype.schedulerExecution;
    /** @type {?} */
    ThfPageJobSchedulerComponent.prototype.schedulerParameters;
    /** @type {?} */
    ThfPageJobSchedulerComponent.prototype.thfPageDynamicLookupService;
    /**
     * @type {?}
     * @private
     */
    ThfPageJobSchedulerComponent.prototype.activatedRoute;
    /**
     * @type {?}
     * @private
     */
    ThfPageJobSchedulerComponent.prototype.thfDialogService;
    /**
     * @type {?}
     * @private
     */
    ThfPageJobSchedulerComponent.prototype.thfNotification;
    /**
     * @type {?}
     * @private
     */
    ThfPageJobSchedulerComponent.prototype.router;
}
