// @totvs/thf-ui/esm2015/lib/components/thf-navbar/thf-navbar-items/thf-navbar-items.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ElementRef, Input, QueryList, ViewChild, ViewChildren } from '@angular/core';
import { NavigationCancel, NavigationEnd, Router } from '@angular/router';
import { ThfNavbarItemComponent } from './thf-navbar-item/thf-navbar-item.component';
export class ThfNavbarItemsComponent {
    /**
     * @param {?} router
     */
    constructor(router) {
        this.router = router;
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        this.routeSubscription.unsubscribe();
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this.subscribeToRoute();
    }
    /**
     * @param {?} item
     * @return {?}
     */
    selectItem(item) {
        this.selectedItem = item;
    }
    /**
     * @private
     * @param {?} urlRouter
     * @return {?}
     */
    checkActiveItemByUrl(urlRouter) {
        this.selectedItem = this.items.find((/**
         * @param {?} item
         * @return {?}
         */
        item => item.link === urlRouter));
    }
    /**
     * @private
     * @return {?}
     */
    checkRouterChildrenFragments() {
        /** @type {?} */
        const childrenPrimary = this.router.parseUrl(this.router.url).root.children['primary'];
        return childrenPrimary ? `/${childrenPrimary.segments.map((/**
         * @param {?} it
         * @return {?}
         */
        it => it.path)).join('/')}` : '';
    }
    /**
     * @private
     * @return {?}
     */
    subscribeToRoute() {
        this.routeSubscription = this.router.events.subscribe((/**
         * @param {?} val
         * @return {?}
         */
        val => {
            if (val instanceof NavigationEnd || val instanceof NavigationCancel) {
                /** @type {?} */
                const urlRouter = this.checkRouterChildrenFragments();
                this.checkActiveItemByUrl(urlRouter);
            }
        }));
    }
}
ThfNavbarItemsComponent.decorators = [
    { type: Component, args: [{
                selector: 'thf-navbar-items',
                template: "<nav>\n  <ul class=\"thf-navbar-items-container\" #navbarItemsContainer>\n    <li *ngFor=\"let item of items\"\n      class=\"thf-navbar-item\"\n      [class.thf-navbar-item-selected]=\"selectedItem === item\">\n      <thf-navbar-item\n        [t-action]=\"item.action\"\n        [t-clickable]=\"selectedItem !== item\"\n        [t-label]=\"item.label\"\n        [t-link]=\"item.link\"\n        (t-click)=\"selectItem(item)\">\n      </thf-navbar-item>\n    </li>\n  </ul>\n</nav>\n"
            }] }
];
/** @nocollapse */
ThfNavbarItemsComponent.ctorParameters = () => [
    { type: Router }
];
ThfNavbarItemsComponent.propDecorators = {
    navbarItemsContainer: [{ type: ViewChild, args: ['navbarItemsContainer', { read: ElementRef, static: true },] }],
    allNavbarItems: [{ type: ViewChildren, args: [ThfNavbarItemComponent, { read: ElementRef },] }],
    items: [{ type: Input, args: ['t-items',] }]
};
if (false) {
    /** @type {?} */
    ThfNavbarItemsComponent.prototype.selectedItem;
    /**
     * @type {?}
     * @private
     */
    ThfNavbarItemsComponent.prototype.routeSubscription;
    /** @type {?} */
    ThfNavbarItemsComponent.prototype.navbarItemsContainer;
    /** @type {?} */
    ThfNavbarItemsComponent.prototype.allNavbarItems;
    /** @type {?} */
    ThfNavbarItemsComponent.prototype.items;
    /**
     * @type {?}
     * @private
     */
    ThfNavbarItemsComponent.prototype.router;
}
