// @totvs/thf-ui/esm5/lib/components/thf-dropdown/thf-dropdown-base.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Input } from '@angular/core';
import { convertToBoolean } from './../../utils/util';
/**
 * \@description
 *
 * O componente `thf-dropdown` pode ser utilizado como um agrupador de ações e / ou opções.
 *
 * > Caso não haja configuração de rotas em sua aplicação, se faz necessário importar o `RouterModule`
 * no módulo principal para o correto funcionamento deste componente:
 *
 * ```
 * import { RouterModule } from '\@angular/router';
 *
 * \@NgModule({
 *   imports: [
 *     ...
 *     RouterModule.forRoot([]),
 *     ThfModule
 *   ],
 *   declarations: [
 *     AppComponent
 *   ],
 *   exports: [],
 *   providers: [],
 *   bootstrap: [
 *     AppComponent
 *   ]
 * })
 * export class AppModule { }
 * ```
 * > Para maiores dúvidas referente à configuração de rotas, acesse em nosso portal /Guias /Começando
 * [/Configurando as rotas do thf-menu](/guides/getting-started).
 */
var ThfDropdownBaseComponent = /** @class */ (function () {
    function ThfDropdownBaseComponent() {
        this._disabled = false;
        this.icon = 'thf-icon-arrow-down';
        this.open = false;
    }
    Object.defineProperty(ThfDropdownBaseComponent.prototype, "actions", {
        get: /**
         * @return {?}
         */
        function () {
            return this._actions;
        },
        /** Lista de ações que serão exibidas no componente. */
        set: /**
         * Lista de ações que serão exibidas no componente.
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._actions = Array.isArray(value) ? value : [];
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ThfDropdownBaseComponent.prototype, "disabled", {
        get: /**
         * @return {?}
         */
        function () {
            return this._disabled;
        },
        /**
         * @optional
         *
         * @description
         *
         * Desabilita o campo.
         *
         * @default `false`
         */
        set: /**
         * \@optional
         *
         * \@description
         *
         * Desabilita o campo.
         *
         * \@default `false`
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._disabled = convertToBoolean(value);
        },
        enumerable: true,
        configurable: true
    });
    ThfDropdownBaseComponent.propDecorators = {
        actions: [{ type: Input, args: ['t-actions',] }],
        disabled: [{ type: Input, args: ['t-disabled',] }],
        label: [{ type: Input, args: ['t-label',] }]
    };
    return ThfDropdownBaseComponent;
}());
export { ThfDropdownBaseComponent };
if (false) {
    /**
     * @type {?}
     * @private
     */
    ThfDropdownBaseComponent.prototype._actions;
    /**
     * @type {?}
     * @private
     */
    ThfDropdownBaseComponent.prototype._disabled;
    /** @type {?} */
    ThfDropdownBaseComponent.prototype.icon;
    /** @type {?} */
    ThfDropdownBaseComponent.prototype.open;
    /**
     * Adiciona um rótulo ao `dropdown`.
     * @type {?}
     */
    ThfDropdownBaseComponent.prototype.label;
}
