// @totvs/thf-ui/esm5/lib/components/thf-slide/thf-slide-circles/thf-slide-circles.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, EventEmitter, Input, Output } from '@angular/core';
/**
 * \@docsPrivate
 *
 * \@description
 *
 * Componente para o conjunto de círculos que faz a passagem de itens do `thf-slide`.
 */
var ThfSlideCirclesComponent = /** @class */ (function () {
    function ThfSlideCirclesComponent() {
        /**
         * Evento emitido ao clicar em um controle.
         */
        this.click = new EventEmitter();
    }
    ThfSlideCirclesComponent.decorators = [
        { type: Component, args: [{
                    selector: 'thf-slide-circles',
                    template: "<div class=\"thf-slide-circles\">\n  <ng-container *ngFor=\"let item of items; let index = index;\">\n    <div\n      class=\"thf-slide-circle\"\n      [class.thf-slide-active-circle]=\"index === currentSlideIndex\"\n      (click)=\"click.emit(index)\">\n    </div>\n  </ng-container>\n</div>\n"
                }] }
    ];
    ThfSlideCirclesComponent.propDecorators = {
        currentSlideIndex: [{ type: Input, args: ['t-current-slide-index',] }],
        items: [{ type: Input, args: ['t-items',] }],
        click: [{ type: Output, args: ['t-click',] }]
    };
    return ThfSlideCirclesComponent;
}());
export { ThfSlideCirclesComponent };
if (false) {
    /**
     * Index do slide atual.
     * @type {?}
     */
    ThfSlideCirclesComponent.prototype.currentSlideIndex;
    /**
     * Itens do slide.
     * @type {?}
     */
    ThfSlideCirclesComponent.prototype.items;
    /**
     * Evento emitido ao clicar em um controle.
     * @type {?}
     */
    ThfSlideCirclesComponent.prototype.click;
}
