// @totvs/thf-ui/esm2015/lib/components/thf-field/thf-upload/thf-upload-file.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ThfUploadStatus } from './thf-upload-status.enum';
// Classe responsável pelo arquivo do THF Upload
export class ThfUploadFile {
    /**
     * @param {?} file
     */
    constructor(file) {
        if (file) {
            this.name = file.name;
            this.extension = this.getExtension(file.name);
            this.size = file.size;
            this.rawFile = file;
            this.uid = this.generateUUID();
            this.status = ThfUploadStatus.None;
        }
    }
    /**
     * @private
     * @param {?} value
     * @return {?}
     */
    getExtension(value) {
        if (value) {
            /** @type {?} */
            const extension = value.substr(value.lastIndexOf('.'));
            return extension.toLowerCase();
        }
        return '';
    }
    /**
     * @private
     * @return {?}
     */
    generateUUID() {
        /**
         * @return {?}
         */
        function s4() {
            return Math.floor((1 + Math.random()) * 0x10000)
                .toString(16)
                .substring(1);
        }
        return s4() + s4() + '-' + s4() + '-' + s4() + '-' +
            s4() + '-' + s4() + s4() + s4();
    }
}
if (false) {
    /** @type {?} */
    ThfUploadFile.prototype.name;
    /** @type {?} */
    ThfUploadFile.prototype.rawFile;
    /** @type {?} */
    ThfUploadFile.prototype.status;
    /** @type {?} */
    ThfUploadFile.prototype.uid;
    /** @type {?} */
    ThfUploadFile.prototype.extension;
    /** @type {?} */
    ThfUploadFile.prototype.size;
}
