// @totvs/thf-templates/esm2015/lib/components/thf-page-background/thf-page-background.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, EventEmitter, Input, Output } from '@angular/core';
import { browserLanguage, convertToBoolean, isTypeof } from './../../utils/util';
/**
 * @docsPrivate
 *
 * @description
 *
 * Componente para definição de cor de fundo e dos logotipos primário e secundário para os templates
 * de `thf-page-login` e demais templates de login.
 */
export class ThfPageBackgroundComponent {
    constructor() {
        this._showSelectLanguage = false;
        this.selectLanguageOptions = [
            { label: 'English', value: 'en' },
            { label: 'Español', value: 'es' },
            { label: 'Português', value: 'pt' }
        ];
        /**
         * Evento disparado ao selecionar alguma opção no seletor de idiomas.
         * Para este evento será passado como parâmetro o valor de idioma selecionado.
         */
        this.selectedLanguage = new EventEmitter();
    }
    /**
     * Caminho para a logomarca localizada na parte superior.
     * @param {?} value
     * @return {?}
     */
    set logo(value) {
        this._logo = isTypeof(value, 'string') && value.trim() ? value : undefined;
    }
    /**
     * @return {?}
     */
    get logo() {
        return this._logo;
    }
    /**
     * \@optional
     *
     * \@description
     *
     * Caminho para a logomarca localizada no rodapé.
     * @param {?} value
     * @return {?}
     */
    set secondaryLogo(value) {
        this._secondaryLogo = isTypeof(value, 'string') && value.trim() ? value : undefined;
    }
    /**
     * @return {?}
     */
    get secondaryLogo() {
        return this._secondaryLogo;
    }
    /**
     * Define se o seletor de idiomas deve ser exibido.
     * @param {?} showSelectLanguage
     * @return {?}
     */
    set showSelectLanguage(showSelectLanguage) {
        this._showSelectLanguage = convertToBoolean(showSelectLanguage);
    }
    /**
     * @return {?}
     */
    get showSelectLanguage() {
        return this._showSelectLanguage;
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this.selectedLanguageOption = browserLanguage();
    }
    /**
     * @return {?}
     */
    onChangeLanguage() {
        this.selectedLanguage.emit(this.selectedLanguageOption);
    }
}
ThfPageBackgroundComponent.decorators = [
    { type: Component, args: [{
                selector: 'thf-page-background',
                template: "<div class=\"thf-page-login-container\">\n  <div class=\"thf-page-login-panel\">\n    <img *ngIf=\"logo\"\n      class=\"thf-page-background-main-logo\"\n      [class.thf-page-background-hide-logo]=\"hideLogo\"\n      alt=\"main-logo\"\n      [src]=\"logo\">\n\n    <div class=\"thf-page-login-body\">\n      <div class=\"thf-page-login-panel-content\">\n        <ng-content></ng-content>\n      </div>\n    </div>\n    <div class=\"thf-page-background-footer thf-sm-12\">\n\n      <thf-divider class=\"thf-page-background-footer-mobile-only\"></thf-divider>\n\n      <div class=\"thf-page-background-footer-content\">\n        <div *ngIf=\"showSelectLanguage\" class=\"thf-page-background-footer-select\">\n          <thf-select\n            name=\"selectedLanguageOption\"\n            [(ngModel)]=\"selectedLanguageOption\"\n            [t-options]=\"selectLanguageOptions\"\n            (t-change) = \"onChangeLanguage()\">\n          </thf-select>\n        </div>\n\n        <div class=\"thf-page-background-secondary-logo\"\n          [ngClass]=\"showSelectLanguage ? 'thf-page-background-secondary-logo-right' : 'thf-page-background-secondary-logo-centered'\">\n\n          <img *ngIf=\"secondaryLogo\"\n            class=\"thf-page-background-secondary-logo-image\"\n            alt=\"secondary-logo\"\n            [src]=\"secondaryLogo\">\n        </div>\n      </div>\n    </div>\n  </div>\n\n  <div *ngIf=\"background\"\n    [ngClass]=\"background ? 'thf-page-login-highlight-image' : 'thf-page-login-highlight-image-off'\"\n    [style.background-image]=\"'url(' + background + ')'\">\n    <div class=\"thf-page-login-highlight-text\">\n      <div class=\"thf-font-display\">{{ highlightInfo }}</div>\n    </div>\n  </div>\n</div>\n"
            }] }
];
ThfPageBackgroundComponent.propDecorators = {
    background: [{ type: Input, args: ['t-background',] }],
    hideLogo: [{ type: Input, args: ['t-hide-logo',] }],
    highlightInfo: [{ type: Input, args: ['t-highlight-info',] }],
    logo: [{ type: Input, args: ['t-logo',] }],
    secondaryLogo: [{ type: Input, args: ['t-secondary-logo',] }],
    showSelectLanguage: [{ type: Input, args: ['t-show-select-language',] }],
    selectedLanguage: [{ type: Output, args: ['t-selected-language',] }]
};
if (false) {
    /**
     * @type {?}
     * @private
     */
    ThfPageBackgroundComponent.prototype._logo;
    /**
     * @type {?}
     * @private
     */
    ThfPageBackgroundComponent.prototype._secondaryLogo;
    /**
     * @type {?}
     * @private
     */
    ThfPageBackgroundComponent.prototype._showSelectLanguage;
    /** @type {?} */
    ThfPageBackgroundComponent.prototype.selectedLanguageOption;
    /** @type {?} */
    ThfPageBackgroundComponent.prototype.selectLanguageOptions;
    /**
     * Insere uma imagem de destaque ao lado direito do container.
     * @type {?}
     */
    ThfPageBackgroundComponent.prototype.background;
    /**
     * Designa se o logotipo deve desaparecer em resoluções menores.
     * @type {?}
     */
    ThfPageBackgroundComponent.prototype.hideLogo;
    /**
     * Texto de destaque sobreposto à imagem de destaque. Essa opção é utilizada em conjunto com o atributo `t-background`.
     * @type {?}
     */
    ThfPageBackgroundComponent.prototype.highlightInfo;
    /**
     * Evento disparado ao selecionar alguma opção no seletor de idiomas.
     * Para este evento será passado como parâmetro o valor de idioma selecionado.
     * @type {?}
     */
    ThfPageBackgroundComponent.prototype.selectedLanguage;
}
