// @totvs/thf-ui/esm5/lib/components/thf-table/thf-table-column-link/thf-table-column-link.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input } from '@angular/core';
import { isExternalLink } from '../../../utils/util';
/**
 * \@docsPrivate
 *
 * \@description
 *
 * Componente responsável por exibir link nas colunas.
 */
var ThfTableColumnLinkComponent = /** @class */ (function () {
    function ThfTableColumnLinkComponent() {
    }
    Object.defineProperty(ThfTableColumnLinkComponent.prototype, "type", {
        get: /**
         * @return {?}
         */
        function () {
            if (this.disabled) {
                return 'disabled';
            }
            if (this.action) {
                return 'action';
            }
            if (isExternalLink(this.link)) {
                return 'externalLink';
            }
            return 'internalLink';
        },
        enumerable: true,
        configurable: true
    });
    ThfTableColumnLinkComponent.decorators = [
        { type: Component, args: [{
                    selector: 'thf-table-column-link',
                    template: "<ng-container [ngSwitch]=\"type\">\n  <p *ngSwitchCase=\"'disabled'\" class=\"thf-table-link-disabled\">{{ value }}</p>\n  <a *ngSwitchCase=\"'action'\" class=\"thf-table-link\" (click)=\"action(value, row)\">{{ value }}</a>\n  <a *ngSwitchCase=\"'externalLink'\" class=\"thf-table-link\" [href]=\"link\" target=\"_blank\">{{ value }}</a>\n  <a *ngSwitchCase=\"'internalLink'\" class=\"thf-table-link\" [routerLink]=\"link\">{{ value }}</a>\n</ng-container>\n"
                }] }
    ];
    ThfTableColumnLinkComponent.propDecorators = {
        action: [{ type: Input, args: ['t-action',] }],
        disabled: [{ type: Input, args: ['t-disabled',] }],
        link: [{ type: Input, args: ['t-link',] }],
        row: [{ type: Input, args: ['t-row',] }],
        value: [{ type: Input, args: ['t-value',] }]
    };
    return ThfTableColumnLinkComponent;
}());
export { ThfTableColumnLinkComponent };
if (false) {
    /** @type {?} */
    ThfTableColumnLinkComponent.prototype.action;
    /** @type {?} */
    ThfTableColumnLinkComponent.prototype.disabled;
    /** @type {?} */
    ThfTableColumnLinkComponent.prototype.link;
    /** @type {?} */
    ThfTableColumnLinkComponent.prototype.row;
    /** @type {?} */
    ThfTableColumnLinkComponent.prototype.value;
}
