// @totvs/thf-ui/esm5/lib/components/thf-toolbar/thf-toolbar-actions/thf-toolbar-actions.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input } from '@angular/core';
import { isTypeof } from '../../../utils/util';
import { ThfControlPositionService } from '../../../services/thf-control-position/thf-control-position.service';
/** @type {?} */
var thfToolbarActionsIconDefault = 'thf-icon-more';
/**
 * \@docsPrivate
 *
 * \@usedBy ThfToolbarComponent
 *
 * \@description
 *
 * O componente `thf-toolbar-actions` tem como objetivo receber uma lista de ações e um ícone que podem ser personalizados.
 */
var ThfToolbarActionsComponent = /** @class */ (function () {
    function ThfToolbarActionsComponent() {
        this._actionsIcon = thfToolbarActionsIconDefault;
    }
    Object.defineProperty(ThfToolbarActionsComponent.prototype, "actionsIcon", {
        get: /**
         * @return {?}
         */
        function () {
            return this._actionsIcon;
        },
        /** Define o ícone das ações. */
        set: /**
         * Define o ícone das ações.
         * @param {?} icon
         * @return {?}
         */
        function (icon) {
            this._actionsIcon = isTypeof(icon, 'string') ? icon : thfToolbarActionsIconDefault;
        },
        enumerable: true,
        configurable: true
    });
    ThfToolbarActionsComponent.decorators = [
        { type: Component, args: [{
                    selector: 'thf-toolbar-actions',
                    template: "<div class=\"thf-toolbar-actions thf-clickable\" (click)=\"popup.toggle()\">\n  <span #toolbarActions class=\"thf-icon {{ actionsIcon }} thf-toolbar-icon\"></span>\n</div>\n\n<thf-popup #popup\n  [t-actions]=\"actions\"\n  [t-target]=\"toolbarActions\">\n</thf-popup>\n",
                    providers: [ThfControlPositionService]
                }] }
    ];
    ThfToolbarActionsComponent.propDecorators = {
        actions: [{ type: Input, args: ['t-actions',] }],
        actionsIcon: [{ type: Input, args: ['t-actions-icon',] }]
    };
    return ThfToolbarActionsComponent;
}());
export { ThfToolbarActionsComponent };
if (false) {
    /**
     * @type {?}
     * @private
     */
    ThfToolbarActionsComponent.prototype._actionsIcon;
    /**
     * Define uma lista de ações.
     * @type {?}
     */
    ThfToolbarActionsComponent.prototype.actions;
}
