// @totvs/thf-ui/esm5/lib/components/thf-field/thf-radio-group/thf-radio-group.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, ElementRef, forwardRef, Input, IterableDiffers, ViewChild } from '@angular/core';
import { NG_VALIDATORS, NG_VALUE_ACCESSOR } from '@angular/forms';
import { removeDuplicatedOptions } from '../../../utils/util';
import { ThfRadioGroupBaseComponent } from './thf-radio-group-base.component';
/**
 * \@docsExtends ThfRadioGroupBaseComponent
 *
 * \@example
 *
 * <example name="thf-radio-group-basic" title="Totvs Radio Group Basic">
 *  <file name="sample-thf-radio-group-basic/sample-thf-radio-group-basic.component.html"> </file>
 *  <file name="sample-thf-radio-group-basic/sample-thf-radio-group-basic.component.ts"> </file>
 *  <file name="sample-thf-radio-group-basic/sample-thf-radio-group-basic.component.e2e-spec.ts"> </file>
 *  <file name="sample-thf-radio-group-basic/sample-thf-radio-group-basic.component.po.ts"> </file>
 * </example>
 *
 * <example name="thf-radio-group-labs" title="Totvs Radio Group Labs">
 *  <file name="sample-thf-radio-group-labs/sample-thf-radio-group-labs.component.html"> </file>
 *  <file name="sample-thf-radio-group-labs/sample-thf-radio-group-labs.component.ts"> </file>
 * </example>
 *
 * <example name="thf-radio-group-translator" title="Totvs Radio Group - Translator">
 *  <file name="sample-thf-radio-group-translator/sample-thf-radio-group-translator.component.html"> </file>
 *  <file name="sample-thf-radio-group-translator/sample-thf-radio-group-translator.component.ts"> </file>
 * </example>
 *
 * <example name="thf-radio-group-translator-reactive-form" title="Totvs Radio Group - Translator Reactive Form">
 *  <file name="sample-thf-radio-group-translator-reactive-form/sample-thf-radio-group-translator-reactive-form.component.html"> </file>
 *  <file name="sample-thf-radio-group-translator-reactive-form/sample-thf-radio-group-translator-reactive-form.component.ts"> </file>
 * </example>
 *
 */
var ThfRadioGroupComponent = /** @class */ (function (_super) {
    tslib_1.__extends(ThfRadioGroupComponent, _super);
    function ThfRadioGroupComponent(differs) {
        var _this = _super.call(this) || this;
        _this.differ = differs.find([]).create(null);
        return _this;
    }
    /**
     * @return {?}
     */
    ThfRadioGroupComponent.prototype.ngDoCheck = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var change = this.differ.diff(this.options);
        if (change) {
            removeDuplicatedOptions(this.options);
        }
    };
    /**
     * @param {?} value
     * @param {?} disabled
     * @return {?}
     */
    ThfRadioGroupComponent.prototype.eventClick = /**
     * @param {?} value
     * @param {?} disabled
     * @return {?}
     */
    function (value, disabled) {
        if (!disabled) {
            this.changeValue(value);
        }
    };
    /**
     * @param {?} value
     * @return {?}
     */
    ThfRadioGroupComponent.prototype.getElementByValue = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        return this.inputEl.nativeElement.querySelector("input[value='" + value + "']");
    };
    /**
     * @param {?} event
     * @param {?} value
     * @return {?}
     */
    ThfRadioGroupComponent.prototype.onKeyUp = /**
     * @param {?} event
     * @param {?} value
     * @return {?}
     */
    function (event, value) {
        /** @type {?} */
        var key = event.keyCode || event.which;
        if (this.isArrowKey(key)) {
            this.changeValue(value);
        }
    };
    /**
     * @private
     * @param {?} key
     * @return {?}
     */
    ThfRadioGroupComponent.prototype.isArrowKey = /**
     * @private
     * @param {?} key
     * @return {?}
     */
    function (key) {
        return key >= 37 && key <= 40;
    };
    ThfRadioGroupComponent.decorators = [
        { type: Component, args: [{
                    selector: 'thf-radio-group',
                    template: "<thf-field-container\n  [t-help]=\"help\"\n  [t-label]=\"label\"\n  [t-optional]=\"!required && optional\">\n\n  <div class=\"thf-row thf-pt-1 thf-pb-1\" #inp>\n    <div *ngFor=\"let option of options\"\n      class=\"thf-radio-group-item thf-md-{{ mdColumns }} thf-lg-{{ columns }}\">\n\n      <input #input\n        class=\"thf-radio-group-input\"\n        type=\"radio\"\n        [attr.name]=\"name\"\n        [checked]=\"value === option.value\"\n        [class.thf-radio-group-input-checked]=\"value === option.value\"\n        [disabled]=\"option.disabled === true || disabled\"\n        [readonly]=\"readonly\"\n        [required]=\"required\"\n        [value]=\"option.value\"\n        (keyup)=\"onKeyUp($event, option.value)\">\n        <label\n          class=\"thf-radio-group-label\"\n          [class.thf-clickable]=\"!input.disabled\"\n          [for]=\"name\"\n          (click)=\"eventClick(option.value, option.disabled === true || disabled)\">\n        {{ option.label }}\n      </label>\n    </div>\n  </div>\n  <thf-field-container-bottom></thf-field-container-bottom>\n</thf-field-container>\n",
                    providers: [
                        {
                            provide: NG_VALUE_ACCESSOR,
                            useExisting: forwardRef((/**
                             * @return {?}
                             */
                            function () { return ThfRadioGroupComponent; })),
                            multi: true,
                        },
                        {
                            provide: NG_VALIDATORS,
                            useExisting: forwardRef((/**
                             * @return {?}
                             */
                            function () { return ThfRadioGroupComponent; })),
                            multi: true,
                        }
                    ]
                }] }
    ];
    /** @nocollapse */
    ThfRadioGroupComponent.ctorParameters = function () { return [
        { type: IterableDiffers }
    ]; };
    ThfRadioGroupComponent.propDecorators = {
        label: [{ type: Input, args: ['t-label',] }],
        help: [{ type: Input, args: ['t-help',] }],
        inputEl: [{ type: ViewChild, args: ['inp', { read: ElementRef, static: true },] }]
    };
    return ThfRadioGroupComponent;
}(ThfRadioGroupBaseComponent));
export { ThfRadioGroupComponent };
if (false) {
    /**
     * Label do campo.
     * @type {?}
     */
    ThfRadioGroupComponent.prototype.label;
    /**
     * Texto de apoio do campo.
     * @type {?}
     */
    ThfRadioGroupComponent.prototype.help;
    /** @type {?} */
    ThfRadioGroupComponent.prototype.inputEl;
    /** @type {?} */
    ThfRadioGroupComponent.prototype.differ;
}
