// @totvs/thf-ui/esm2015/lib/components/thf-field/thf-lookup/services/thf-lookup-filter.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { isTypeof } from '../../../../utils/util';
/**
 * \@docsPrivate
 *
 * \@description
 *
 * Serviço padrão utilizado para filtrar os dados do componente thf-lookup.
 */
export class ThfLookupFilterService {
    /**
     * @param {?} httpClient
     */
    constructor(httpClient) {
        this.httpClient = httpClient;
    }
    /**
     * @param {?} filter
     * @param {?} page
     * @param {?=} pageSize
     * @param {?=} filterParams
     * @return {?}
     */
    getFilteredData(filter, page, pageSize, filterParams) {
        /** @type {?} */
        const validatedFilterParams = this.validateParams(filterParams);
        return this.httpClient.get(this.url, { params: Object.assign({ page: page.toString(), pageSize: pageSize.toString() }, validatedFilterParams, { filter }) });
    }
    /**
     * @param {?} value
     * @param {?=} filterParams
     * @return {?}
     */
    getObjectByValue(value, filterParams) {
        /** @type {?} */
        const validatedFilterParams = this.validateParams(filterParams);
        return this.httpClient.get(`${this.url}/${value}`, { params: validatedFilterParams });
    }
    /**
     * @param {?} url
     * @return {?}
     */
    setUrl(url) {
        this.url = url;
    }
    /**
     * @private
     * @param {?} params
     * @return {?}
     */
    validateParams(params) {
        return isTypeof(params, 'object') && !Array.isArray(params) ? params : undefined;
    }
}
ThfLookupFilterService.decorators = [
    { type: Injectable }
];
/** @nocollapse */
ThfLookupFilterService.ctorParameters = () => [
    { type: HttpClient }
];
if (false) {
    /**
     * @type {?}
     * @private
     */
    ThfLookupFilterService.prototype.url;
    /**
     * @type {?}
     * @private
     */
    ThfLookupFilterService.prototype.httpClient;
}
