// @totvs/thf-ui/esm2015/lib/components/thf-field/thf-lookup/services/thf-lookup-modal.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable, EventEmitter } from '@angular/core';
import { ThfComponentInjectorService } from '../../../../services/thf-component-injector/thf-component-injector.service';
import { ThfLookupModalComponent } from '../thf-lookup-modal/thf-lookup-modal.component';
/**
 * \@docsPrivate
 *
 * Serviço responsável por controlar a abertura do componente Thf Lookup Modal.
 */
export class ThfLookupModalService {
    /**
     * @param {?} thfComponentInjector
     */
    constructor(thfComponentInjector) {
        this.thfComponentInjector = thfComponentInjector;
        this.componentRef = null;
        this.selectValueEvent = new EventEmitter();
    }
    /**
     * Método responsável por abrir a modal de busca das informações.
     *
     * @param {?} params
     * @return {?}
     */
    openModal(params) {
        const { service, columns, filterParams, title, literals } = params;
        this.componentRef = this.thfComponentInjector.createComponentInApplication(ThfLookupModalComponent);
        this.componentRef.instance.title = title;
        this.componentRef.instance.columns = columns;
        this.componentRef.instance.filterService = service;
        this.componentRef.instance.filterParams = filterParams;
        this.componentRef.instance.literals = literals;
        this.componentRef.instance.model.subscribe((/**
         * @param {?} $event
         * @return {?}
         */
        $event => {
            this.selectValue($event);
        }));
        this.componentRef.changeDetectorRef.detectChanges();
        this.componentRef.instance.openModal();
    }
    // Este metodo é chamado quando é selecionado um item na lookup modal.
    /**
     * @param {?} value
     * @return {?}
     */
    selectValue(value) {
        if (value) {
            this.selectValueEvent.emit(value);
        }
        this.componentRef.destroy();
    }
}
ThfLookupModalService.decorators = [
    { type: Injectable }
];
/** @nocollapse */
ThfLookupModalService.ctorParameters = () => [
    { type: ThfComponentInjectorService }
];
if (false) {
    /**
     * @type {?}
     * @private
     */
    ThfLookupModalService.prototype.componentRef;
    /** @type {?} */
    ThfLookupModalService.prototype.selectValueEvent;
    /**
     * @type {?}
     * @private
     */
    ThfLookupModalService.prototype.thfComponentInjector;
}
