// @totvs/thf-ui/esm2015/lib/components/thf-widget/thf-widget.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ViewContainerRef } from '@angular/core';
import { ThfWidgetBaseComponent } from './thf-widget-base.component';
/**
 * \@docsExtends ThfWidgetBaseComponent
 *
 * \@example
 *
 * <example name="thf-widget-basic" title="Totvs Widget Basic">
 *  <file name="sample-thf-widget-basic/sample-thf-widget-basic.component.html"> </file>
 *  <file name="sample-thf-widget-basic/sample-thf-widget-basic.component.ts"> </file>
 * </example>
 *
 * <example name="thf-widget-labs" title="Totvs Widget Labs">
 *  <file name="sample-thf-widget-labs/sample-thf-widget-labs.component.html"> </file>
 *  <file name="sample-thf-widget-labs/sample-thf-widget-labs.component.ts"> </file>
 * </example>
 *
 * <example name="thf-widget-finance-dashboard" title="Totvs Widget - Finance dashboard">
 *  <file name="sample-thf-widget-finance-dashboard/sample-thf-widget-finance-dashboard.component.html"> </file>
 *  <file name="sample-thf-widget-finance-dashboard/sample-thf-widget-finance-dashboard.component.ts"> </file>
 * </example>
 *
 * <example name="thf-widget-card" title="Totvs Widget - Card">
 *  <file name="sample-thf-widget-card/sample-thf-widget-card.component.html"> </file>
 *  <file name="sample-thf-widget-card/sample-thf-widget-card.component.ts"> </file>
 * </example>
 *
 */
export class ThfWidgetComponent extends ThfWidgetBaseComponent {
    /**
     * @param {?} viewRef
     */
    constructor(viewRef) {
        super();
    }
    /**
     * @return {?}
     */
    get showTitleAction() {
        return !!this.titleAction.observers[0];
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this.setHeight(this.height);
    }
    /**
     * @return {?}
     */
    hasTitleHelpOrSetting() {
        return !!this.title || !!this.help || !!this.setting.observers[0];
    }
    /**
     * @param {?} event
     * @return {?}
     */
    onClick(event) {
        if (!this.disabled) {
            this.click.emit(event);
        }
    }
    /**
     * @param {?} event
     * @return {?}
     */
    openHelp(event) {
        if (!this.disabled) {
            event.stopPropagation();
            window.open(this.help, '_blank');
        }
    }
    /**
     * @param {?} event
     * @return {?}
     */
    runPrimaryAction(event) {
        if (!this.disabled) {
            event.stopPropagation();
            this.primaryAction.emit();
        }
    }
    /**
     * @param {?} event
     * @return {?}
     */
    runSecondaryAction(event) {
        if (!this.disabled) {
            event.stopPropagation();
            this.secondaryAction.emit();
        }
    }
    /**
     * @param {?} event
     * @return {?}
     */
    runTitleAction(event) {
        if (!this.disabled) {
            event.stopPropagation();
            this.titleAction.emit();
        }
    }
    /**
     * @param {?} height
     * @return {?}
     */
    setHeight(height) {
        if (height) {
            /** @type {?} */
            let bodyHeight = height;
            /** @type {?} */
            const hasSettingOrHelp = this.setting.observers.length > 0 || this.help;
            /** @type {?} */
            const footerBorder = 1;
            /** @type {?} */
            const footerHeight = 40;
            /** @type {?} */
            const settingHeight = 37;
            /** @type {?} */
            const shadowHeight = 2;
            /** @type {?} */
            const titleHeight = 50;
            if (this.noShadow) {
                bodyHeight -= shadowHeight;
            }
            if (hasSettingOrHelp && !this.title) {
                bodyHeight -= settingHeight;
            }
            if (this.title) {
                bodyHeight -= titleHeight;
            }
            if (this.primaryLabel) {
                bodyHeight -= footerHeight + footerBorder;
            }
            this.containerHeight = `${bodyHeight}px`;
        }
        else {
            this.containerHeight = `auto`;
        }
    }
    /**
     * @param {?} event
     * @return {?}
     */
    settingOutput(event) {
        if (!this.disabled) {
            event.stopPropagation();
            this.setting.emit();
        }
    }
}
ThfWidgetComponent.decorators = [
    { type: Component, args: [{
                selector: 'thf-widget',
                template: "<div\n  [class.thf-clickable]=\"click.observers.length\"\n  [class.thf-widget]=\"!primary\"\n  [class.thf-widget-disabled]=\"disabled\"\n  [class.thf-widget-primary]=\"primary\"\n  [class.thf-widget-no-shadow]=\"noShadow\"\n  [style.background-image]=\"background ? 'url(' + background + ')' : undefined\"\n  (click)=\"onClick($event)\">\n\n  <div *ngIf=\"hasTitleHelpOrSetting()\" class=\"thf-widget-header\">\n\n    <span\n      *ngIf=\"showTitleAction; else noTitleAction\"\n      class=\"thf-widget-title-action\"\n      (click)=\"runTitleAction($event)\">\n      {{ title }}\n    </span>\n    <ng-template #noTitleAction>{{ title }}</ng-template>\n\n    <div *ngIf=\"setting.observers[0] || help\" class=\"thf-pull-right\">\n      <span id=\"settingAction\"\n        *ngIf=\"setting.observers[0]\"\n        class=\"thf-clickable thf-icon thf-icon-settings\"\n        (click)=\"settingOutput($event)\">\n      </span>\n      <span id=\"helpLink\"\n        *ngIf=\"help\"\n        class=\"thf-clickable thf-icon thf-icon-help\"\n        (click)=\"openHelp($event)\">\n      </span>\n    </div>\n  </div>\n\n  <thf-container\n    t-no-border\n    [t-height]=\"containerHeight\">\n    <div class=\"thf-widget-body\">\n      <ng-content></ng-content>\n    </div>\n  </thf-container>\n\n  <div *ngIf=\"primaryLabel\" class=\"thf-widget-footer\">\n\n    <div class=\"thf-widget-xl\" *ngIf=\"!secondaryLabel\">\n      <a class=\"thf-widget-action\" id='primaryAct' (click)=\"runPrimaryAction($event)\">{{ primaryLabel }}</a>\n    </div>\n\n    <div class=\"thf-widget-md\" *ngIf=\"secondaryLabel\">\n      <a class=\"thf-widget-action\" id='primaryAct' (click)=\"runPrimaryAction($event)\">{{ primaryLabel }}</a>\n    </div>\n    <div class=\"thf-widget-md\" *ngIf=\"secondaryLabel\">\n      <a class=\"thf-widget-action\" id='secondaryAct' (click)=\"runSecondaryAction($event)\">{{ secondaryLabel }}</a>\n    </div>\n\n  </div>\n</div>\n"
            }] }
];
/** @nocollapse */
ThfWidgetComponent.ctorParameters = () => [
    { type: ViewContainerRef }
];
