// @totvs/thf-ui/esm5/lib/components/thf-navbar/thf-navbar-item-navigation/thf-navbar-item-navigation.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, EventEmitter, Input, Output } from '@angular/core';
var ThfNavbarItemNavigationComponent = /** @class */ (function () {
    function ThfNavbarItemNavigationComponent() {
        this.click = new EventEmitter();
    }
    ThfNavbarItemNavigationComponent.decorators = [
        { type: Component, args: [{
                    selector: 'thf-navbar-item-navigation',
                    template: "<div class=\"thf-navbar-item-navigation-icon\"\n  [ngClass]=\"{'thf-navbar-item-navigation-icon-disabled': disableLeft , 'thf-clickable': !disableLeft }\"\n  tabindex=\"0\"\n  (click)=\"disableLeft ? undefined : click.emit('left')\" >\n  <span class=\"thf-icon thf-icon-arrow-left\"></span>\n</div>\n<div class=\"thf-navbar-item-navigation-icon\"\n  [ngClass]=\"{'thf-navbar-item-navigation-icon-disabled': disableRight , 'thf-clickable': !disableRight }\"  \n  tabindex=\"0\"\n  (click)=\"disableRight ? undefined : click.emit('right')\">\n  <span class=\"thf-icon thf-icon-arrow-right\"></span>\n</div>\n"
                }] }
    ];
    ThfNavbarItemNavigationComponent.propDecorators = {
        disableLeft: [{ type: Input, args: ['t-disable-left',] }],
        disableRight: [{ type: Input, args: ['t-disable-right',] }],
        click: [{ type: Output, args: ['t-click',] }]
    };
    return ThfNavbarItemNavigationComponent;
}());
export { ThfNavbarItemNavigationComponent };
if (false) {
    /** @type {?} */
    ThfNavbarItemNavigationComponent.prototype.disableLeft;
    /** @type {?} */
    ThfNavbarItemNavigationComponent.prototype.disableRight;
    /** @type {?} */
    ThfNavbarItemNavigationComponent.prototype.click;
}
