// @totvs/thf-ui/esm5/lib/thf.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { ThfComponentsModule } from './components/components.module';
import { ThfDirectivesModule } from './directives/directives.module';
import { ThfGuardsModule } from './guards/guards.module';
import { ThfInterceptorsModule } from './interceptors/interceptors.module';
import { ThfPipesModule } from './pipes/pipes.module';
import { ThfServicesModule } from './services/services.module';
import { ThfNotificationService } from './services/thf-notification/thf-notification.service';
var ThfModule = /** @class */ (function () {
    function ThfModule() {
    }
    ThfModule.decorators = [
        { type: NgModule, args: [{
                    declarations: [],
                    imports: [
                        ThfComponentsModule,
                        ThfDirectivesModule,
                        ThfGuardsModule,
                        ThfInterceptorsModule,
                        ThfPipesModule,
                        ThfServicesModule
                    ],
                    exports: [
                        ThfComponentsModule,
                        ThfDirectivesModule,
                        ThfInterceptorsModule,
                        ThfGuardsModule,
                        ThfPipesModule,
                        ThfServicesModule
                    ],
                    providers: [ThfNotificationService],
                    bootstrap: []
                },] }
    ];
    return ThfModule;
}());
export { ThfModule };
