// @totvs/thf-ui/esm2015/lib/components/thf-navbar/thf-navbar-items/thf-navbar-item/thf-navbar-item.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, EventEmitter, Output } from '@angular/core';
import { isExternalLink } from '../../../../utils/util';
export class ThfNavbarItemComponent {
    constructor() {
        this.click = new EventEmitter();
    }
    /**
     * @return {?}
     */
    get type() {
        if (isExternalLink(this.link)) {
            return 'externalLink';
        }
        return 'internalLink';
    }
    /**
     * @return {?}
     */
    itemClick() {
        if (this.action) {
            this.action({ label: this.label, link: this.link });
        }
        this.click.emit();
    }
}
ThfNavbarItemComponent.decorators = [
    { type: Component, args: [{
                selector: 'thf-navbar-item',
                template: "<ng-container [ngSwitch]=\"type\">\n\n  <a *ngSwitchCase=\"'externalLink'\"\n    class=\"thf-navbar-item-link\"\n    [class.thf-clickable]=\"clickable\"\n    [href]=\"link\" \n    target=\"_blank\"\n    (click)=\"itemClick()\">{{ label }}</a>\n\n  <a *ngSwitchCase=\"'internalLink'\"\n    class=\"thf-navbar-item-link\"\n    [class.thf-clickable]=\"clickable\"\n    [routerLink]=\"link\"\n    (click)=\"itemClick()\">{{ label }}</a>\n\n</ng-container>\n"
            }] }
];
ThfNavbarItemComponent.propDecorators = {
    action: [{ type: Input, args: ['t-action',] }],
    clickable: [{ type: Input, args: ['t-clickable',] }],
    label: [{ type: Input, args: ['t-label',] }],
    link: [{ type: Input, args: ['t-link',] }],
    click: [{ type: Output, args: ['t-click',] }]
};
if (false) {
    /** @type {?} */
    ThfNavbarItemComponent.prototype.action;
    /** @type {?} */
    ThfNavbarItemComponent.prototype.clickable;
    /** @type {?} */
    ThfNavbarItemComponent.prototype.label;
    /** @type {?} */
    ThfNavbarItemComponent.prototype.link;
    /** @type {?} */
    ThfNavbarItemComponent.prototype.click;
}
