// @totvs/thf-ui/esm5/lib/components/thf-avatar/thf-avatar.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component } from '@angular/core';
import { ThfAvatarBaseComponent } from './thf-avatar-base.component';
/**
 * \@docsExtends ThfAvatarBaseComponent
 *
 * \@example
 *
 * <example name="thf-avatar-basic" title="Totvs Avatar Basic" >
 *  <file name="sample-thf-avatar-basic/sample-thf-avatar-basic.component.html"> </file>
 *  <file name="sample-thf-avatar-basic/sample-thf-avatar-basic.component.ts"> </file>
 *  <file name="sample-thf-avatar-basic/sample-thf-avatar-basic.component.e2e-spec.ts"> </file>
 *  <file name="sample-thf-avatar-basic/sample-thf-avatar-basic.component.po.ts"> </file>
 * </example>
 *
 * <example name="thf-avatar-labs" title="Totvs Avatar Labs" >
 *  <file name="sample-thf-avatar-labs/sample-thf-avatar-labs.component.html"> </file>
 *  <file name="sample-thf-avatar-labs/sample-thf-avatar-labs.component.ts"> </file>
 * </example>
 *
 * <example name="thf-avatar-business-card" title="Totvs Avatar - Business Card" >
 *  <file name="sample-thf-avatar-business-card/sample-thf-avatar-business-card.component.html"> </file>
 *  <file name="sample-thf-avatar-business-card/sample-thf-avatar-business-card.component.ts"> </file>
 * </example>
 */
var ThfAvatarComponent = /** @class */ (function (_super) {
    tslib_1.__extends(ThfAvatarComponent, _super);
    function ThfAvatarComponent() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /**
     * @return {?}
     */
    ThfAvatarComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        if (!this.src) {
            this.src = undefined;
        }
    };
    /**
     * @return {?}
     */
    ThfAvatarComponent.prototype.onError = /**
     * @return {?}
     */
    function () {
        this.src = undefined;
    };
    ThfAvatarComponent.decorators = [
        { type: Component, args: [{
                    selector: 'thf-avatar',
                    template: "<div class=\"thf-avatar thf-avatar-{{ size }}\">\n  <ng-container *ngIf=\"src; then sourceImage; else defaultIcon\"></ng-container>\n</div>\n\n<ng-template #sourceImage>\n  <img class=\"thf-avatar-image\" [src]=\"src\" alt=\"\" (error)=\"onError()\">\n</ng-template>\n\n<ng-template #defaultIcon>\n  <span class=\"thf-icon thf-icon-user thf-avatar-default-icon\"></span>\n</ng-template>\n"
                }] }
    ];
    return ThfAvatarComponent;
}(ThfAvatarBaseComponent));
export { ThfAvatarComponent };
