// @totvs/thf-ui/esm5/lib/components/thf-grid/thf-grid-cell-action/thf-grid-cell-action.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input } from '@angular/core';
var ThfGridCellActionComponent = /** @class */ (function () {
    function ThfGridCellActionComponent() {
        this.position = '';
    }
    /**
     * @param {?} event
     * @return {?}
     */
    ThfGridCellActionComponent.prototype.onKeyDownContent = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        // console.log('onKeyDownContent: ', event);
        // ENTER
        if (event.keyCode === 13) {
            event.preventDefault();
            // this.openActions(this.value);
            return;
        }
    };
    ThfGridCellActionComponent.decorators = [
        { type: Component, args: [{
                    selector: 'thf-grid-cell-action',
                    template: "<div>\n  <div #contentElement [id]=\"position\"\n    class=\"thf-grid-cell-action-content\"\n    tabindex=\"0\"\n    (keydown)=\"onKeyDownContent($event)\">\n    <span>{{ value || '&nbsp;' }}</span>\n  </div>\n</div>\n\n",
                    styles: [".thf-grid-cell-action-content{border-bottom:1px solid #eceeee;border-left:none;border-right:1px solid #eceeee;border-top:none;font-family:NunitoSans;font-size:14px;height:44px;line-height:1.71;margin:0;max-height:44px;max-width:56px;min-height:44px;min-width:56px;padding:10px 16px;text-align:center;width:56px}.thf-grid-cell-action-content:focus{border:none;outline:#0c9abe solid 1px}"]
                }] }
    ];
    /** @nocollapse */
    ThfGridCellActionComponent.ctorParameters = function () { return []; };
    ThfGridCellActionComponent.propDecorators = {
        position: [{ type: Input, args: ['t-position',] }],
        value: [{ type: Input, args: ['t-value',] }]
    };
    return ThfGridCellActionComponent;
}());
export { ThfGridCellActionComponent };
if (false) {
    /** @type {?} */
    ThfGridCellActionComponent.prototype.position;
    /** @type {?} */
    ThfGridCellActionComponent.prototype.value;
}
