// @totvs/thf-ui/esm2015/lib/components/thf-chart/thf-chart-types/thf-chart-dynamic-type.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ElementRef, ViewChild } from '@angular/core';
import { Subject } from 'rxjs';
/** @type {?} */
const Padding = 24;
/**
 * @abstract
 */
export class ThfChartDynamicTypeComponent {
    constructor() {
        this.onSerieClick = new Subject();
        this.onSerieHover = new Subject();
        this.series = [];
    }
    /**
     * @param {?} chartWrapperElement
     * @param {?} chartHeaderElement
     * @param {?} chartLegendElement
     * @return {?}
     */
    calculateSVGContainerDimensions(chartWrapperElement, chartHeaderElement, chartLegendElement) {
        /** @type {?} */
        const svgContainerHeightCalc = this.height - chartHeaderElement - chartLegendElement - (Padding * 2);
        this.svgHeight = svgContainerHeightCalc <= 0 ? 0 : svgContainerHeightCalc;
        this.centerX = chartWrapperElement / 2;
    }
    /**
     * @return {?}
     */
    calculateTotalValue() {
        this.totalValue = this.series.reduce((/**
         * @param {?} previousValue
         * @param {?} serie
         * @return {?}
         */
        (previousValue, serie) => previousValue + serie.value), 0);
    }
}
ThfChartDynamicTypeComponent.propDecorators = {
    chartBody: [{ type: ViewChild, args: ['chartBody', { static: true },] }],
    svgContainer: [{ type: ViewChild, args: ['svgContainer', { static: true },] }]
};
if (false) {
    /**
     * @type {?}
     * @protected
     */
    ThfChartDynamicTypeComponent.prototype.windowResizeListener;
    /**
     * @type {?}
     * @protected
     */
    ThfChartDynamicTypeComponent.prototype.windowScrollListener;
    /** @type {?} */
    ThfChartDynamicTypeComponent.prototype.centerX;
    /** @type {?} */
    ThfChartDynamicTypeComponent.prototype.chartElementCategory;
    /** @type {?} */
    ThfChartDynamicTypeComponent.prototype.chartElementValue;
    /** @type {?} */
    ThfChartDynamicTypeComponent.prototype.chartHeader;
    /** @type {?} */
    ThfChartDynamicTypeComponent.prototype.chartLegend;
    /** @type {?} */
    ThfChartDynamicTypeComponent.prototype.chartWrapper;
    /** @type {?} */
    ThfChartDynamicTypeComponent.prototype.colors;
    /** @type {?} */
    ThfChartDynamicTypeComponent.prototype.height;
    /** @type {?} */
    ThfChartDynamicTypeComponent.prototype.onSerieClick;
    /** @type {?} */
    ThfChartDynamicTypeComponent.prototype.onSerieHover;
    /** @type {?} */
    ThfChartDynamicTypeComponent.prototype.series;
    /** @type {?} */
    ThfChartDynamicTypeComponent.prototype.svgElement;
    /** @type {?} */
    ThfChartDynamicTypeComponent.prototype.svgHeight;
    /** @type {?} */
    ThfChartDynamicTypeComponent.prototype.target;
    /** @type {?} */
    ThfChartDynamicTypeComponent.prototype.tooltipElement;
    /** @type {?} */
    ThfChartDynamicTypeComponent.prototype.tooltipText;
    /** @type {?} */
    ThfChartDynamicTypeComponent.prototype.totalValue;
    /** @type {?} */
    ThfChartDynamicTypeComponent.prototype.chartBody;
    /** @type {?} */
    ThfChartDynamicTypeComponent.prototype.svgContainer;
}
