// @totvs/thf-ui/esm2015/lib/components/thf-table/thf-table-subtitle-circle/thf-table-subtitle-circle.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input } from '@angular/core';
import { ThfColorPaletteService } from './../../../services/thf-color-palette/thf-color-palette.service';
/**
 * \@docsPrivate
 *
 * \@description
 *
 * Componente para a criação da representação da legenda, em formato de círculo.
 */
export class ThfTableSubtitleCircleComponent {
    /**
     * @param {?} thfColorPaletteService
     */
    constructor(thfColorPaletteService) {
        this.thfColorPaletteService = thfColorPaletteService;
        /**
         * Esconde a tooltip.
         */
        this.hideTitle = false;
    }
    /**
     * Objeto com os dados da legenda.
     * @param {?} subtitle
     * @return {?}
     */
    set subtitle(subtitle) {
        if (subtitle) {
            subtitle.color = this.thfColorPaletteService.getColor(subtitle);
        }
        this._subtitle = subtitle;
    }
    /**
     * @return {?}
     */
    get subtitle() {
        return this._subtitle;
    }
}
ThfTableSubtitleCircleComponent.decorators = [
    { type: Component, args: [{
                selector: 'thf-table-subtitle-circle',
                template: "<span\n  [title]=\"hideTitle ? '' : subtitle?.label\"\n  class=\"thf-table-subtitle-circle thf-{{ subtitle?.color }}\">\n  {{ subtitle?.content }}\n</span>\n"
            }] }
];
/** @nocollapse */
ThfTableSubtitleCircleComponent.ctorParameters = () => [
    { type: ThfColorPaletteService }
];
ThfTableSubtitleCircleComponent.propDecorators = {
    subtitle: [{ type: Input, args: ['t-subtitle',] }],
    hideTitle: [{ type: Input, args: ['t-hide-title',] }]
};
if (false) {
    /**
     * @type {?}
     * @private
     */
    ThfTableSubtitleCircleComponent.prototype._subtitle;
    /**
     * Esconde a tooltip.
     * @type {?}
     */
    ThfTableSubtitleCircleComponent.prototype.hideTitle;
    /**
     * @type {?}
     * @private
     */
    ThfTableSubtitleCircleComponent.prototype.thfColorPaletteService;
}
