// @totvs/thf-ui/esm2015/lib/components/thf-field/thf-number/thf-number-base.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ThfInputGeneric } from '../thf-input-generic/thf-input-generic';
/**
 * @abstract
 */
export class ThfNumberBaseComponent extends ThfInputGeneric {
    /**
     * @param {?} elementRef
     */
    constructor(elementRef) {
        super(elementRef);
        this.type = 'number';
    }
    /**
     * @param {?} e
     * @return {?}
     */
    eventOnInput(e) {
        if (!this.mask) {
            /** @type {?} */
            let value = e.target.value;
            /** @type {?} */
            const valueMaxlength = this.validMaxLength(this.maxlength, value);
            if (value !== valueMaxlength) {
                value = valueMaxlength;
                this.inputEl.nativeElement.value = value;
            }
            this.callOnChange(this.formatNumber(value));
        }
    }
    /**
     * @param {?} maxlength
     * @param {?} value
     * @return {?}
     */
    validMaxLength(maxlength, value) {
        if (maxlength && value.length > maxlength) {
            /** @type {?} */
            const substringValue = value.toString().substring(0, maxlength);
            if (substringValue && this.isEndWithDot(substringValue)) {
                return substringValue.toString().substring(0, maxlength - 1);
            }
            return substringValue;
        }
        return value;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    writeValueModel(value) {
        if (this.inputEl) {
            if (value || value === 0) {
                if (this.mask) {
                    this.inputEl.nativeElement.value = this.objMask.controlFormatting(String(value));
                    // Se o model for definido como formatado, então precisa atualizá-lo no primeiro acesso
                    if (this.objMask.formatModel) {
                        this.onChangePropagate(this.objMask.valueToModel);
                    }
                }
                else {
                    this.inputEl.nativeElement.value = value;
                }
            }
            else { // Se for o valor for undefined, deve limpar o campo
                this.inputEl.nativeElement.value = '';
            }
        }
        // Emite evento quando o model é atualizado, inclusive a primeira vez
        this.changeModel.emit(value);
    }
    /**
     * @private
     * @param {?} value
     * @return {?}
     */
    isEndWithDot(value) {
        return value && value.lastIndexOf('.') === value.length - 1;
    }
    /**
     * @private
     * @param {?} value
     * @return {?}
     */
    formatNumber(value) {
        return value ? Number(value) : null;
    }
}
if (false) {
    /** @type {?} */
    ThfNumberBaseComponent.prototype.type;
}
