// @totvs/thf-ui/esm2015/lib/components/thf-field/thf-rich-text/thf-rich-text.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ElementRef, forwardRef } from '@angular/core';
import { NG_VALIDATORS, NG_VALUE_ACCESSOR } from '@angular/forms';
import { ThfRichTextBaseComponent } from './thf-rich-text-base.component';
/**
 * \@docsExtends ThfRichTextBaseComponent
 *
 * \@example
 *
 * <example name="thf-rich-text-basic" title="Totvs Rich Text Basic">
 *   <file name="sample-thf-rich-text-basic/sample-thf-rich-text-basic.component.html"> </file>
 *   <file name="sample-thf-rich-text-basic/sample-thf-rich-text-basic.component.ts"> </file>
 * </example>
 *
 * <example name="thf-rich-text-recipe" title="Totvs Rich Text Recipe">
 *   <file name="sample-thf-rich-text-recipe/sample-thf-rich-text-recipe.component.html"> </file>
 *   <file name="sample-thf-rich-text-recipe/sample-thf-rich-text-recipe.component.ts"> </file>
 * </example>
 */
export class ThfRichTextComponent extends ThfRichTextBaseComponent {
    /**
     * @param {?} element
     */
    constructor(element) {
        super();
        this.element = element;
        this.listener = this.validateClassesForRequired.bind(this);
    }
    /**
     * @return {?}
     */
    get errorMsg() {
        return (this.errorMessage !== '' && !this.value && this.required && this.invalid) ? this.errorMessage : '';
    }
    /**
     * @return {?}
     */
    ngAfterViewInit() {
        // Se não tem ngModel ou reactive form adiciona validação com classes css
        if (!this.onChangeModel) {
            this.element.nativeElement.addEventListener('keyup', this.listener);
            this.element.nativeElement.addEventListener('cut', this.listener);
            this.element.nativeElement.addEventListener('paste', this.listener);
        }
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        if (!this.onChangeModel) {
            this.element.nativeElement.removeEventListener('keyup', this.listener);
            this.element.nativeElement.removeEventListener('cut', this.listener);
            this.element.nativeElement.removeEventListener('paste', this.listener);
        }
    }
    /**
     * @param {?} value
     * @return {?}
     */
    updateValue(value) {
        this.value = value;
        this.invalid = !value;
        this.updateModel(this.value);
    }
    /**
     * @private
     * @return {?}
     */
    validateClassesForRequired() {
        setTimeout((/**
         * @return {?}
         */
        () => {
            /** @type {?} */
            const value = this.value;
            /** @type {?} */
            const element = this.element.nativeElement;
            if (!value && this.required) {
                element.classList.add('ng-invalid');
                element.classList.add('ng-dirty');
            }
            else {
                element.classList.remove('ng-invalid');
            }
        }));
    }
}
ThfRichTextComponent.decorators = [
    { type: Component, args: [{
                selector: 'thf-rich-text',
                template: "<thf-field-container\n[t-help]=\"help\"\n[t-label]=\"label\"\n[t-optional]=\"!required && optional\">\n\n  <div class=\"thf-field-container-content\">\n\n    <thf-rich-text-body #richTextBody\n      [t-height]=\"height\"\n      [t-model-value]=\"value\"\n      [t-placeholder]=\"placeholder\"\n      [t-readonly]=\"readonly\"\n      (t-commands)=\"richTextToolbar.setButtonsStates($event)\"\n      (t-value)=\"updateValue($event)\">\n    </thf-rich-text-body>\n\n    <thf-rich-text-toolbar #richTextToolbar\n      [t-readonly]=\"readonly\"\n      (t-command)=\"richTextBody.executeCommand($event)\">\n    </thf-rich-text-toolbar>\n\n  </div>\n\n  <thf-field-container-bottom [t-error-pattern]=\"errorMsg\"></thf-field-container-bottom>\n\n</thf-field-container>\n",
                providers: [
                    {
                        provide: NG_VALUE_ACCESSOR,
                        useExisting: forwardRef((/**
                         * @return {?}
                         */
                        () => ThfRichTextComponent)),
                        multi: true,
                    },
                    {
                        provide: NG_VALIDATORS,
                        useExisting: forwardRef((/**
                         * @return {?}
                         */
                        () => ThfRichTextComponent)),
                        multi: true,
                    }
                ]
            }] }
];
/** @nocollapse */
ThfRichTextComponent.ctorParameters = () => [
    { type: ElementRef }
];
if (false) {
    /**
     * @type {?}
     * @private
     */
    ThfRichTextComponent.prototype.listener;
    /**
     * @type {?}
     * @private
     */
    ThfRichTextComponent.prototype.element;
}
