// @totvs/thf-ui/esm2015/lib/components/thf-dynamic/thf-dynamic-view/thf-dynamic-view.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component } from '@angular/core';
import { CurrencyPipe, DatePipe, DecimalPipe, TitleCasePipe } from '@angular/common';
import { ThfTimePipe } from '../../../pipes/thf-time/thf-time.pipe';
import { ThfDynamicViewBaseComponent } from './thf-dynamic-view-base.component';
/**
 * \@docsExtends ThfDynamicViewBaseComponent
 *
 * \@example
 *
 * <example name="thf-dynamic-view-basic" title="Totvs Dynamic View Basic">
 *  <file name="sample-thf-dynamic-view-basic/sample-thf-dynamic-view-basic.component.html"> </file>
 *  <file name="sample-thf-dynamic-view-basic/sample-thf-dynamic-view-basic.component.ts"> </file>
 * </example>
 *
 * <example name="thf-dynamic-view-employee" title="Totvs Dynamic View - Employee">
 *  <file name="sample-thf-dynamic-view-employee/sample-thf-dynamic-view-employee.component.html"> </file>
 *  <file name="sample-thf-dynamic-view-employee/sample-thf-dynamic-view-employee.component.ts"> </file>
 * </example>
 */
export class ThfDynamicViewComponent extends ThfDynamicViewBaseComponent {
    /**
     * @param {?} currencyPipe
     * @param {?} datePipe
     * @param {?} decimalPipe
     * @param {?} timePipe
     * @param {?} titleCasePipe
     */
    constructor(currencyPipe, datePipe, decimalPipe, timePipe, titleCasePipe) {
        super(currencyPipe, datePipe, decimalPipe, timePipe, titleCasePipe);
    }
    /**
     * @param {?} changes
     * @return {?}
     */
    ngOnChanges(changes) {
        if (changes.fields || changes.value || changes.showAllValue) {
            this.visibleFields = this.getVisibleFields();
        }
    }
    /**
     * @private
     * @return {?}
     */
    getVisibleFields() {
        if (this.showAllValue) {
            return this.getMergedFields();
        }
        return this.value && this.fields.length ? this.getConfiguredFields() : this.getValueFields();
    }
}
ThfDynamicViewComponent.decorators = [
    { type: Component, args: [{
                selector: 'thf-dynamic-view',
                template: "<div class=\"thf-row\" *ngIf=\"visibleFields.length\">\n\n  <ng-template ngFor let-field [ngForOf]=\"visibleFields\">\n\n    <thf-divider *ngIf=\"field?.divider?.trim()\" class=\"thf-sm-12\" [t-label]=\"field.divider\">\n    </thf-divider>\n\n    <ng-container *ngTemplateOutlet=\"!field.tag ? thfInfo : thfTag; context: { $implicit: field }\">\n    </ng-container>\n\n  </ng-template>\n\n</div>\n\n<ng-template #thfInfo let-field>\n  <thf-info [ngClass]=\"field.cssClass\"\n    [t-label]=\"field.label\"\n    [t-value]=\"field.value\">\n  </thf-info>\n</ng-template>\n\n<ng-template #thfTag let-field>\n  <thf-tag [ngClass]=\"field.cssClass\"\n    [t-label]=\"field.label\"\n    [t-value]=\"field.value\">\n  </thf-tag>\n</ng-template>\n"
            }] }
];
/** @nocollapse */
ThfDynamicViewComponent.ctorParameters = () => [
    { type: CurrencyPipe },
    { type: DatePipe },
    { type: DecimalPipe },
    { type: ThfTimePipe },
    { type: TitleCasePipe }
];
