// @totvs/thf-ui/esm2015/lib/components/thf-page/thf-page-edit/thf-page-edit.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ViewContainerRef } from '@angular/core';
import { ThfPageEditBaseComponent } from './thf-page-edit-base.component';
import { callAction, hasAction } from '../thf-page-util/thf-page-util';
/**
 * \@docsExtends ThfPageEditBaseComponent
 *
 * \@example
 *
 * <example name="thf-page-edit-basic" title="Totvs Page Edit Basic">
 *  <file name="sample-thf-page-edit-basic/sample-thf-page-edit-basic.component.html"> </file>
 *  <file name="sample-thf-page-edit-basic/sample-thf-page-edit-basic.component.ts"> </file>
 * </example>
 *
 * <example name="thf-page-edit-labs" title="Totvs Page Edit Labs">
 *  <file name="sample-thf-page-edit-labs/sample-thf-page-edit-labs.component.html"> </file>
 *  <file name="sample-thf-page-edit-labs/sample-thf-page-edit-labs.component.ts"> </file>
 * </example>
 *
 * <example name="thf-page-edit-user" title="Totvs Page Edit - User">
 *  <file name="sample-thf-page-edit-user/sample-thf-page-edit-user.component.html"> </file>
 *  <file name="sample-thf-page-edit-user/sample-thf-page-edit-user.component.ts"> </file>
 * </example>
 */
export class ThfPageEditComponent extends ThfPageEditBaseComponent {
    /**
     * @param {?} viewRef
     */
    constructor(viewRef) {
        super();
        this.hasAction = hasAction;
        this.callAction = callAction;
        this.parentContext = viewRef['_view']['component'];
    }
    /**
     * @param {?} icon
     * @return {?}
     */
    getIcon(icon) {
        if (icon === 'cancel') {
            return this.isPrimaryAction('cancel') ? 'thf-icon-close' : '';
        }
        if (icon === 'saveNew') {
            return this.isPrimaryAction('saveNew') ? 'thf-icon-ok' : '';
        }
        return '';
    }
    /**
     * @param {?} type
     * @return {?}
     */
    getType(type) {
        /** @type {?} */
        const isCancelPrimaryAction = type === 'cancel' && this.isPrimaryAction('cancel');
        /** @type {?} */
        const isSaveNewPrimaryAction = type === 'saveNew' && this.isPrimaryAction('saveNew');
        return isCancelPrimaryAction || isSaveNewPrimaryAction ? 'primary' : 'default';
    }
    /**
     * @return {?}
     */
    hasAnyAction() {
        return hasAction('cancel', this.parentContext) || hasAction('saveNew', this.parentContext) || hasAction('save', this.parentContext);
    }
    /**
     * @return {?}
     */
    hasPageHeader() {
        return !!(this.title || this.hasAnyAction() || (this.breadcrumb && this.breadcrumb.items.length));
    }
    /**
     * @private
     * @param {?} action
     * @return {?}
     */
    isPrimaryAction(action) {
        /** @type {?} */
        const hasSaveAction = !hasAction('save', this.parentContext);
        if (action === 'saveNew') {
            return hasSaveAction;
        }
        if (action === 'cancel') {
            return !hasAction('saveNew', this.parentContext) && hasSaveAction;
        }
        return false;
    }
}
ThfPageEditComponent.decorators = [
    { type: Component, args: [{
                selector: 'thf-page-edit',
                template: "<thf-page>\n  <!-- HEADER -->\n  <thf-page-header\n    *ngIf=\"hasPageHeader()\"\n    [t-breadcrumb]=\"breadcrumb\"\n    [t-title]=\"title\">\n\n    <!-- OPERATIONS -->\n    <div *ngIf=\"hasAnyAction()\" class=\"thf-page-header-actions\">\n      <thf-button\n        *ngIf=\"hasAction('cancel', parentContext)\"\n        [t-icon]=\"getIcon('cancel')\"\n        [t-label]=\"literals.cancel\"\n        [t-type]=\"getType('cancel')\"\n        (t-click)=\"callAction('cancel', parentContext)\">\n      </thf-button>\n\n      <thf-button\n        *ngIf=\"hasAction('saveNew', parentContext)\"\n        [t-disabled]=\"disableSubmit\"\n        [t-icon]=\"getIcon('saveNew')\"\n        [t-label]=\"literals.saveNew\"\n        [t-type]=\"getType('saveNew')\"\n        (t-click)=\"callAction('saveNew', parentContext)\">\n      </thf-button>\n\n      <thf-button\n        *ngIf=\"hasAction('save', parentContext)\"\n        t-icon=\"thf-icon-ok\"\n        t-type=\"primary\"\n        [t-disabled]=\"disableSubmit\"\n        [t-label]=\"literals.save\"\n        (t-click)=\"callAction('save', parentContext)\">\n      </thf-button>\n    </div>\n\n  </thf-page-header>\n\n  <!-- CONTENT -->\n  <thf-page-content>\n    <ng-content>\n    </ng-content>\n  </thf-page-content>\n\n</thf-page>\n"
            }] }
];
/** @nocollapse */
ThfPageEditComponent.ctorParameters = () => [
    { type: ViewContainerRef }
];
if (false) {
    /** @type {?} */
    ThfPageEditComponent.prototype.hasAction;
    /** @type {?} */
    ThfPageEditComponent.prototype.callAction;
    /**
     * @type {?}
     * @private
     */
    ThfPageEditComponent.prototype.parentContext;
}
