// @totvs/thf-ui/esm5/lib/components/thf-menu/services/thf-menu-items.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { Subject } from 'rxjs';
/**
 * \@docsPrivate
 *
 * \@description
 *
 * Serviço que implementa a comunicação entre os items do thf-menu.
 */
var ThfMenuItemsService = /** @class */ (function () {
    function ThfMenuItemsService() {
        this.subjectParent = new Subject();
        this.subjectChild = new Subject();
    }
    // Envia informações do click do thf-menu-item para o thf-menu
    // Envia informações do click do thf-menu-item para o thf-menu
    /**
     * @param {?} menu
     * @return {?}
     */
    ThfMenuItemsService.prototype.sendToParentMenuClicked = 
    // Envia informações do click do thf-menu-item para o thf-menu
    /**
     * @param {?} menu
     * @return {?}
     */
    function (menu) {
        this.subjectChild.next(menu);
    };
    // Recebe do thf-menu-item sua informação de click.
    // Recebe do thf-menu-item sua informação de click.
    /**
     * @return {?}
     */
    ThfMenuItemsService.prototype.receiveFromChildMenuClicked = 
    // Recebe do thf-menu-item sua informação de click.
    /**
     * @return {?}
     */
    function () {
        return this.subjectChild.asObservable();
    };
    // Envia para os thf-menu-item a resposta do processamento de click de um thf-menu-item.
    // Envia para os thf-menu-item a resposta do processamento de click de um thf-menu-item.
    /**
     * @param {?} menu
     * @return {?}
     */
    ThfMenuItemsService.prototype.sendToChildMenuClicked = 
    // Envia para os thf-menu-item a resposta do processamento de click de um thf-menu-item.
    /**
     * @param {?} menu
     * @return {?}
     */
    function (menu) {
        this.subjectParent.next(menu);
    };
    // Recebe do thf-menu as informações processadas do click de um thf-menu-item.
    // Recebe do thf-menu as informações processadas do click de um thf-menu-item.
    /**
     * @return {?}
     */
    ThfMenuItemsService.prototype.receiveFromParentMenuClicked = 
    // Recebe do thf-menu as informações processadas do click de um thf-menu-item.
    /**
     * @return {?}
     */
    function () {
        return this.subjectParent.asObservable();
    };
    ThfMenuItemsService.decorators = [
        { type: Injectable }
    ];
    return ThfMenuItemsService;
}());
export { ThfMenuItemsService };
if (false) {
    /**
     * @type {?}
     * @private
     */
    ThfMenuItemsService.prototype.subjectParent;
    /**
     * @type {?}
     * @private
     */
    ThfMenuItemsService.prototype.subjectChild;
}
