// @totvs/thf-templates/esm5/lib/components/thf-page-login/thf-page-login-popover/thf-page-login-popover.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { ChangeDetectorRef, Component, EventEmitter, Input, Output } from '@angular/core';
import { browserLanguage, isExternalLink, isTypeof, thfLocaleDefault } from '../../../utils/util';
import { thfPageLoginLiteralsDefault } from './../thf-page-login-base.component';
var ThfPageLoginPopoverComponent = /** @class */ (function () {
    function ThfPageLoginPopoverComponent(changeDetector) {
        this.changeDetector = changeDetector;
        /**
         * se 't-recovery' for do tipo Function ou ThfPageLoginRecovery, emite para o método 'openUrl' do componente 'thf-page-login'
         */
        this.forgotPassword = new EventEmitter();
    }
    Object.defineProperty(ThfPageLoginPopoverComponent.prototype, "recovery", {
        get: /**
         * @return {?}
         */
        function () {
            return this._recovery;
        },
        /** exibe o link de 'esqueci minha senha' e verifica se o valor é um link interno ou externo */
        set: /**
         * exibe o link de 'esqueci minha senha' e verifica se o valor é um link interno ou externo
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._recovery = value;
            if (isTypeof(value, 'string')) {
                this.recoveryType = isExternalLink(value) ? 'externalLink' : 'internalLink';
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ThfPageLoginPopoverComponent.prototype, "remainingAttempts", {
        get: /**
         * @return {?}
         */
        function () {
            return this._remainingAttempts;
        },
        /** define se a mensagem deverá ser exibida caso seja maior que 0(zero) */
        set: /**
         * define se a mensagem deverá ser exibida caso seja maior que 0(zero)
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._remainingAttempts = value;
            this.getLiterals(this.selectedLanguage);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ThfPageLoginPopoverComponent.prototype, "selectedLanguage", {
        get: /**
         * @return {?}
         */
        function () {
            return this._selectedLanguage;
        },
        /** define o idioma da mensagem conforme selecionado no 'thf-page-login' */
        set: /**
         * define o idioma da mensagem conforme selecionado no 'thf-page-login'
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._selectedLanguage = value;
            this.getLiterals(value);
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    ThfPageLoginPopoverComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.getLiterals(this.selectedLanguage);
    };
    /**
     * @param {?} recovery
     * @return {?}
     */
    ThfPageLoginPopoverComponent.prototype.onForgotPasswordClick = /**
     * @param {?} recovery
     * @return {?}
     */
    function (recovery) {
        this.forgotPassword.emit(recovery);
    };
    /**
     * @private
     * @param {?=} language
     * @return {?}
     */
    ThfPageLoginPopoverComponent.prototype.getLiterals = /**
     * @private
     * @param {?=} language
     * @return {?}
     */
    function (language) {
        language = language || browserLanguage();
        this.literalParams = this.remainingAttempts;
        this.literals = tslib_1.__assign({}, thfPageLoginLiteralsDefault[thfLocaleDefault], thfPageLoginLiteralsDefault[language]);
        this.changeDetector.detectChanges();
    };
    ThfPageLoginPopoverComponent.decorators = [
        { type: Component, args: [{
                    selector: 'thf-page-login-popover',
                    template: "<div class=\"thf-page-login-popover-container\">\n  <div class=\"thf-page-login-popover\">\n    <div class=\"thf-page-login-popover-arrow\"></div>\n    <div class=\"thf-page-login-popover-content\">\n      <p class=\"thf-font-text-bold thf-page-login-popover-title\">\n        {{ literals?.titlePopover }}\n      </p>\n      <p class=\"thf-font-text\">\n        {{ literals?.forgotYourPassword }}\n      </p>\n      <p class=\"thf-font-text\">\n        {{ literals?.ifYouTryHarder }}\n        <span class=\"thf-font-text-bold thf-page-login-popover-attempts\">\n          {{ literals?.attempts | thfI18n:literalParams }}\n        </span>{{ literals?.yourUserWillBeBlocked }}\n      </p>\n      <br>\n      <p class=\"thf-font-text\">\n        {{ literals?.createANewPasswordNow }}\n      </p>\n      <div *ngIf=\"recovery\" class=\"thf-font-text-bold thf-mt-1 thf-page-login-popover-link-container\">\n        <a *ngIf=\"recoveryType === 'internalLink'\"\n          class=\"thf-page-login-popover-link\"\n          [routerLink]=\"recovery\">\n          {{ literals?.iForgotMyPassword }}\n        </a>\n        <a *ngIf=\"recoveryType === 'externalLink'\"\n          class=\"thf-page-login-popover-link\"\n          [href]=\"recovery\"\n          target=\"_blank\">\n          {{ literals?.iForgotMyPassword }}\n        </a>\n        <a *ngIf=\"!recoveryType\"\n          class=\"thf-page-login-popover-link\"\n          (click)=\"onForgotPasswordClick(recovery)\">\n          {{ literals?.iForgotMyPassword }}\n        </a>\n      </div>\n    </div>\n  </div>\n</div>\n"
                }] }
    ];
    /** @nocollapse */
    ThfPageLoginPopoverComponent.ctorParameters = function () { return [
        { type: ChangeDetectorRef }
    ]; };
    ThfPageLoginPopoverComponent.propDecorators = {
        recovery: [{ type: Input, args: ['t-recovery',] }],
        remainingAttempts: [{ type: Input, args: ['t-remaining-attempts',] }],
        selectedLanguage: [{ type: Input, args: ['t-selected-language',] }],
        forgotPassword: [{ type: Output, args: ['t-forgot-password',] }]
    };
    return ThfPageLoginPopoverComponent;
}());
export { ThfPageLoginPopoverComponent };
if (false) {
    /** @type {?} */
    ThfPageLoginPopoverComponent.prototype.literals;
    /** @type {?} */
    ThfPageLoginPopoverComponent.prototype.literalParams;
    /** @type {?} */
    ThfPageLoginPopoverComponent.prototype.recoveryType;
    /**
     * @type {?}
     * @private
     */
    ThfPageLoginPopoverComponent.prototype._recovery;
    /**
     * @type {?}
     * @private
     */
    ThfPageLoginPopoverComponent.prototype._remainingAttempts;
    /**
     * @type {?}
     * @private
     */
    ThfPageLoginPopoverComponent.prototype._selectedLanguage;
    /**
     * se 't-recovery' for do tipo Function ou ThfPageLoginRecovery, emite para o método 'openUrl' do componente 'thf-page-login'
     * @type {?}
     */
    ThfPageLoginPopoverComponent.prototype.forgotPassword;
    /**
     * @type {?}
     * @private
     */
    ThfPageLoginPopoverComponent.prototype.changeDetector;
}
