// @totvs/thf-ui/esm5/lib/components/thf-page/thf-page-list/thf-page-list.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, Renderer2, ViewChild, ViewContainerRef } from '@angular/core';
import { Router } from '@angular/router';
import { browserLanguage, callFunction, isTypeof } from '../../../utils/util';
import { ThfPageContentComponent } from '../thf-page-content/thf-page-content.component';
import { ThfPageListBaseComponent } from './thf-page-list-base.component';
/**
 * \@docsExtends ThfPageListBaseComponent
 *
 * \@example
 *
 * <example name="thf-page-list-basic" title="Totvs Page List Basic">
 *  <file name="sample-thf-page-list-basic/sample-thf-page-list-basic.component.html"> </file>
 *  <file name="sample-thf-page-list-basic/sample-thf-page-list-basic.component.ts"> </file>
 * </example>
 *
 * <example name="thf-page-list-labs" title="Totvs Page List Labs">
 *  <file name="sample-thf-page-list-labs/sample-thf-page-list-labs.component.html"> </file>
 *  <file name="sample-thf-page-list-labs/sample-thf-page-list-labs.component.ts"> </file>
 * </example>
 *
 * <example name="thf-page-list-hiring-processes" title="Totvs Page List - Hiring Processes">
 *  <file name="sample-thf-page-list-hiring-processes/sample-thf-page-list-hiring-processes.component.html"> </file>
 *  <file name="sample-thf-page-list-hiring-processes/sample-thf-page-list-hiring-processes.component.ts"> </file>
 *  <file name="sample-thf-page-list-hiring-processes/sample-thf-page-list-hiring-processes.service.ts"> </file>
 * </example>
 */
var ThfPageListComponent = /** @class */ (function (_super) {
    tslib_1.__extends(ThfPageListComponent, _super);
    function ThfPageListComponent(viewRef, renderer, router) {
        var _this = _super.call(this) || this;
        _this.renderer = renderer;
        _this.router = router;
        _this.limitPrimaryActions = 3;
        _this.isRecalculate = true;
        _this.maxWidthMobile = 480;
        _this.callFunction = callFunction;
        _this.parentRef = viewRef['_view']['component'];
        _this.initializeListeners();
        return _this;
    }
    /**
     * @return {?}
     */
    ThfPageListComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.advancedSearch = this.initFixedLiterals();
    };
    /**
     * @return {?}
     */
    ThfPageListComponent.prototype.ngAfterContentInit = /**
     * @return {?}
     */
    function () {
        this.setIsMobile();
        this.setDropdownActions();
    };
    /**
     * @param {?} changes
     * @return {?}
     */
    ThfPageListComponent.prototype.ngOnChanges = /**
     * @param {?} changes
     * @return {?}
     */
    function (changes) {
        this.setDropdownActions();
    };
    /**
     * @return {?}
     */
    ThfPageListComponent.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        this.removeListeners();
    };
    /**
     * @param {?} action
     * @return {?}
     */
    ThfPageListComponent.prototype.actionIsDisabled = /**
     * @param {?} action
     * @return {?}
     */
    function (action) {
        return isTypeof(action.disabled, 'function') ? action.disabled(action) : action.disabled;
    };
    /**
     * @param {?} item
     * @return {?}
     */
    ThfPageListComponent.prototype.callAction = /**
     * @param {?} item
     * @return {?}
     */
    function (item) {
        if (item.url) {
            this.router.navigate([item.url]);
        }
        else if (item.action) {
            callFunction(item.action, this.parentRef);
        }
    };
    /**
     * @return {?}
     */
    ThfPageListComponent.prototype.hasPageHeader = /**
     * @return {?}
     */
    function () {
        return !!(this.title || (this.actions && this.actions.length) || (this.breadcrumb && this.breadcrumb.items.length));
    };
    /**
     * @private
     * @param {?} event
     * @return {?}
     */
    ThfPageListComponent.prototype.onResize = /**
     * @private
     * @param {?} event
     * @return {?}
     */
    function (event) {
        /** @type {?} */
        var width = ((/** @type {?} */ (event.target))).innerWidth;
        if (width < this.maxWidthMobile) {
            this.isMobile = true;
            this.limitPrimaryActions = 2;
            this.setDropdownActions();
        }
        else {
            this.isMobile = false;
            this.limitPrimaryActions = 3;
            this.setDropdownActions();
        }
    };
    /**
     * @private
     * @return {?}
     */
    ThfPageListComponent.prototype.setIsMobile = /**
     * @private
     * @return {?}
     */
    function () {
        if (window.innerWidth < this.maxWidthMobile) {
            this.isMobile = true;
            this.limitPrimaryActions = 2;
        }
    };
    /**
     * @return {?}
     */
    ThfPageListComponent.prototype.setDropdownActions = /**
     * @return {?}
     */
    function () {
        if (this.actions.length > this.limitPrimaryActions) {
            this.dropdownActions = this.actions.slice(this.limitPrimaryActions - 1);
        }
    };
    /**
     * @param {?} field
     * @return {?}
     */
    ThfPageListComponent.prototype.callActionFilter = /**
     * @param {?} field
     * @return {?}
     */
    function (field) {
        this.callFunction(this.filter[field], this.parentRef);
    };
    /**
     * @return {?}
     */
    ThfPageListComponent.prototype.initFixedLiterals = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var locale = browserLanguage();
        /** @type {?} */
        var literal = {
            pt: {
                advancedSearch: 'Busca avançada'
            },
            en: {
                advancedSearch: 'Advanced search'
            },
            es: {
                advancedSearch: 'Búsqueda avanzada'
            }
        };
        return literal[locale].advancedSearch;
    };
    /**
     * @param {?} key
     * @return {?}
     */
    ThfPageListComponent.prototype.onkeypress = /**
     * @param {?} key
     * @return {?}
     */
    function (key) {
        if (key === 13) {
            this.callActionFilter('action');
        }
    };
    /**
     * @param {?} newModel
     * @return {?}
     */
    ThfPageListComponent.prototype.changeModel = /**
     * @param {?} newModel
     * @return {?}
     */
    function (newModel) {
        this.parentRef[this.filter.ngModel] = newModel;
    };
    // Recebe evento change do disclaimer e recalcula tela
    // Recebe evento change do disclaimer e recalcula tela
    /**
     * @param {?} disclaimers
     * @return {?}
     */
    ThfPageListComponent.prototype.onChangeDisclaimerGroup = 
    // Recebe evento change do disclaimer e recalcula tela
    /**
     * @param {?} disclaimers
     * @return {?}
     */
    function (disclaimers) {
        if ((disclaimers && disclaimers.length && this.isRecalculate) || (disclaimers.length === 0 && !this.isRecalculate)) {
            this.thfPageContent.recalculateHeaderSize();
            this.isRecalculate = !this.isRecalculate;
        }
        if (this.disclaimerGroup && this.disclaimerGroup.change) {
            this.disclaimerGroup.change(disclaimers);
        }
    };
    /**
     * @private
     * @return {?}
     */
    ThfPageListComponent.prototype.initializeListeners = /**
     * @private
     * @return {?}
     */
    function () {
        var _this = this;
        this.resizeListener = this.renderer.listen('window', 'resize', (/**
         * @param {?} event
         * @return {?}
         */
        function (event) {
            _this.onResize(event);
        }));
    };
    /**
     * @private
     * @return {?}
     */
    ThfPageListComponent.prototype.removeListeners = /**
     * @private
     * @return {?}
     */
    function () {
        this.resizeListener();
    };
    ThfPageListComponent.decorators = [
        { type: Component, args: [{
                    selector: 'thf-page-list',
                    template: "<thf-page>\n\n  <!-- HEADER -->\n  <thf-page-header\n    *ngIf=\"hasPageHeader()\"\n    [class.thf-page-list-header-padding]=\"filter && !actions.length\"\n    [t-breadcrumb]=\"breadcrumb\"\n    [t-title]=\"title\">\n\n    <!-- OPERATIONS -->\n    <div class=\"thf-page-list-operations\">\n      <div class=\"thf-page-list-actions\"\n        [class.thf-page-list-actions-padding]=\"filter\">\n        <thf-button *ngIf=\"actions[0]\"\n          t-type=\"primary\"\n          [t-disabled]=\"actionIsDisabled(actions[0])\"\n          [t-icon]=\"actions[0].icon\"\n          [t-label]=\"actions[0].label\"\n          (t-click)=\"callAction(actions[0])\">\n        </thf-button>\n\n        <thf-button *ngIf=\"actions[1] && (actions.length === 2 || !isMobile)\"\n          [t-disabled]=\"actionIsDisabled(actions[1])\"\n          [t-label]=\"actions[1].label\"\n          (t-click)=\"callAction(actions[1])\">\n        </thf-button>\n\n        <thf-button *ngIf=\"actions.length == 3 && actions[2] && !isMobile\"\n          [t-disabled]=\"actionIsDisabled(actions[2])\"\n          [t-label]=\"actions[2].label\"\n          (t-click)=\"callAction(actions[2])\">\n        </thf-button>\n\n        <thf-dropdown *ngIf=\"actions.length > limitPrimaryActions\"\n          [t-actions]=\"dropdownActions\"\n          [t-label]=\"literals.otherActions\">\n        </thf-dropdown>\n      </div>\n\n      <!-- FILTER -->\n      <div class=\"thf-page-list-filter-wrapper\" *ngIf=\"filter\">\n\n        <div class=\"thf-field-container-content thf-page-filter-content\">\n          <div class=\"thf-field-icon-container-right\">\n            <span class=\"thf-icon thf-icon-search thf-field-icon\"\n              (click)=\"callActionFilter('action')\">\n            </span>\n          </div>\n\n          <input class=\"thf-input thf-input-icon-right\"\n            name=\"model\"\n            type=\"text\"\n            [ngModel]=\"parentRef[filter.ngModel]\"\n            [placeholder]=\"filter.placeholder || ''\"\n            (keypress)=\"onkeypress($event.keyCode)\"\n            (ngModelChange)=\"changeModel($event)\">\n        </div>\n\n        <div class=\"thf-page-list-filter-search\" *ngIf=\"filter.advancedAction\">\n          <span class=\"thf-page-list-filter-search-link\"\n            tabindex=\"0\"\n            (click)=\"callActionFilter('advancedAction')\"\n            (keydown.enter)=\"callActionFilter('advancedAction')\">\n              {{advancedSearch}}\n          </span>\n        </div>\n\n      </div>\n    </div>\n\n    <!-- DISCLAIMER -->\n    <thf-disclaimer-group *ngIf=\"!!disclaimerGroup\"\n      [class.thf-page-list-disclaimer-group]=\"!!disclaimerGroup?.disclaimers?.length\"\n      [t-disclaimers]=\"disclaimerGroup?.disclaimers\"\n      [t-hide-remove-all]=\"disclaimerGroup?.hideRemoveAll\"\n      [t-title]=\"disclaimerGroup?.title\"\n      (t-change)=\"onChangeDisclaimerGroup($event)\">\n    </thf-disclaimer-group>\n\n  </thf-page-header>\n\n  <!-- CONTENT -->\n  <thf-page-content>\n    <ng-content></ng-content>\n  </thf-page-content>\n\n</thf-page>\n"
                }] }
    ];
    /** @nocollapse */
    ThfPageListComponent.ctorParameters = function () { return [
        { type: ViewContainerRef },
        { type: Renderer2 },
        { type: Router }
    ]; };
    ThfPageListComponent.propDecorators = {
        thfPageContent: [{ type: ViewChild, args: [ThfPageContentComponent, { static: true },] }]
    };
    return ThfPageListComponent;
}(ThfPageListBaseComponent));
export { ThfPageListComponent };
if (false) {
    /** @type {?} */
    ThfPageListComponent.prototype.advancedSearch;
    /** @type {?} */
    ThfPageListComponent.prototype.dropdownActions;
    /** @type {?} */
    ThfPageListComponent.prototype.isMobile;
    /** @type {?} */
    ThfPageListComponent.prototype.limitPrimaryActions;
    /**
     * @type {?}
     * @private
     */
    ThfPageListComponent.prototype.isRecalculate;
    /**
     * @type {?}
     * @private
     */
    ThfPageListComponent.prototype.maxWidthMobile;
    /**
     * @type {?}
     * @private
     */
    ThfPageListComponent.prototype.parentRef;
    /** @type {?} */
    ThfPageListComponent.prototype.callFunction;
    /** @type {?} */
    ThfPageListComponent.prototype.thfPageContent;
    /** @type {?} */
    ThfPageListComponent.prototype.renderer;
    /**
     * @type {?}
     * @private
     */
    ThfPageListComponent.prototype.router;
}
