// @totvs/thf-ui/esm5/lib/components/thf-navbar/thf-navbar-items/thf-navbar-items.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ElementRef, Input, QueryList, ViewChild, ViewChildren } from '@angular/core';
import { NavigationCancel, NavigationEnd, Router } from '@angular/router';
import { ThfNavbarItemComponent } from './thf-navbar-item/thf-navbar-item.component';
var ThfNavbarItemsComponent = /** @class */ (function () {
    function ThfNavbarItemsComponent(router) {
        this.router = router;
    }
    /**
     * @return {?}
     */
    ThfNavbarItemsComponent.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        this.routeSubscription.unsubscribe();
    };
    /**
     * @return {?}
     */
    ThfNavbarItemsComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.subscribeToRoute();
    };
    /**
     * @param {?} item
     * @return {?}
     */
    ThfNavbarItemsComponent.prototype.selectItem = /**
     * @param {?} item
     * @return {?}
     */
    function (item) {
        this.selectedItem = item;
    };
    /**
     * @private
     * @param {?} urlRouter
     * @return {?}
     */
    ThfNavbarItemsComponent.prototype.checkActiveItemByUrl = /**
     * @private
     * @param {?} urlRouter
     * @return {?}
     */
    function (urlRouter) {
        this.selectedItem = this.items.find((/**
         * @param {?} item
         * @return {?}
         */
        function (item) { return item.link === urlRouter; }));
    };
    /**
     * @private
     * @return {?}
     */
    ThfNavbarItemsComponent.prototype.checkRouterChildrenFragments = /**
     * @private
     * @return {?}
     */
    function () {
        /** @type {?} */
        var childrenPrimary = this.router.parseUrl(this.router.url).root.children['primary'];
        return childrenPrimary ? "/" + childrenPrimary.segments.map((/**
         * @param {?} it
         * @return {?}
         */
        function (it) { return it.path; })).join('/') : '';
    };
    /**
     * @private
     * @return {?}
     */
    ThfNavbarItemsComponent.prototype.subscribeToRoute = /**
     * @private
     * @return {?}
     */
    function () {
        var _this = this;
        this.routeSubscription = this.router.events.subscribe((/**
         * @param {?} val
         * @return {?}
         */
        function (val) {
            if (val instanceof NavigationEnd || val instanceof NavigationCancel) {
                /** @type {?} */
                var urlRouter = _this.checkRouterChildrenFragments();
                _this.checkActiveItemByUrl(urlRouter);
            }
        }));
    };
    ThfNavbarItemsComponent.decorators = [
        { type: Component, args: [{
                    selector: 'thf-navbar-items',
                    template: "<nav>\n  <ul class=\"thf-navbar-items-container\" #navbarItemsContainer>\n    <li *ngFor=\"let item of items\"\n      class=\"thf-navbar-item\"\n      [class.thf-navbar-item-selected]=\"selectedItem === item\">\n      <thf-navbar-item\n        [t-action]=\"item.action\"\n        [t-clickable]=\"selectedItem !== item\"\n        [t-label]=\"item.label\"\n        [t-link]=\"item.link\"\n        (t-click)=\"selectItem(item)\">\n      </thf-navbar-item>\n    </li>\n  </ul>\n</nav>\n"
                }] }
    ];
    /** @nocollapse */
    ThfNavbarItemsComponent.ctorParameters = function () { return [
        { type: Router }
    ]; };
    ThfNavbarItemsComponent.propDecorators = {
        navbarItemsContainer: [{ type: ViewChild, args: ['navbarItemsContainer', { read: ElementRef, static: true },] }],
        allNavbarItems: [{ type: ViewChildren, args: [ThfNavbarItemComponent, { read: ElementRef },] }],
        items: [{ type: Input, args: ['t-items',] }]
    };
    return ThfNavbarItemsComponent;
}());
export { ThfNavbarItemsComponent };
if (false) {
    /** @type {?} */
    ThfNavbarItemsComponent.prototype.selectedItem;
    /**
     * @type {?}
     * @private
     */
    ThfNavbarItemsComponent.prototype.routeSubscription;
    /** @type {?} */
    ThfNavbarItemsComponent.prototype.navbarItemsContainer;
    /** @type {?} */
    ThfNavbarItemsComponent.prototype.allNavbarItems;
    /** @type {?} */
    ThfNavbarItemsComponent.prototype.items;
    /**
     * @type {?}
     * @private
     */
    ThfNavbarItemsComponent.prototype.router;
}
