// @totvs/thf-ui/esm2015/lib/components/thf-field/thf-url/thf-url.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ElementRef, forwardRef } from '@angular/core';
import { NG_VALUE_ACCESSOR, NG_VALIDATORS } from '@angular/forms';
import { ThfInputGeneric } from '../thf-input-generic/thf-input-generic';
/**
 * \@docsExtends ThfInputBaseComponent
 *
 * \@description
 *
 * thf-url é um input específico para receber URL, com o pattern já configurado.
 *
 * \@example
 *
 * <example name="thf-url-basic" title="Totvs Url Basic">
 *   <file name="sample-thf-url-basic/sample-thf-url-basic.component.html"> </file>
 *   <file name="sample-thf-url-basic/sample-thf-url-basic.component.ts"> </file>
 * </example>
 *
 * <example name="thf-url-labs" title="Totvs Url Labs">
 *   <file name="sample-thf-url-labs/sample-thf-url-labs.component.html"> </file>
 *   <file name="sample-thf-url-labs/sample-thf-url-labs.component.ts"> </file>
 * </example>
 *
 * <example name="thf-url-shortener" title="Totvs Url - Shortener">
 *   <file name="sample-thf-url-shortener/sample-thf-url-shortener.component.html"> </file>
 *   <file name="sample-thf-url-shortener/sample-thf-url-shortener.component.ts"> </file>
 * </example>
 *
 */
export class ThfUrlComponent extends ThfInputGeneric {
    /**
     * @param {?} el
     */
    constructor(el) {
        super(el);
        this.icon = 'thf-icon-world';
        this.maxlength = 254;
        this.type = 'url';
        this.pattern = '^((https|http):\\/\\/)?' + // protocol
            '((([a-z\\d]([a-z\\d-]*[a-z\\d])*)\\.)+[a-z]{2,}|' + // domain name
            '((\\d{1,3}\\.){3}\\d{1,3}))' + // OR ip (v4) address
            '(\\:\\d+)?(\\/[-a-z\\d%_.~+]*)*' + // port and path
            '(\\?[;&a-z\\d%_.~+=-]*)?' + // query string
            '(\\#[-a-z\\d_]*)?$';
        this.mask = '';
        this.listener = this.validateClassesForPattern.bind(this);
    }
    /**
     * @return {?}
     */
    ngAfterViewInit() {
        // Se não tem ngModel ou reactive form adiciona validação com classes css
        setTimeout((/**
         * @return {?}
         */
        () => {
            if (!this.onChangePropagate) {
                this.inputEl.nativeElement.addEventListener('keyup', this.listener);
            }
        }));
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        if (!this.onChangePropagate) {
            this.inputEl.nativeElement.removeEventListener('keyup', this.listener);
        }
    }
    /**
     * @param {?} c
     * @return {?}
     */
    extraValidation(c) {
        return null;
    }
}
ThfUrlComponent.decorators = [
    { type: Component, args: [{
                selector: 'thf-url',
                template: "<thf-field-container\n  [t-help]=\"help\"\n  [t-label]=\"label\"\n  [t-optional]=\"!required && optional\">\n\n  <div class=\"thf-field-container-content\">\n    <div *ngIf=\"icon\" class=\"thf-field-icon-container-left\">\n      <span class=\"thf-icon thf-field-icon {{ icon }}\" [class.thf-field-icon-disabled]=\"disabled\"></span>\n    </div>\n\n    <input #inp\n      class=\"thf-input\"\n      [attr.name]=\"name\"\n      [class.thf-input-icon-left]=\"icon\"\n      [class.thf-input-icon-right]=\"clean\"\n      [disabled]=\"disabled\"\n      [placeholder]=\"placeholder\"\n      [readonly]=\"readonly\"\n      [required]=\"required\"\n      [type]=\"type\"\n      (blur)=\"eventOnBlur($event)\"\n      (click)=\"eventOnClick($event)\"\n      (focus)=\"eventOnFocus($event)\"\n      (input)=\"eventOnInput($event)\">\n\n    <div class=\"thf-field-icon-container-right\">\n      <thf-clean [t-element-ref]=\"inputEl\" (t-change-event)=\"clear($event)\"></thf-clean>\n    </div>\n  </div>\n\n  <thf-field-container-bottom\n    [t-error-pattern]=\"getErrorPattern()\">\n  </thf-field-container-bottom>\n\n</thf-field-container>\n",
                providers: [
                    {
                        provide: NG_VALUE_ACCESSOR,
                        useExisting: forwardRef((/**
                         * @return {?}
                         */
                        () => ThfUrlComponent)),
                        multi: true
                    },
                    {
                        provide: NG_VALIDATORS,
                        useExisting: forwardRef((/**
                         * @return {?}
                         */
                        () => ThfUrlComponent)),
                        multi: true
                    }
                ]
            }] }
];
/** @nocollapse */
ThfUrlComponent.ctorParameters = () => [
    { type: ElementRef }
];
if (false) {
    /** @type {?} */
    ThfUrlComponent.prototype.icon;
    /** @type {?} */
    ThfUrlComponent.prototype.maxlength;
    /** @type {?} */
    ThfUrlComponent.prototype.type;
    /** @type {?} */
    ThfUrlComponent.prototype.pattern;
    /** @type {?} */
    ThfUrlComponent.prototype.mask;
    /**
     * @type {?}
     * @private
     */
    ThfUrlComponent.prototype.listener;
}
