// @totvs/thf-ui/esm5/lib/interceptors/interceptors.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { ThfHttpInterceptorModule } from './thf-http-interceptor/thf-http-interceptor.module';
import { ThfHttpRequestModule } from './thf-http-request/thf-http-request.module';
var ThfInterceptorsModule = /** @class */ (function () {
    function ThfInterceptorsModule() {
    }
    ThfInterceptorsModule.decorators = [
        { type: NgModule, args: [{
                    declarations: [],
                    imports: [
                        ThfHttpInterceptorModule,
                        ThfHttpRequestModule
                    ],
                    exports: [
                        ThfHttpInterceptorModule,
                        ThfHttpRequestModule
                    ],
                    providers: [],
                    bootstrap: []
                },] }
    ];
    return ThfInterceptorsModule;
}());
export { ThfInterceptorsModule };
