// @totvs/thf-ui/esm5/lib/components/thf-field/thf-rich-text/thf-rich-text-toolbar/thf-rich-text-toolbar.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, ElementRef, EventEmitter, Output, ViewChild, Input } from '@angular/core';
import { ThfLanguageService } from '../../../../services/thf-language/thf-language.service';
import { thfRichTextLiteralsDefault } from '../thf-rich-text-literals';
var ThfRichTextToolbarComponent = /** @class */ (function () {
    function ThfRichTextToolbarComponent(languageService) {
        this.languageService = languageService;
        this.literals = tslib_1.__assign({}, thfRichTextLiteralsDefault[this.languageService.getShortLanguage()]);
        this.alignButtons = [
            {
                command: 'justifyleft',
                icon: 'thf-icon-align-left',
                tooltip: this.literals.left,
                action: this.emitAlignCommand.bind(this, 'justifyleft'),
                selected: true
            },
            {
                command: 'justifycenter',
                icon: 'thf-icon-align-center',
                tooltip: this.literals.center,
                action: this.emitAlignCommand.bind(this, 'justifycenter')
            },
            {
                command: 'justifyright',
                icon: 'thf-icon-align-right',
                tooltip: this.literals.right,
                action: this.emitAlignCommand.bind(this, 'justifyright')
            },
            {
                command: 'justifyfull',
                icon: 'thf-icon-align-justify',
                tooltip: this.literals.justify,
                action: this.emitAlignCommand.bind(this, 'justifyfull')
            }
        ];
        this.formatButtons = [
            {
                command: 'bold',
                icon: 'thf-icon-text-bold',
                tooltip: this.literals.bold,
                action: this.emitCommand.bind(this, 'bold')
            },
            {
                command: 'italic',
                icon: 'thf-icon-text-italic',
                tooltip: this.literals.italic,
                action: this.emitCommand.bind(this, 'italic')
            },
            {
                command: 'underline',
                icon: 'thf-icon-text-underline',
                tooltip: this.literals.underline,
                action: this.emitCommand.bind(this, 'underline')
            }
        ];
        this.listButtons = [
            {
                command: 'insertUnorderedList',
                icon: 'thf-icon-list',
                tooltip: this.literals.unorderedList,
                action: this.emitCommand.bind(this, 'insertUnorderedList')
            }
        ];
        this.command = new EventEmitter();
    }
    Object.defineProperty(ThfRichTextToolbarComponent.prototype, "readonly", {
        get: /**
         * @return {?}
         */
        function () {
            return this._readonly;
        },
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._readonly = value;
            this.toggleDisableButtons(this._readonly);
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    ThfRichTextToolbarComponent.prototype.ngAfterViewInit = /**
     * @return {?}
     */
    function () {
        this.removeButtonFocus();
    };
    /**
     * @param {?} commands
     * @return {?}
     */
    ThfRichTextToolbarComponent.prototype.setButtonsStates = /**
     * @param {?} commands
     * @return {?}
     */
    function (commands) {
        if (!this.readonly) {
            this.alignButtons.forEach((/**
             * @param {?} button
             * @return {?}
             */
            function (button) { button.selected = commands.includes(button.command); }));
            this.formatButtons.forEach((/**
             * @param {?} button
             * @return {?}
             */
            function (button) { button.selected = commands.includes(button.command); }));
            this.listButtons[0].selected = commands.includes(this.listButtons[0].command);
        }
    };
    /**
     * @private
     * @param {?} command
     * @return {?}
     */
    ThfRichTextToolbarComponent.prototype.emitAlignCommand = /**
     * @private
     * @param {?} command
     * @return {?}
     */
    function (command) {
        /** @type {?} */
        var index = this.alignButtons.findIndex((/**
         * @param {?} btn
         * @return {?}
         */
        function (btn) { return btn.command === command; }));
        if (this.alignButtons[index].selected) {
            this.alignButtons[index].selected = false;
        }
        this.command.emit(command);
    };
    /**
     * @private
     * @param {?} command
     * @return {?}
     */
    ThfRichTextToolbarComponent.prototype.emitCommand = /**
     * @private
     * @param {?} command
     * @return {?}
     */
    function (command) {
        this.command.emit(command);
    };
    /**
     * @private
     * @return {?}
     */
    ThfRichTextToolbarComponent.prototype.removeButtonFocus = /**
     * @private
     * @return {?}
     */
    function () {
        /** @type {?} */
        var buttons = this.toolbarElement.nativeElement.querySelectorAll('button');
        buttons.forEach((/**
         * @param {?} button
         * @return {?}
         */
        function (button) { button.setAttribute('tabindex', '-1'); }));
    };
    /**
     * @private
     * @param {?} state
     * @return {?}
     */
    ThfRichTextToolbarComponent.prototype.toggleDisableButtons = /**
     * @private
     * @param {?} state
     * @return {?}
     */
    function (state) {
        this.alignButtons.forEach((/**
         * @param {?} button
         * @return {?}
         */
        function (button) { button.disabled = state; }));
        this.formatButtons.forEach((/**
         * @param {?} button
         * @return {?}
         */
        function (button) { button.disabled = state; }));
        this.listButtons[0].disabled = state;
    };
    ThfRichTextToolbarComponent.decorators = [
        { type: Component, args: [{
                    selector: 'thf-rich-text-toolbar',
                    template: "<div class=\"thf-rich-text-toolbar\" #toolbarElement>\n  <div class=\"thf-rich-text-toolbar-button-align\">\n    <thf-button-group [t-buttons]=\"formatButtons\" t-toggle=\"multiple\"></thf-button-group>\n  </div>\n\n  <div class=\"thf-rich-text-toolbar-button-align\">\n    <thf-button-group [t-buttons]=\"alignButtons\" t-toggle=\"single\"></thf-button-group>\n  </div>\n\n  <div class=\"thf-rich-text-toolbar-button-align\">\n    <thf-button-group [t-buttons]=\"listButtons\" t-toggle=\"single\"></thf-button-group>\n  </div>\n</div>\n"
                }] }
    ];
    /** @nocollapse */
    ThfRichTextToolbarComponent.ctorParameters = function () { return [
        { type: ThfLanguageService }
    ]; };
    ThfRichTextToolbarComponent.propDecorators = {
        toolbarElement: [{ type: ViewChild, args: ['toolbarElement', { static: true },] }],
        readonly: [{ type: Input, args: ['t-readonly',] }],
        command: [{ type: Output, args: ['t-command',] }]
    };
    return ThfRichTextToolbarComponent;
}());
export { ThfRichTextToolbarComponent };
if (false) {
    /**
     * @type {?}
     * @private
     */
    ThfRichTextToolbarComponent.prototype._readonly;
    /** @type {?} */
    ThfRichTextToolbarComponent.prototype.literals;
    /** @type {?} */
    ThfRichTextToolbarComponent.prototype.alignButtons;
    /** @type {?} */
    ThfRichTextToolbarComponent.prototype.formatButtons;
    /** @type {?} */
    ThfRichTextToolbarComponent.prototype.listButtons;
    /** @type {?} */
    ThfRichTextToolbarComponent.prototype.toolbarElement;
    /** @type {?} */
    ThfRichTextToolbarComponent.prototype.command;
    /**
     * @type {?}
     * @private
     */
    ThfRichTextToolbarComponent.prototype.languageService;
}
