// @totvs/thf-ui/esm5/lib/components/thf-tabs/thf-tab-dropdown/thf-tab-dropdown.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ElementRef, EventEmitter, Input, Output, ViewChild } from '@angular/core';
import { ThfPopoverComponent } from '../../thf-popover/thf-popover.component';
/**
 * \@docsPrivate
 *
 * \@description
 *
 * Componente para agrupamento de componentes `thf-tab-button` que os rendereiza em uma um caixa de diálogo
 * no formato de lista.
 */
var ThfTabDropdownComponent = /** @class */ (function () {
    function ThfTabDropdownComponent() {
        // Lista de abas
        this.tabs = [];
        // Evento que será emitido ao ativar uma aba
        this.activated = new EventEmitter();
        // Evento que será emitido a aba for desabilitada ou ocultada
        this.changeState = new EventEmitter();
        // Evento de click
        this.click = new EventEmitter();
    }
    /**
     * @return {?}
     */
    ThfTabDropdownComponent.prototype.getContainerClass = /**
     * @return {?}
     */
    function () {
        return this.small ? 'thf-tab-dropdown-container-sm' : 'thf-tab-dropdown-container';
    };
    /**
     * @return {?}
     */
    ThfTabDropdownComponent.prototype.togglePopover = /**
     * @return {?}
     */
    function () {
        this.popover.isHidden ? this.popover.open() : this.popover.close();
    };
    ThfTabDropdownComponent.decorators = [
        { type: Component, args: [{
                    selector: 'thf-tab-dropdown',
                    template: "<div #tabDrop\n  class=\"thf-tab-dropdown-content thf-tab-button-md\"\n  [attr.tabindex]=\"0\"\n  [class.thf-tab-button-sm]=\"small\"\n  (keyup.enter)=\"togglePopover()\">\n  <div class=\"thf-tab-button-label\">\n    {{ label }}\n    <span class=\"thf-icon\" [ngClass]=\"(popover.isHidden ? 'thf-icon-arrow-down' : 'thf-icon-arrow-up')\">\n    </span>\n  </div>\n</div>\n\n<thf-popover #popover\n  t-hide-arrow\n  t-position=\"bottom\"\n  [t-target]=\"tabDropdown\">\n\n  <div [ngClass]=\"getContainerClass()\">\n    <thf-tab-button *ngFor=\"let tab of tabs\"\n      class=\"thf-tab-dropdown-button\"\n      [class.thf-tab-dropdown-button-active]=\"tab.active\"\n      [t-active]=\"tab.active\"\n      [t-disabled]=\"tab.disabled\"\n      [t-hide]=\"tab.hide\"\n      [t-id]=\"tab.id\"\n      [t-label]=\"tab.label\"\n      [t-small]=\"small\"\n      (t-activated)=\"activated.emit(tab)\"\n      (t-change-state)=\"changeState.emit(tab)\"\n      (t-click)=\"popover.close();click.emit(tab);\">\n    </thf-tab-button>\n  </div>\n\n</thf-popover>\n"
                }] }
    ];
    ThfTabDropdownComponent.propDecorators = {
        popover: [{ type: ViewChild, args: ['popover', { static: true },] }],
        tabDropdown: [{ type: ViewChild, args: ['tabDrop', { read: ElementRef, static: true },] }],
        label: [{ type: Input, args: ['t-label',] }],
        small: [{ type: Input, args: ['t-small',] }],
        tabs: [{ type: Input, args: ['t-tabs',] }],
        activated: [{ type: Output, args: ['t-activated',] }],
        changeState: [{ type: Output, args: ['t-change-state',] }],
        click: [{ type: Output, args: ['t-click',] }]
    };
    return ThfTabDropdownComponent;
}());
export { ThfTabDropdownComponent };
if (false) {
    /** @type {?} */
    ThfTabDropdownComponent.prototype.popover;
    /** @type {?} */
    ThfTabDropdownComponent.prototype.tabDropdown;
    /** @type {?} */
    ThfTabDropdownComponent.prototype.label;
    /** @type {?} */
    ThfTabDropdownComponent.prototype.small;
    /** @type {?} */
    ThfTabDropdownComponent.prototype.tabs;
    /** @type {?} */
    ThfTabDropdownComponent.prototype.activated;
    /** @type {?} */
    ThfTabDropdownComponent.prototype.changeState;
    /** @type {?} */
    ThfTabDropdownComponent.prototype.click;
}
