// @totvs/thf-ui/esm2015/lib/components/thf-breadcrumb/thf-breadcrumb-item/thf-breadcrumb-item.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input } from '@angular/core';
/**
 * \@docsPrivate
 *
 * \@description
 *
 * Componente que renderiza cada item do thf-breadcrumb.
 */
export class ThfBreadcrumbItemComponent {
    constructor() {
        // Especifica se item é o link ativo.
        this.itemActive = false;
    }
}
ThfBreadcrumbItemComponent.decorators = [
    { type: Component, args: [{
                selector: 'thf-breadcrumb-item',
                template: "<li *ngIf=\"!itemActive\"\n  [class.thf-breadcrumb-item]=\"link || action\"\n  [class.thf-breadcrumb-item-unclickable]=\"!link && !action\">\n\n  <a *ngIf=\"link\" [routerLink]=\"link\">\n    <label class=\"thf-breadcrumb-label\">{{ label }}</label>\n    <div class=\"thf-breadcrumb-arrow\"></div>\n  </a>\n\n  <a *ngIf=\"!link && action\" (click)=\"action(label)\">\n    <label class=\"thf-breadcrumb-label\">{{ label }}</label>\n    <div class=\"thf-breadcrumb-arrow\"></div>\n  </a>\n\n  <a *ngIf=\"!link && !action\">\n    <label class=\"thf-breadcrumb-label\">{{ label }}</label>\n    <div class=\"thf-breadcrumb-arrow\"></div>\n  </a>\n</li>\n\n<li *ngIf=\"itemActive\" class=\"thf-breadcrumb-item-unclickable\">\n  <label class=\"thf-breadcrumb-label\">{{ label }}</label>\n</li>\n"
            }] }
];
ThfBreadcrumbItemComponent.propDecorators = {
    action: [{ type: Input, args: ['t-action',] }],
    label: [{ type: Input, args: ['t-label',] }],
    link: [{ type: Input, args: ['t-link',] }],
    itemActive: [{ type: Input, args: ['t-item-active',] }]
};
if (false) {
    /** @type {?} */
    ThfBreadcrumbItemComponent.prototype.action;
    /** @type {?} */
    ThfBreadcrumbItemComponent.prototype.label;
    /** @type {?} */
    ThfBreadcrumbItemComponent.prototype.link;
    /** @type {?} */
    ThfBreadcrumbItemComponent.prototype.itemActive;
}
