// @totvs/thf-templates/esm2015/lib/components/thf-page-dynamic-search/thf-page-dynamic-search.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ViewChild } from '@angular/core';
import { ThfDynamicFieldType } from '@totvs/thf-ui';
import { getBrowserLanguage } from '../../utils/util';
import { ThfAdvancedFilterComponent } from './thf-advanced-filter/thf-advanced-filter.component';
import { ThfPageDynamicSearchBaseComponent } from './thf-page-dynamic-search-base.component';
/**
 * \@docsExtends ThfPageDynamicSearchBaseComponent
 *
 * \@example
 *
 * <example name="thf-page-dynamic-search-basic" title="Totvs Page Dynamic Search Basic">
 *  <file name="sample-thf-page-dynamic-search-basic/sample-thf-page-dynamic-search-basic.component.html"> </file>
 *  <file name="sample-thf-page-dynamic-search-basic/sample-thf-page-dynamic-search-basic.component.ts"> </file>
 * </example>
 *
 * <example name="thf-page-dynamic-search-hiring-processes" title="Totvs Page Dynamic Search - Hiring processes">
 *  <file name="sample-thf-page-dynamic-search-hiring-processes/sample-thf-page-dynamic-search-hiring-processes.component.html"> </file>
 *  <file name="sample-thf-page-dynamic-search-hiring-processes/sample-thf-page-dynamic-search-hiring-processes.component.ts"> </file>
 *  <file name="sample-thf-page-dynamic-search-hiring-processes/sample-thf-page-dynamic-search-hiring-processes.service.ts"> </file>
 * </example>
 */
export class ThfPageDynamicSearchComponent extends ThfPageDynamicSearchBaseComponent {
    constructor() {
        super(...arguments);
        this._disclaimerGroup = {
            change: this.onChangeDisclaimerGroup.bind(this),
            disclaimers: [],
            title: this.literals.disclaimerGroupTitle
        };
        this._filterSettings = {
            action: 'onAction',
            advancedAction: 'onAdvancedAction',
            ngModel: 'quickFilter',
            placeholder: this.literals.filterSettingsPlaceholder
        };
        // Flag to control when changeDisclaimerGroup should be called
        this.changeDisclaimersEnabled = false;
    }
    /**
     * @return {?}
     */
    get disclaimerGroup() {
        return Object.assign({}, this._disclaimerGroup);
    }
    /**
     * @return {?}
     */
    get filterSettings() {
        this._filterSettings.advancedAction = this.filters.length === 0 ? undefined : 'onAdvancedAction';
        return Object.assign({}, this._filterSettings);
    }
    /**
     * @return {?}
     */
    onAction() {
        this.changeDisclaimersEnabled = false;
        this._disclaimerGroup.disclaimers = [
            { property: 'search', label: `${this.literals.quickSearchLabel} ${this.quickFilter}`, value: this.quickFilter }
        ];
        if (this.quickSearch.observers && this.quickSearch.observers.length > 0) {
            this.quickSearch.emit(this.quickFilter);
        }
        this.quickFilter = undefined;
    }
    /**
     * @return {?}
     */
    onAdvancedAction() {
        this.thfAdvancedFilter.open();
    }
    /**
     * @param {?} filters
     * @return {?}
     */
    onAdvancedSearch(filters) {
        this.changeDisclaimersEnabled = false;
        this._disclaimerGroup.disclaimers = this.setDisclaimers(filters);
        this.advancedSearch.emit(filters);
    }
    /**
     * @private
     * @param {?} date
     * @return {?}
     */
    formatDate(date) {
        /** @type {?} */
        const year = parseInt(date.substr(0, 4), 10);
        /** @type {?} */
        const month = parseInt(date.substr(5, 2), 10);
        /** @type {?} */
        const day = parseInt(date.substr(8, 2), 10);
        return new Date(year, month - 1, day).toLocaleDateString(getBrowserLanguage());
    }
    /**
     * @private
     * @param {?} fields
     * @param {?} fieldName
     * @return {?}
     */
    getFieldByProperty(fields, fieldName) {
        return fields.find((/**
         * @param {?} field
         * @return {?}
         */
        (field) => field.property === fieldName));
    }
    /**
     * @private
     * @param {?} disclaimers
     * @return {?}
     */
    onChangeDisclaimerGroup(disclaimers) {
        this.changeDisclaimersEnabled ? this.changeDisclaimers.emit(disclaimers) : this.changeDisclaimersEnabled = true;
    }
    /**
     * @private
     * @param {?} filters
     * @return {?}
     */
    setDisclaimers(filters) {
        /** @type {?} */
        const disclaimers = [];
        Object.keys(filters).forEach((/**
         * @param {?} filter
         * @return {?}
         */
        filter => {
            /** @type {?} */
            const field = this.getFieldByProperty(this.filters, filter);
            /** @type {?} */
            const value = field.type === ThfDynamicFieldType.Date ? this.formatDate(filters[filter]) : filters[filter];
            disclaimers.push({
                label: `${field.label}: ${value}`,
                property: filter,
                value: filters[filter]
            });
        }));
        return disclaimers;
    }
}
ThfPageDynamicSearchComponent.decorators = [
    { type: Component, args: [{
                selector: 'thf-page-dynamic-search',
                template: "<thf-page-list\n  [t-actions]=\"actions\"\n  [t-breadcrumb]=\"breadcrumb\"\n  [t-disclaimer-group]=\"disclaimerGroup\"\n  [t-filter]=\"filterSettings\"\n  [t-title]=\"title\">\n\n  <thf-advanced-filter\n    [t-filters]=\"filters\"\n    (t-search-event)=\"onAdvancedSearch($event)\">\n  </thf-advanced-filter>\n\n  <ng-content></ng-content>\n\n</thf-page-list>\n"
            }] }
];
ThfPageDynamicSearchComponent.propDecorators = {
    thfAdvancedFilter: [{ type: ViewChild, args: [ThfAdvancedFilterComponent, { static: true },] }]
};
if (false) {
    /**
     * @type {?}
     * @private
     */
    ThfPageDynamicSearchComponent.prototype._disclaimerGroup;
    /**
     * @type {?}
     * @private
     */
    ThfPageDynamicSearchComponent.prototype._filterSettings;
    /**
     * @type {?}
     * @private
     */
    ThfPageDynamicSearchComponent.prototype.changeDisclaimersEnabled;
    /**
     * @type {?}
     * @private
     */
    ThfPageDynamicSearchComponent.prototype.quickFilter;
    /** @type {?} */
    ThfPageDynamicSearchComponent.prototype.thfAdvancedFilter;
}
