// @totvs/thf-templates/esm2015/lib/components/thf-page-blocked-user/thf-page-blocked-user-contacts/thf-page-blocked-user-contacts.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ChangeDetectorRef, Component, ElementRef, Input, ViewChild } from '@angular/core';
/** @type {?} */
const thfPageBlockedUserContactItemMargin = 16;
export class ThfPageBlockedUserContactsComponent {
    /**
     * @param {?} changeDetector
     */
    constructor(changeDetector) {
        this.changeDetector = changeDetector;
        this.overflowItem = false;
    }
    /**
     * @return {?}
     */
    ngAfterViewInit() {
        this.checkContactItemWidth();
    }
    /**
     * @param {?} changes
     * @return {?}
     */
    ngOnChanges(changes) {
        if (changes.email || changes.phone) {
            this.checkContactItemWidth();
        }
    }
    /**
     * @private
     * @return {?}
     */
    checkContactItemWidth() {
        if (!this.email || !this.phone) {
            this.overflowItem = true;
            return;
        }
        else {
            setTimeout((/**
             * @return {?}
             */
            () => {
                /** @type {?} */
                const phoneWidth = this.phoneItem.nativeElement.offsetWidth;
                /** @type {?} */
                const mailWidth = this.mailItem.nativeElement.offsetWidth;
                /** @type {?} */
                const contactGroupHalfWidth = this.contactGroup.nativeElement.offsetWidth / 2;
                this.overflowItem = phoneWidth > contactGroupHalfWidth || mailWidth > contactGroupHalfWidth - thfPageBlockedUserContactItemMargin;
            }));
        }
        this.changeDetector.detectChanges();
    }
}
ThfPageBlockedUserContactsComponent.decorators = [
    { type: Component, args: [{
                selector: 'thf-page-blocked-user-contacts',
                template: "<div #contactGroup class=\"thf-page-blocked-user-contact-group thf-row thf-mb-3\">\n  <div *ngIf=\"phone\"\n    class=\"thf-page-blocked-user-contact-item\"\n    [ngClass]=\"overflowItem ? 'thf-md-12' : 'thf-md-6'\">\n    <a\n      class=\"thf-page-blocked-user-link thf-clickable\"\n      [class.thf-page-blocked-user-contact-content-inline]=\"overflowItem\"\n      href=\"tel:{{ phone }}\"\n      target=\"_self\">\n      <div class=\"thf-page-blocked-user-contact-group-item\">\n        <span class=\"thf-page-blocked-user-contact-icon thf-icon thf-icon-telephone thf-pr-1\"></span>\n        <span #phoneItem class=\"thf-page-blocked-user-contact-text thf-font-text\">{{ phone }}</span>\n      </div>\n    </a>\n  </div>\n  <div *ngIf=\"email\"\n    class=\"thf-page-blocked-user-contact-item thf-page-blocked-user-contact-mail\"\n    [ngClass]=\"overflowItem ? 'thf-md-12' : 'thf-md-6'\">\n    <a\n      class=\"thf-page-blocked-user-link thf-clickable\"\n      [class.thf-page-blocked-user-contact-content-inline]=\"overflowItem\"\n      href=\"mailto:{{ email }}\"\n      target=\"_self\">\n      <div class=\"thf-page-blocked-user-contact-group-item\">\n        <span class=\"thf-page-blocked-user-contact-icon thf-icon thf-icon-mail thf-pr-1 \"></span>\n        <span #mailItem class=\"thf-page-blocked-user-contact-text thf-font-text\">{{ email }}</span>\n      </div>\n    </a>\n  </div>\n</div>"
            }] }
];
/** @nocollapse */
ThfPageBlockedUserContactsComponent.ctorParameters = () => [
    { type: ChangeDetectorRef }
];
ThfPageBlockedUserContactsComponent.propDecorators = {
    email: [{ type: Input, args: ['t-email',] }],
    phone: [{ type: Input, args: ['t-phone',] }],
    contactGroup: [{ type: ViewChild, args: ['contactGroup', { static: true },] }],
    mailItem: [{ type: ViewChild, args: ['mailItem', { static: true },] }],
    phoneItem: [{ type: ViewChild, args: ['phoneItem', { static: true },] }]
};
if (false) {
    /** @type {?} */
    ThfPageBlockedUserContactsComponent.prototype.literals;
    /** @type {?} */
    ThfPageBlockedUserContactsComponent.prototype.overflowItem;
    /**
     * @type {?}
     * @private
     */
    ThfPageBlockedUserContactsComponent.prototype.mailText;
    /**
     * @type {?}
     * @private
     */
    ThfPageBlockedUserContactsComponent.prototype.phoneText;
    /** @type {?} */
    ThfPageBlockedUserContactsComponent.prototype.email;
    /** @type {?} */
    ThfPageBlockedUserContactsComponent.prototype.phone;
    /** @type {?} */
    ThfPageBlockedUserContactsComponent.prototype.contactGroup;
    /** @type {?} */
    ThfPageBlockedUserContactsComponent.prototype.mailItem;
    /** @type {?} */
    ThfPageBlockedUserContactsComponent.prototype.phoneItem;
    /**
     * @type {?}
     * @private
     */
    ThfPageBlockedUserContactsComponent.prototype.changeDetector;
}
