// @totvs/thf-ui/esm2015/lib/components/thf-tabs/thf-tabs.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ChangeDetectorRef, Component, ContentChildren, QueryList, ViewChild } from '@angular/core';
import { ThfTabComponent } from './thf-tab/thf-tab.component';
import { ThfTabDropdownComponent } from './thf-tab-dropdown/thf-tab-dropdown.component';
import { ThfTabsBaseComponent } from './thf-tabs-base.component';
/** @type {?} */
const thfTabsMaxNumberOfTabs = 5;
/**
 * \@docsExtends ThfTabsBaseComponent
 *
 * \@example
 *
 * <example name="thf-tabs-basic" title="Totvs Tabs Basic">
 *  <file name="sample-thf-tabs-basic/sample-thf-tabs-basic.component.html"> </file>
 *  <file name="sample-thf-tabs-basic/sample-thf-tabs-basic.component.ts"> </file>
 * </example>
 *
 * <example name="thf-tabs-labs" title="Totvs Tabs Labs">
 *  <file name="sample-thf-tabs-labs/sample-thf-tabs-labs.component.html"> </file>
 *  <file name="sample-thf-tabs-labs/sample-thf-tabs-labs.component.ts"> </file>
 * </example>
 *
 * <example name="thf-tabs-travel" title="Totvs Tabs - Travel">
 *  <file name="sample-thf-tabs-travel/sample-thf-tabs-travel.component.html"> </file>
 *  <file name="sample-thf-tabs-travel/sample-thf-tabs-travel.component.ts"> </file>
 * </example>
 *
 * <example name="thf-tabs-business-conf" title="Totvs Tabs - Business Conference">
 *  <file name="sample-thf-tabs-business-conf/sample-thf-tabs-business-conf.component.html"> </file>
 *  <file name="sample-thf-tabs-business-conf/sample-thf-tabs-business-conf.component.ts"> </file>
 * </example>
 */
export class ThfTabsComponent extends ThfTabsBaseComponent {
    /**
     * @param {?} changeDetector
     */
    constructor(changeDetector) {
        super();
        this.changeDetector = changeDetector;
        this.maxNumberOfTabs = thfTabsMaxNumberOfTabs;
    }
    /**
     * @return {?}
     */
    get isShowTabDropdown() {
        return this.visibleTabs.length > this.maxNumberOfTabs;
    }
    // tabs que serão apresentadas na aba "Mais"
    /**
     * @return {?}
     */
    get overflowedTabs() {
        return this.visibleTabs.filter((/**
         * @param {?} _tab
         * @param {?} index
         * @return {?}
         */
        (_tab, index) => index > (this.maxNumberOfTabs - 2)));
    }
    /**
     * @return {?}
     */
    get visibleTabs() {
        return this.tabs.filter((/**
         * @param {?} tab
         * @return {?}
         */
        tab => !tab.hide));
    }
    /**
     * @return {?}
     */
    closePopover() {
        /** @type {?} */
        const containsPopoverVisible = this.tabDropdown && this.tabDropdown.popover && !this.tabDropdown.popover.isHidden;
        if (containsPopoverVisible) {
            this.tabDropdown.popover.close();
        }
    }
    /**
     * @param {?} tab
     * @return {?}
     */
    isVisibleTab(tab) {
        /** @type {?} */
        const visibleTabIndex = this.visibleTabs.findIndex((/**
         * @param {?} visibleTab
         * @return {?}
         */
        visibleTab => visibleTab.id === tab.id));
        return (this.visibleTabs.length <= this.maxNumberOfTabs) || (visibleTabIndex < (this.maxNumberOfTabs - 1));
    }
    // Função disparada quando alguma tab ficar ativa
    /**
     * @param {?} tab
     * @return {?}
     */
    onTabActive(tab) {
        this.previousActiveTab = this.tabs.find((/**
         * @param {?} tabChild
         * @return {?}
         */
        tabChild => tabChild.active && tabChild.id !== tab.id));
        this.deactivateTab();
    }
    // funcao será disparada quando mudar o estado do thfTab para desabilitado ou escondido.
    /**
     * @param {?} tab
     * @return {?}
     */
    onTabChangeState(tab) {
        if (tab.active) {
            tab.active = false;
            this.activeDistinctTab();
            this.changeDetector.detectChanges();
        }
    }
    // selectiona a aba informada por parametro, caso houver click faz a emição do evento.
    /**
     * @param {?} tab
     * @return {?}
     */
    selectedTab(tab) {
        tab.active = true;
        if (tab.click) {
            tab.click.emit(tab);
        }
        this.changeDetector.detectChanges();
    }
    /**
     * @param {?} _i
     * @param {?} tab
     * @return {?}
     */
    trackByFn(_i, tab) {
        return tab.id;
    }
    // ativa a previousActiveTab ou primeira tab encontrada.
    /**
     * @private
     * @return {?}
     */
    activeDistinctTab() {
        if (this.previousActiveTab) {
            this.previousActiveTab.active = true;
        }
        else {
            this.activeFirstTab();
        }
    }
    // Ativa a primeira Tab que não estiver desabilitada ou escondida.
    /**
     * @private
     * @return {?}
     */
    activeFirstTab() {
        this.tabs.some((/**
         * @param {?} tabChild
         * @return {?}
         */
        tabChild => {
            if (!tabChild.disabled && !tabChild.hide) {
                tabChild.active = true;
                return true;
            }
        }));
    }
    // desativa previousActiveTab e dispara a detecção de mudança.
    /**
     * @private
     * @return {?}
     */
    deactivateTab() {
        if (this.previousActiveTab) {
            this.previousActiveTab.active = false;
            this.changeDetector.detectChanges();
        }
    }
}
ThfTabsComponent.decorators = [
    { type: Component, args: [{
                selector: 'thf-tabs',
                template: "<div class=\"thf-tabs-header\">\n\n  <ng-container *ngFor=\"let tab of tabs; trackBy: trackByFn\">\n    <thf-tab-button *ngIf=\"isVisibleTab(tab)\"\n      class=\"thf-tab-button\"\n      [t-active]=\"tab.active\"\n      [t-disabled]=\"tab.disabled\"\n      [t-id]=\"tab.id\"\n      [t-hide]=\"tab.hide\"\n      [t-label]=\"tab.label\"\n      [t-small]=\"small\"\n      (keyup.enter)=\"closePopover()\"\n      (t-activated)=\"onTabActive(tab)\"\n      (t-change-state)=\"onTabChangeState(tab)\"\n      (t-click)=\"selectedTab(tab)\">\n    </thf-tab-button>\n  </ng-container>\n\n  <thf-tab-dropdown #tabDropdown\n    *ngIf=\"isShowTabDropdown\"\n    class=\"thf-tab-button thf-tab-dropdown\"\n    t-label=\"Mais\"\n    [t-small]=\"small\"\n    [t-tabs]=\"overflowedTabs\"\n    (t-activated)=\"onTabActive($event)\"\n    (t-change-state)=\"onTabChangeState($event)\"\n    (t-click)=\"selectedTab($event)\">\n  </thf-tab-dropdown>\n</div>\n\n<div class=\"thf-tabs-container\">\n  <ng-content></ng-content>\n</div>\n\n"
            }] }
];
/** @nocollapse */
ThfTabsComponent.ctorParameters = () => [
    { type: ChangeDetectorRef }
];
ThfTabsComponent.propDecorators = {
    tabs: [{ type: ContentChildren, args: [ThfTabComponent,] }],
    tabDropdown: [{ type: ViewChild, args: ['tabDropdown', { static: true },] }]
};
if (false) {
    /** @type {?} */
    ThfTabsComponent.prototype.maxNumberOfTabs;
    /**
     * @type {?}
     * @private
     */
    ThfTabsComponent.prototype.previousActiveTab;
    /** @type {?} */
    ThfTabsComponent.prototype.tabs;
    /** @type {?} */
    ThfTabsComponent.prototype.tabDropdown;
    /**
     * @type {?}
     * @private
     */
    ThfTabsComponent.prototype.changeDetector;
}
